{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conformational ensembles generation, data representation and visualization of predicted flexibility properties using BioExcel Building Blocks (biobb) and FlexDyn tools\n",
    "\n",
    "## Workflow included in the [ELIXIR 3D-Bioinfo](https://elixir-europe.org/communities/3d-bioinfo) Implementation Study:\n",
    "\n",
    "### _Building on PDBe-KB to chart and characterize the conformation landscape of native proteins_\n",
    "\n",
    "***\n",
    "\n",
    "This tutorial aims to illustrate the process of generating **protein conformational ensembles** from **3D structures** and analysing its **molecular flexibility**, step by step, using the **BioExcel Building Blocks library (biobb)**. \n",
    "\n",
    "## Conformational landscape of native proteins\n",
    "\n",
    "**Proteins** are **dynamic** systems that adopt multiple **conformational states**, a property essential for many **biological processes** (e.g. binding other proteins, nucleic acids, small molecule ligands, or switching between functionaly active and inactive states). Characterizing the different **conformational states** of proteins and the transitions between them is therefore critical for gaining insight into their **biological function** and can help explain the effects of genetic variants in **health** and **disease** and the action of drugs.\n",
    "\n",
    "**Structural biology** has become increasingly efficient in sampling the different **conformational states** of proteins. The **PDB** has currently archived more than **170,000 individual structures**, but over two thirds of these structures represent **multiple conformations** of the same or related protein, observed in different crystal forms, when interacting with other proteins or other macromolecules, or upon binding small molecule ligands. Charting this conformational diversity across the PDB can therefore be employed to build a useful approximation of the **conformational landscape** of native proteins.\n",
    "\n",
    "A number of resources and **tools** describing and characterizing various often complementary aspects of protein **conformational diversity** in known structures have been developed, notably by groups in Europe. These tools include algorithms with varying degree of sophistication, for aligning the 3D structures of individual protein chains or domains, of protein assemblies, and evaluating their degree of **structural similarity**. Using such tools one can **align structures pairwise**, compute the corresponding **similarity matrix**, and identify ensembles of **structures/conformations** with a defined **similarity level** that tend to recur in different PDB entries, an operation typically performed using **clustering** methods. Such workflows are at the basis of resources such as **CATH, Contemplate, or PDBflex** that offer access to **conformational ensembles** comprised of similar **conformations** clustered according to various criteria. Other types of tools focus on differences between **protein conformations**, identifying regions of proteins that undergo large **collective displacements** in different PDB entries, those that act as **hinges or linkers**, or regions that are inherently **flexible**.\n",
    "\n",
    "To build a meaningful approximation of the **conformational landscape** of native proteins, the **conformational ensembles** (and the differences between them), identified on the basis of **structural similarity/dissimilarity** measures alone, need to be **biophysically characterized**. This may be approached at **two different levels**. \n",
    "- At the **biological level**, it is important to link observed **conformational ensembles**, to their **functional roles** by evaluating the correspondence with **protein family classifications** based on sequence information and **functional annotations** in public databases e.g. Uniprot, PDKe-Knowledge Base (KB). These links should provide valuable mechanistic insights into how the **conformational and dynamic properties** of proteins are exploited by evolution to regulate their **biological function**. <br><br>\n",
    "\n",
    "- At the **physical level** one needs to introduce **energetic consideration** to evaluate the likelihood that the identified **conformational ensembles** represent **conformational states** that the protein (or domain under study) samples in isolation. Such evaluation is notoriously **challenging** and can only be roughly approximated by using **computational methods** to evaluate the extent to which the observed **conformational ensembles** can be reproduced by algorithms that simulate the **dynamic behavior** of protein systems. These algorithms include the computationally expensive **classical molecular dynamics (MD) simulations** to sample local thermal fluctuations but also faster more approximate methods such as **Elastic Network Models** and **Normal Node Analysis** (NMA) to model low energy **collective motions**. Alternatively, **enhanced sampling molecular dynamics** can be used to model complex types of **conformational changes** but at a very high computational cost. \n",
    "\n",
    "The **ELIXIR 3D-Bioinfo Implementation Study** *Building on PDBe-KB to chart and characterize the conformation landscape of native proteins* focuses on:\n",
    "\n",
    "1. Mapping the **conformational diversity** of proteins and their homologs across the PDB. \n",
    "2. Characterize the different **flexibility properties** of protein regions, and link this information to sequence and functional annotation.\n",
    "3. Benchmark **computational methods** that can predict a biophysical description of protein motions.\n",
    "\n",
    "This notebook is part of the third objective, where a list of **computational resources** that are able to predict **protein flexibility** and **conformational ensembles** have been collected, evaluated, and integrated in reproducible and interoperable workflows using the **BioExcel Building Blocks library**. Note that the list is not meant to be exhaustive, it is built following the expertise of the implementation study partners.\n",
    "\n",
    "The list of the selected **tools** is given in the following table, classified based on the underlying **theoretical methods types**, and presented together with the corresponding publication DOIs and URLs:\n",
    "\n",
    "\n",
    "| Tool | URL | Reference | Conda | Type\n",
    "| :-: | :-: | :-: | :-: | :-: |\n",
    "| **Concoord** | [URL](https://www3.mpibpc.mpg.de/groups/de_groot/concoord/concoord.html)| [Reference](https://doi.org/10.1002/(sici)1097-0134(199710)29:2%3C240::aid-prot11%3E3.0.co;2-o) | [Conda](https://anaconda.org/bioconda/concoord) | Atomistic intra-molecular interactions |\n",
    "| **ProDy** | [URL](https://prody.csb.pitt.edu/) | [Reference](https://doi.org/10.1093/bioinformatics/btab187) | [Conda](https://anaconda.org/conda-forge/prody) | Vibrational Analysis\n",
    "| **FlexServ** | [URL](https://mmb.irbbarcelona.org/FlexServ/) | [Reference](https://doi.org/10.1093/bioinformatics/btp304) | [Conda](https://anaconda.org/bioconda/flexserv) | Vibrational Analysis, Coarse-Grained MD\n",
    "| | | | | |\n",
    "| **NOLB** | [URL](https://team.inria.fr/nano-d/software/nolb-normal-modes) | [Reference](https://doi.org/10.1101/2020.12.31.424963) | [Conda](https://anaconda.org/bioconda/nolb) | Vibrational Analysis, Atomistic intra-molecular interactions\n",
    "| **iMod** | [URL](https://imods.chaconlab.org/) | [Reference](https://doi.org/10.1093/nar/gku339) | [Conda](https://anaconda.org/bioconda/imods) | Vibrational Analysis, Atomistic intra-molecular interactions\n",
    "\n",
    "where the **theoretical methods types** are:\n",
    "\n",
    "* **Vibrational analysis**: \n",
    "Tools computing the vibrational normal modes of protein three-dimensional structures, taking as input PDB or mmCIF files. The movements associated with the vibrational modes (i.e. the eigenvectors) are known to be good descriptors of proteins dynamics, and of their flexibility. The type of output delivered by the different modes varies from protein motions to conformational ensembles.\n",
    "\n",
    "\n",
    "* **Coarse-grained molecular simulations**: \n",
    "Tools in this category make use of coarse-grained representations of the proteins and of molecular simulation techniques to generate conformational ensembles.   \n",
    "\n",
    "\n",
    "* **Atomistic intra-molecular interactions**: \n",
    "Such tools make use of a potential to represent intra-molecular interactions and predict likely conformational changes in a protein structure.\n",
    "\n",
    "The notebook is divided in two main blocks:\n",
    "\n",
    "* **Conformational ensemble generation**: where the different selected methods are used to build **conformational ensembles** that are then visualized using the NGL viewer.\n",
    "* **Macromolecular flexibility analysis**: where the generated ensembles are analysed to extract **flexibility properties** that will be used in subsequent comparisons with the **conformational diversity** and **flexiblity** observed in the **PDB database**. \n",
    "\n",
    "***\n",
    "\n",
    "## Settings\n",
    "\n",
    "### Biobb modules used\n",
    "\n",
    " - [biobb_flexserv](https://github.com/bioexcel/biobb_flexserv): Tools to compute biomolecular flexibility on protein 3D structures.\n",
    " - [biobb_flexdyn](https://github.com/bioexcel/biobb_flexdyn): Tools to study the conformational landscape of native proteins.\n",
    " - [biobb_io](https://github.com/bioexcel/biobb_io): Tools to fetch biomolecular data from public databases.\n",
    " - [biobb_structure_utils](https://github.com/bioexcel/biobb_structure_utils): Tools to modify or extract information from a PDB structure.\n",
    " - [biobb_analysis](https://github.com/bioexcel/biobb_analysis): Tools to analyse Molecular Dynamics trajectories.\n",
    " - [biobb_gromacs](https://github.com/bioexcel/biobb_gromacs): Tools to setup and run Molecular Dynamics simulations with GROMACS MD package.\n",
    "  \n",
    "### Auxiliary libraries used\n",
    "\n",
    "* [jupyter](https://jupyter.org/): Free software, open standards, and web services for interactive computing across all programming languages.\n",
    "* [plotly](https://plot.ly/python/offline/): Python interactive graphing library integrated in Jupyter notebooks.\n",
    "* [nglview](https://nglviewer.org/#nglview): Jupyter/IPython widget to interactively view molecular structures and trajectories in notebooks.\n",
    "* [simpletraj](https://github.com/arose/simpletraj): Lightweight coordinate-only trajectory reader based on code from GROMACS, MDAnalysis and VMD.\n",
    "* [pandas](https://pandas.pydata.org/): Open source data analysis and manipulation tool, built on top of the Python programming language.\n",
    "\n",
    "### Conda Installation and Launch\n",
    "\n",
    "Take into account that, for this specific workflow, there are two environment files, one for linux OS and the other for mac OS:\n",
    "\n",
    "```console\n",
    "git clone https://github.com/bioexcel/biobb_wf_flexdyn.git\n",
    "cd biobb_wf_flexdyn\n",
    "conda env create -f conda_env/environment.yml\n",
    "conda activate biobb_wf_flexdyn\n",
    "jupyter-notebook biobb_wf_flexdyn/notebooks/biobb_wf_flexdyn.ipynb\n",
    "```\n",
    "\n",
    "***\n",
    "## Pipeline steps\n",
    " 1. [Input Parameters](#input)\n",
    " 2. [Fetching PDB structure](#fetch)\n",
    " 3. [Conformational Ensemble Generation](#ensemble)\n",
    "  1. [CONCOORD](#concoord)\n",
    "  2. [PRODY](#prody)\n",
    "  3. [FLEXSERV](#flexserv)\n",
    "       1. [Brownian Dynamics (BD)](#bd)\n",
    "       2. [Discrete Molecular Dynamics (DMD)](#dmd)\n",
    "       3. [Normal Model Analysis (NMA)](#nma)\n",
    "  4. [NOLB](#nolb)\n",
    "  5. [iMOD](#imod)\n",
    " 4. [Meta-trajectory & Clustering](#metatraj)  \n",
    " 5. [Macromolecular Flexibility Analyses](#flexibility)\n",
    "  1. [Principal Component Analysis](#pca)\n",
    "  2. [B-Factor x Principal Components](#bfactor)\n",
    "  3. [Hinge Points Prediction](#hinges)\n",
    "  4. [Apparent Stiffness](#stiffness)\n",
    "  5. [Collectivity index](#collectivity)\n",
    " \n",
    "\n",
    "***\n",
    "<table>\n",
    "    <tr style=\"background: white;\"><td>\n",
    "        <img src=\"https://bioexcel.eu/wp-content/uploads/2019/04/Bioexcell_logo_1080px_transp.png\" alt=\"Bioexcel2 logo\"\n",
    "            title=\"Bioexcel logo\" width=\"400\" />\n",
    "    </td>\n",
    "        <td style=\"width: 100px;\"></td>\n",
    "    <td>\n",
    "        <img src=\"https://cdn.eventscase.com/elixirevents.eventscase.com/uploads/users/711598/uploads/d0c5cffb09bfcac4df9bea35a8bc66bc225bf1a59701eb40a86d0449c81671cfc9cbb1163a5bc53d2dbd89a2e16bc4c0cb7d.610bafb9be1e9.png\" alt=\"3D-Bioinfo logo\"\n",
    "            title=\"3D-Bioinfo logo\" width=\"250\" />\n",
    "    </td></tr>\n",
    "</table>\n",
    "\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initializing colab\n",
    "\n",
    "The three cells below are used only in case this notebook is executed via **Google Colab**. Take into account that, for running conda on **Google Colab**, the **condacolab** library must be installed. As [explained here](https://pypi.org/project/condacolab/), the installation requires a **kernel restart**, so when running this notebook in **Google Colab**, don't run all cells until this **installation** is properly **finished** and the **kernel** has **restarted**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only executed when using google colab\n",
    "import sys\n",
    "if 'google.colab' in sys.modules:\n",
    "  !git clone https://github.com/bioexcel/biobb_wf_flexdyn.git\n",
    "  !pip install -q condacolab\n",
    "  import condacolab\n",
    "  condacolab.install()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: \n",
    "  # Install conda environment\n",
    "  !mamba env update -n base -f biobb_wf_flexdyn/conda_env/environment.yml\n",
    "  # Fix Python path to prioritize conda packages\n",
    "  conda_path = '/usr/local'\n",
    "  sys.path.insert(0, f'{conda_path}/lib/python3.11/site-packages')\n",
    "  # Enable widgets for colab\n",
    "  from google.colab import output \n",
    "  output.enable_custom_widget_manager() \n",
    "  # Change working dir\n",
    "  import os \n",
    "  os.chdir(\"biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks\") \n",
    "  print(f\"New working directory: {os.getcwd()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store original Colab paths\n",
    "ORIGINAL_PATHS = [p for p in sys.path if '/usr/local/lib/python3.12' in p or 'dist-packages' in p]\n",
    "CONDA_PATHS = [p for p in sys.path if '/usr/local/lib/python3.11' in p]\n",
    "\n",
    "def use_colab_plotly():\n",
    "    \"\"\"Switch to Colab's original plotly\"\"\"\n",
    "    # Move Colab paths to the front\n",
    "    for path in reversed(ORIGINAL_PATHS):\n",
    "        if path in sys.path:\n",
    "            sys.path.remove(path)\n",
    "        sys.path.insert(0, path)\n",
    "    \n",
    "    # Reload plotly and nbformat from Colab's environment\n",
    "    if 'plotly' in sys.modules:\n",
    "        del sys.modules['plotly']\n",
    "        del sys.modules['plotly.graph_objs']\n",
    "        # del sys.modules['plotly.io']\n",
    "        # Delete all plotly submodules\n",
    "        to_delete = [m for m in sys.modules if m.startswith('plotly')]\n",
    "        for m in to_delete:\n",
    "            del sys.modules[m]\n",
    "    if 'nbformat' in sys.modules:\n",
    "        del sys.modules['nbformat']\n",
    "\n",
    "def use_conda_env():\n",
    "    \"\"\"Switch back to conda environment\"\"\"\n",
    "    # Move conda paths to the front\n",
    "    for path in reversed(CONDA_PATHS):\n",
    "        if path in sys.path:\n",
    "            sys.path.remove(path)\n",
    "        sys.path.insert(0, path)\n",
    "    \n",
    "    # Clean up Colab's plotly/nbformat so conda versions load next time\n",
    "    if 'plotly' in sys.modules:\n",
    "        to_delete = [m for m in sys.modules if m.startswith('plotly')]\n",
    "        for m in to_delete:\n",
    "            del sys.modules[m]\n",
    "    if 'nbformat' in sys.modules:\n",
    "        del sys.modules['nbformat']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"input\"></a>\n",
    "## Input parameters\n",
    "**Input parameters** needed:\n",
    "\n",
    " - **Auxiliary libraries**: Libraries to be used in the workflow are imported once at the beginning\n",
    "<br><br>\n",
    " - **pdbCode**: PDB code of the protein structure (e.g. 1AKE, [https://doi.org/10.2210/pdb1AKE/pdb](https://doi.org/10.2210/pdb1AKE/pdb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import nglview\n",
    "import simpletraj\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import ipywidgets\n",
    "import json\n",
    "import zipfile\n",
    "from IPython.display import display, Markdown\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "pdbCode = \"1ake\"\n",
    "num_frames = 300"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"fetch\"></a>\n",
    "***\n",
    "## Fetching PDB structure\n",
    "Downloading **PDB structure** with the **protein molecule** from the RCSB PDB database.<br>\n",
    "Alternatively, a **PDB file** can be used as starting structure. <br>\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [Pdb](https://biobb-io.readthedocs.io/en/latest/api.html#module-api.pdb) from **biobb_io.api.pdb**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:28:39,489 [MainThread  ] [INFO ]  Module: biobb_io.api.pdb Version: 5.2.0\n",
      "2026-02-12 09:28:39,490 [MainThread  ] [INFO ]  Downloading 1ake from: https://www.ebi.ac.uk/pdbe/entry-files/download/pdb1ake.ent\n",
      "2026-02-12 09:28:39,854 [MainThread  ] [INFO ]  Writting pdb to: 1ake.pdb\n",
      "2026-02-12 09:28:39,857 [MainThread  ] [INFO ]  Filtering lines NOT starting with one of these words: ['ATOM', 'MODEL', 'ENDMDL']\n",
      "2026-02-12 09:28:39,886 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Downloading desired PDB file \n",
    "# Import module\n",
    "from biobb_io.api.pdb import pdb\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "downloaded_pdb = pdbCode+'.pdb'\n",
    "\n",
    "prop = {\n",
    "    'pdb_code': pdbCode,\n",
    "    'api_id' : 'pdbe'\n",
    "}\n",
    "\n",
    "#Create and launch bb\n",
    "pdb(output_pdb_path=downloaded_pdb,\n",
    "    properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Selecting a model\n",
    "If needed, select a single **model** from the **PDB file**. <br>\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [extract_model](https://biobb-structure-utils.readthedocs.io/en/latest/utils.html#module-utils.extract_model) from **biobb_structure_utils.utils.extract_model**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:28:42,147 [MainThread  ] [INFO ]  Module: biobb_structure_utils.utils.extract_model Version: 5.2.0\n",
      "2026-02-12 09:28:42,151 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d3796c0-6ea1-493e-a472-4e4e43b0974a\n",
      "2026-02-12 09:28:42,152 [MainThread  ] [INFO ]  Copy to stage: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/1ake.pdb --> sandbox_8d3796c0-6ea1-493e-a472-4e4e43b0974a\n",
      "2026-02-12 09:28:42,155 [MainThread  ] [INFO ]  Creating 49d6169a-77be-49c4-b04d-b23ea24200a3 temporary folder\n",
      "2026-02-12 09:28:42,156 [MainThread  ] [INFO ]  Launching command (it may take a while): check_structure -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d3796c0-6ea1-493e-a472-4e4e43b0974a/1ake.pdb -o 49d6169a-77be-49c4-b04d-b23ea24200a3/model1.pdb --force_save models --select 1\n",
      "2026-02-12 09:28:43,220 [MainThread  ] [INFO ]  Command 'check_structure -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d3796c0-6ea1-493e-a472-4e4e43b0974a/1ake.pdb -o 49d6169a-77be-49c4-b04d-b23ea24200a3/model1.pdb --force_save models --select 1' finalized with exit code 0\n",
      "2026-02-12 09:28:43,222 [MainThread  ] [INFO ]  ================================================================================\n",
      "=                   BioBB structure checking utility v3.15.6                   =\n",
      "=            P. Andrio, A. Hospital, G. Bayarri, J.L. Gelpi 2018-23            =\n",
      "================================================================================\n",
      "\n",
      "Warning: sequence features may not be available, use --sequence for  external fasta input\n",
      "Structure /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d3796c0-6ea1-493e-a472-4e4e43b0974a/1ake.pdb loaded\n",
      " PDB id:  \n",
      " Title: \n",
      " Experimental method: unknown\n",
      " Resolution (A): N.A.\n",
      "\n",
      " Num. models: 1\n",
      " Num. chains: 2 (A: Protein, B: Protein)\n",
      " Num. residues:  428\n",
      " Num. residues with ins. codes:  0\n",
      " Num. residues with H atoms: 0\n",
      " Num. HETATM residues:  0\n",
      " Num. ligands or modified residues:  0\n",
      " Num. water mol.:  0\n",
      " Num. atoms:  3312\n",
      "Running models. Options: --select 1\n",
      "Detected 1 Model(s)\n",
      "Found Single model\n",
      "Structure not modified, saving due to --force_save option\n",
      "Final Num. models: 1\n",
      "Final Num. chains: 2 (A: Protein, B: Protein)\n",
      "Final Num. residues:  428\n",
      "Final Num. residues with ins. codes:  0\n",
      "Final Num. residues with H atoms: 0\n",
      "Final Num. HETATM residues:  0\n",
      "Final Num. ligands or modified residues:  0\n",
      "Final Num. water mol.:  0\n",
      "Final Num. atoms:  3312\n",
      "Structure saved on 49d6169a-77be-49c4-b04d-b23ea24200a3/model1.pdb\n",
      "\n",
      "2026-02-12 09:28:43,225 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d3796c0-6ea1-493e-a472-4e4e43b0974a', '49d6169a-77be-49c4-b04d-b23ea24200a3']\n",
      "2026-02-12 09:28:43,225 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_structure_utils.utils.extract_model import extract_model\n",
    "\n",
    "pdb_model = pdbCode+'_model.pdb'\n",
    "\n",
    "prop = {\n",
    "    'models': [ 1 ]\n",
    "}\n",
    "\n",
    "extract_model(input_structure_path=downloaded_pdb,\n",
    "              output_structure_path=pdb_model,\n",
    "              properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Selecting a monomer\n",
    "If needed, select a single **monomer** (chain) from the **PDB file**. <br>\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [extract_chain](https://biobb-structure-utils.readthedocs.io/en/latest/utils.html#module-utils.extract_chain) from **biobb_structure_utils.utils.extract_chain**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:28:46,445 [MainThread  ] [INFO ]  Module: biobb_structure_utils.utils.extract_chain Version: 5.2.0\n",
      "2026-02-12 09:28:46,450 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_462c5826-0f32-47d7-a72f-ad946ebcf4a0\n",
      "2026-02-12 09:28:46,452 [MainThread  ] [INFO ]  Copy to stage: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/1ake_model.pdb --> sandbox_462c5826-0f32-47d7-a72f-ad946ebcf4a0\n",
      "2026-02-12 09:28:46,455 [MainThread  ] [INFO ]  Selected Chains: A\n",
      "2026-02-12 09:28:46,456 [MainThread  ] [INFO ]  Launching command (it may take a while): check_structure -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/1ake_model.pdb -o 1ake_monomer.pdb --force_save chains --select A\n",
      "2026-02-12 09:28:46,819 [MainThread  ] [INFO ]  Command 'check_structure -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/1ake_model.pdb -o 1ake_monomer.pdb --force_save chains --select A' finalized with exit code 0\n",
      "2026-02-12 09:28:46,820 [MainThread  ] [INFO ]  ================================================================================\n",
      "=                   BioBB structure checking utility v3.15.6                   =\n",
      "=            P. Andrio, A. Hospital, G. Bayarri, J.L. Gelpi 2018-23            =\n",
      "================================================================================\n",
      "\n",
      "Warning: sequence features may not be available, use --sequence for  external fasta input\n",
      "Structure /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/1ake_model.pdb loaded\n",
      " PDB id:  \n",
      " Title: \n",
      " Experimental method: unknown\n",
      " Resolution (A): N.A.\n",
      "\n",
      " Num. models: 1\n",
      " Num. chains: 2 (A: Protein, B: Protein)\n",
      " Num. residues:  428\n",
      " Num. residues with ins. codes:  0\n",
      " Num. residues with H atoms: 0\n",
      " Num. HETATM residues:  0\n",
      " Num. ligands or modified residues:  0\n",
      " Num. water mol.:  0\n",
      " Num. atoms:  3312\n",
      "Running chains. Options: --select A\n",
      "Detected 2 Chain(s)\n",
      " A: Protein\n",
      " B: Protein\n",
      "Selecting chain(s) A\n",
      "Final Num. models: 1\n",
      "Final Num. chains: 1 (A: Protein)\n",
      "Final Num. residues:  214\n",
      "Final Num. residues with ins. codes:  0\n",
      "Final Num. residues with H atoms: 0\n",
      "Final Num. HETATM residues:  0\n",
      "Final Num. ligands or modified residues:  0\n",
      "Final Num. water mol.:  0\n",
      "Final Num. atoms:  1656\n",
      "Structure saved on 1ake_monomer.pdb\n",
      "\n",
      "2026-02-12 09:28:46,821 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_462c5826-0f32-47d7-a72f-ad946ebcf4a0']\n",
      "2026-02-12 09:28:46,822 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_structure_utils.utils.extract_chain import extract_chain\n",
    "\n",
    "monomer = pdbCode+'_monomer.pdb'\n",
    "\n",
    "prop = {\n",
    "    'chains': [ 'A' ]\n",
    "}\n",
    "\n",
    "extract_chain(input_structure_path=pdb_model,\n",
    "            output_structure_path=monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the protein\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b0f214be4db444a1bb9444ff98424d01",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Show protein\n",
    "view = nglview.show_structure_file(monomer)\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Generating the reduced (coarse-grained) structure (backbone)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_mask](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_mask) from **biobb_analysis.ambertools.cpptraj_mask**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:28:53,037 [MainThread  ] [INFO ]  Module: biobb_analysis.ambertools.cpptraj_mask Version: 5.2.0\n",
      "2026-02-12 09:28:53,038 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea\n",
      "2026-02-12 09:28:53,040 [MainThread  ] [INFO ]  Copy to stage: 1ake_monomer.pdb --> sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea\n",
      "2026-02-12 09:28:53,042 [MainThread  ] [INFO ]  Copy to stage: 1ake_monomer.pdb --> sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea\n",
      "2026-02-12 09:28:53,043 [MainThread  ] [INFO ]  Launching command (it may take a while): cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/cbc2c60f-db06-4c87-954b-a20677bce90einstructions.in\n",
      "2026-02-12 09:29:01,052 [MainThread  ] [INFO ]  Command 'cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/cbc2c60f-db06-4c87-954b-a20677bce90einstructions.in' finalized with exit code 0\n",
      "2026-02-12 09:29:01,054 [MainThread  ] [INFO ]  \n",
      "CPPTRAJ: Trajectory Analysis. V6.18.1 (AmberTools)\n",
      "    ___  ___  ___  ___\n",
      "     | \\/ | \\/ | \\/ | \n",
      "    _|_/\\_|_/\\_|_/\\_|_\n",
      "\n",
      "| Date/time: 02/12/26 09:29:00\n",
      "| Available memory: 8.497 GB\n",
      "\n",
      "INPUT: Reading input from '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/cbc2c60f-db06-4c87-954b-a20677bce90einstructions.in'\n",
      "  [parm /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_monomer.pdb]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_monomer.pdb' as PDB File\n",
      "\tReading bond info from CONECT records.\n",
      "\tNot reading bond info from LINK records.\n",
      "Warning: Atom 1288 CD in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1289 CD in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1290 NE in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1291 NE in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1292 CZ in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1293 CZ in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1294 NH1 in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1295 NH1 in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1296 NH2 in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1297 NH2 in res ARG 167 A has alternate location specifier B\n",
      "\tDetermining bond info from distances.\n",
      "Warning: Determining bond length parameters from element types for '1ake_monomer.pdb'.\n",
      "  [trajin /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_monomer.pdb 1 -1 1]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_monomer.pdb' as PDB\n",
      "Warning: PDB has 10 records wih alternate atom location IDs.\n",
      "  [strip !@C,CA,N,O,C3',O3',C4',C5',O5',P]\n",
      "    STRIP: Stripping atoms in mask [!@C,CA,N,O,C3',O3',C4',C5',O5',P]\n",
      "  [trajout /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_backbone.pdb pdb]\n",
      "\tWriting '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_backbone.pdb' as PDB\n",
      "---------- RUN BEGIN -------------------------------------------------\n",
      "\n",
      "PARAMETER FILES (1 total):\n",
      " 0: 1ake_monomer.pdb, 1661 atoms, 214 res, box: None, 1 mol\n",
      "\n",
      "INPUT TRAJECTORIES (1 total):\n",
      " 0: '1ake_monomer.pdb' is a PDB file, Parm 1ake_monomer.pdb (reading 1 of 1)\n",
      "  Coordinate processing will occur on 1 frames.\n",
      "\n",
      "OUTPUT TRAJECTORIES (1 total):\n",
      "  '1ake_backbone.pdb' (1 frames) is a PDB file\n",
      "\n",
      "BEGIN TRAJECTORY PROCESSING:\n",
      ".....................................................\n",
      "ACTION SETUP FOR PARM '1ake_monomer.pdb' (1 actions):\n",
      "  0: [strip !@C,CA,N,O,C3',O3',C4',C5',O5',P]\n",
      "\tStripping 805 atoms.\n",
      "\tStripped topology: 856 atoms, 214 res, box: None, 1 mol\n",
      ".....................................................\n",
      "ACTIVE OUTPUT TRAJECTORIES (1):\n",
      "  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea/1ake_backbone.pdb (coordinates)\n",
      "----- 1ake_monomer.pdb (1-1, 1) -----\n",
      "100% Complete.\n",
      "\n",
      "Read 1 frames and processed 1 frames.\n",
      "TIME: Avg. throughput= 408.3299 frames / second.\n",
      "\n",
      "ACTION OUTPUT:\n",
      "TIME: Analyses took 0.0000 seconds.\n",
      "\n",
      "RUN TIMING:\n",
      "TIME:\t\tInit               : 0.0000 s (  0.62%)\n",
      "TIME:\t\tTrajectory Process : 0.0024 s ( 94.96%)\n",
      "TIME:\t\tAction Post        : 0.0000 s (  0.04%)\n",
      "TIME:\t\tAnalysis           : 0.0000 s (  0.08%)\n",
      "TIME:\t\tData File Write    : 0.0000 s (  0.08%)\n",
      "TIME:\t\tOther              : 0.0001 s (  0.04%)\n",
      "TIME:\tRun Total 0.0026 s\n",
      "---------- RUN END ---------------------------------------------------\n",
      "TIME: Total execution time: 0.0996 seconds.\n",
      "--------------------------------------------------------------------------------\n",
      "To cite CPPTRAJ use:\n",
      "Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for\n",
      "  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.\n",
      "  Theory Comput., 2013, 9 (7), pp 3084-3095.\n",
      "\n",
      "\n",
      "2026-02-12 09:29:01,056 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_f1df24d6-317f-4c63-97c7-d637710e23ea']\n",
      "2026-02-12 09:29:01,056 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_mask import cpptraj_mask\n",
    "\n",
    "prot_backbone = pdbCode + \"_backbone.pdb\"\n",
    "\n",
    "prop = {\n",
    "    'mask': 'backbone',\n",
    "    'format': 'pdb'\n",
    "}\n",
    "\n",
    "cpptraj_mask(input_top_path=monomer,\n",
    "            input_traj_path=monomer,\n",
    "            output_cpptraj_path=prot_backbone,\n",
    "            properties=prop)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the protein\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4930dbf681194a949f4d6cd0b1219e25",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Show protein\n",
    "view = nglview.show_structure_file(prot_backbone)\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Generating the reduced (coarse-grained) structure (alpha carbons)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_mask](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_mask) from **biobb_analysis.ambertools.cpptraj_mask**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:29:15,015 [MainThread  ] [INFO ]  Module: biobb_analysis.ambertools.cpptraj_mask Version: 5.2.0\n",
      "2026-02-12 09:29:15,018 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927\n",
      "2026-02-12 09:29:15,019 [MainThread  ] [INFO ]  Copy to stage: 1ake_monomer.pdb --> sandbox_7281d602-50bc-46d1-84d1-7cc89da67927\n",
      "2026-02-12 09:29:15,023 [MainThread  ] [INFO ]  Copy to stage: 1ake_monomer.pdb --> sandbox_7281d602-50bc-46d1-84d1-7cc89da67927\n",
      "2026-02-12 09:29:15,026 [MainThread  ] [INFO ]  Launching command (it may take a while): cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/bd769cbf-5d0a-4f1a-ba5f-56317c1c7dd9instructions.in\n",
      "2026-02-12 09:29:15,194 [MainThread  ] [INFO ]  Command 'cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/bd769cbf-5d0a-4f1a-ba5f-56317c1c7dd9instructions.in' finalized with exit code 0\n",
      "2026-02-12 09:29:15,196 [MainThread  ] [INFO ]  \n",
      "CPPTRAJ: Trajectory Analysis. V6.18.1 (AmberTools)\n",
      "    ___  ___  ___  ___\n",
      "     | \\/ | \\/ | \\/ | \n",
      "    _|_/\\_|_/\\_|_/\\_|_\n",
      "\n",
      "| Date/time: 02/12/26 09:29:15\n",
      "| Available memory: 8.412 GB\n",
      "\n",
      "INPUT: Reading input from '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/bd769cbf-5d0a-4f1a-ba5f-56317c1c7dd9instructions.in'\n",
      "  [parm /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_monomer.pdb]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_monomer.pdb' as PDB File\n",
      "\tReading bond info from CONECT records.\n",
      "\tNot reading bond info from LINK records.\n",
      "Warning: Atom 1288 CD in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1289 CD in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1290 NE in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1291 NE in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1292 CZ in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1293 CZ in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1294 NH1 in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1295 NH1 in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1296 NH2 in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1297 NH2 in res ARG 167 A has alternate location specifier B\n",
      "\tDetermining bond info from distances.\n",
      "Warning: Determining bond length parameters from element types for '1ake_monomer.pdb'.\n",
      "  [trajin /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_monomer.pdb 1 -1 1]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_monomer.pdb' as PDB\n",
      "Warning: PDB has 10 records wih alternate atom location IDs.\n",
      "  [strip !@CA]\n",
      "    STRIP: Stripping atoms in mask [!@CA]\n",
      "  [trajout /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_ca.pdb pdb]\n",
      "\tWriting '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_ca.pdb' as PDB\n",
      "---------- RUN BEGIN -------------------------------------------------\n",
      "\n",
      "PARAMETER FILES (1 total):\n",
      " 0: 1ake_monomer.pdb, 1661 atoms, 214 res, box: None, 1 mol\n",
      "\n",
      "INPUT TRAJECTORIES (1 total):\n",
      " 0: '1ake_monomer.pdb' is a PDB file, Parm 1ake_monomer.pdb (reading 1 of 1)\n",
      "  Coordinate processing will occur on 1 frames.\n",
      "\n",
      "OUTPUT TRAJECTORIES (1 total):\n",
      "  '1ake_ca.pdb' (1 frames) is a PDB file\n",
      "\n",
      "BEGIN TRAJECTORY PROCESSING:\n",
      ".....................................................\n",
      "ACTION SETUP FOR PARM '1ake_monomer.pdb' (1 actions):\n",
      "  0: [strip !@CA]\n",
      "\tStripping 1447 atoms.\n",
      "\tStripped topology: 214 atoms, 214 res, box: None, 214 mol\n",
      ".....................................................\n",
      "ACTIVE OUTPUT TRAJECTORIES (1):\n",
      "  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927/1ake_ca.pdb (coordinates)\n",
      "----- 1ake_monomer.pdb (1-1, 1) -----\n",
      "100% Complete.\n",
      "\n",
      "Read 1 frames and processed 1 frames.\n",
      "TIME: Avg. throughput= 750.7508 frames / second.\n",
      "\n",
      "ACTION OUTPUT:\n",
      "TIME: Analyses took 0.0000 seconds.\n",
      "\n",
      "RUN TIMING:\n",
      "TIME:\t\tInit               : 0.0000 s (  1.08%)\n",
      "TIME:\t\tTrajectory Process : 0.0013 s ( 90.06%)\n",
      "TIME:\t\tAction Post        : 0.0000 s (  0.07%)\n",
      "TIME:\t\tAnalysis           : 0.0000 s (  0.14%)\n",
      "TIME:\t\tData File Write    : 0.0000 s (  0.14%)\n",
      "TIME:\t\tOther              : 0.0001 s (  0.09%)\n",
      "TIME:\tRun Total 0.0015 s\n",
      "---------- RUN END ---------------------------------------------------\n",
      "TIME: Total execution time: 0.0960 seconds.\n",
      "--------------------------------------------------------------------------------\n",
      "To cite CPPTRAJ use:\n",
      "Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for\n",
      "  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.\n",
      "  Theory Comput., 2013, 9 (7), pp 3084-3095.\n",
      "\n",
      "\n",
      "2026-02-12 09:29:15,198 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_7281d602-50bc-46d1-84d1-7cc89da67927']\n",
      "2026-02-12 09:29:15,199 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_mask import cpptraj_mask\n",
    "\n",
    "prot_ca = pdbCode + \"_ca.pdb\"\n",
    "\n",
    "prop = {\n",
    "    'mask': 'c-alpha',\n",
    "    'format': 'pdb'\n",
    "}\n",
    "\n",
    "cpptraj_mask(input_top_path=monomer,\n",
    "            input_traj_path=monomer,\n",
    "            output_cpptraj_path=prot_ca,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the protein\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d7157fb576834c2dbe154873da914763",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Show protein\n",
    "view = nglview.show_structure_file(prot_ca)\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ensemble\"></a>\n",
    "***\n",
    "## Conformational Ensemble Generation\n",
    "\n",
    "The next section is presenting a list of **computational resources** able to predict **protein flexibility** and **conformational ensembles** (the list is not meant to be exhaustive).\n",
    "\n",
    "The notebook is intended to give an overview of the tools and resources considered relevant for the **implementation study**, namely tools that can be used to **generate** or **characterize conformational ensembles of proteins**. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"concoord\"></a>\n",
    "***\n",
    "## CONCOORD\n",
    "\n",
    "**CONCOORD** is a method to generate **protein conformations** around a known structure based on **geometric restrictions**. **Principal component analyses** of **Molecular Dynamics** (MD) simulations of proteins have indicated that **collective degrees of freedom** dominate **protein conformational fluctuations**. These large-scale **collective motions** have been shown essential to **protein function** in a number of cases. The notion that **internal constraints** and other **configurational barriers** restrict **protein dynamics** to a limited number of **collective degrees of freedom** has led to the design of the **CONCOORD** method to predict these modes without doing explicit, more CPU intensive, **MD simulations**.\n",
    "\n",
    "The **CONCOORD** method consists of two stages: The first stage is the identification of all **interatomic interactions** in the starting structure. These interactions are divided among different categories, depending on the strength of the interaction. **Covalent bonds** form the **tightest interactions** and **long-range non-bonded interactions** are among the **weakest interactions**. Based on the strength of the interaction, a specific geometric freedom is given to each interacting pair. In this way, a set of **upper** and **lower geometric bounds** is obtained for all **interacting pairs of atoms**. The second stage consists of **generating structures** other than the starting structure that **fulfill all geometric bounds**. This is achieved by iteratively applying corrections to randomly generated coordinates untill all bounds are fulfilled.\n",
    "\n",
    "\n",
    "References: <br>\n",
    "\n",
    "**CONCOORD website**: https://www3.mpibpc.mpg.de/groups/de_groot/concoord/concoord.html\n",
    "\n",
    "**Prediction of Protein Conformational Freedom From Distance Constraints.**<br>\n",
    "*Proteins, 1997, 29(2):240-51.*<br>\n",
    "*Available at: https://doi.org/10.1002/(sici)1097-0134(199710)29:2%3C240::aid-prot11%3E3.0.co;2-o*\n",
    "\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [concoord_dist](https://biobb-flexdyn.readthedocs.io/en/latest/flexdyn.html#module-concoord.concoord_dist) from **biobb_flexdyn.concoord.concoord_dist**\n",
    " - [concoord_disco](https://biobb-flexdyn.readthedocs.io/en/latest/flexdyn.html#module-concoord.concoord_disco) from **biobb_flexdyn.concoord.concoord_disco**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:33:35,410 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_1a70d2a6-0f10-4266-bad8-f7d021f2fb5c\n",
      "2026-02-12 09:33:35,411 [MainThread  ] [INFO ]  Copy to stage: 1ake_monomer.pdb --> sandbox_1a70d2a6-0f10-4266-bad8-f7d021f2fb5c\n",
      "2026-02-12 09:33:35,414 [MainThread  ] [INFO ]  Copy to stage: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/978a6a32-0bc5-4100-a6fa-4610d563497e.stdin --> sandbox_1a70d2a6-0f10-4266-bad8-f7d021f2fb5c\n",
      "2026-02-12 09:33:35,416 [MainThread  ] [INFO ]  Launching command (it may take a while): cd  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_1a70d2a6-0f10-4266-bad8-f7d021f2fb5c ; dist -op 1ake_dist.pdb -og 1ake_dist.gro -od 1ake_dist.dat -p 1ake_monomer.pdb -c 4.0 -m 50 -damp 1.0 < 978a6a32-0bc5-4100-a6fa-4610d563497e.stdin\n",
      "2026-02-12 09:33:35,732 [MainThread  ] [INFO ]  Command 'cd  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_1a70d2a6-0f10-4266-bad8-f7d021f2fb5c ; dist -op 1ake_dist.pdb -og 1ake_dist.gro -od 1ake_dist.dat -p 1ake_monomer.pdb -c 4.0 -m 50 -damp 1.0 < 978a6a32-0bc5-4100-a6fa-4610d563497e.stdin' finalized with exit code 0\n",
      "2026-02-12 09:33:35,734 [MainThread  ] [INFO ]  Select a set of Van der Waals parameters:\n",
      "1: OPLS-UA (united atoms) parameters\n",
      "2: OPLS-AA (all atoms) parameters\n",
      "3: PROLSQ repel parameters\n",
      "4: Yamber2 parameters\n",
      "5: Li et al. parameters\n",
      "6: OPLS-X parameters (recommended for NMR structure determination)\n",
      "Selected parameter set 1\n",
      "copying /Users/gbayarri/miniconda3/envs/biobb_wf_flexdyn/share/concoord/lib/ATOMS_oplsua.DAT to ATOMS.DAT in current working directory\n",
      "copying /Users/gbayarri/miniconda3/envs/biobb_wf_flexdyn/share/concoord/lib/MARGINS_oplsua.DAT to MARGINS.DAT in current working directory\n",
      "Select a set of bond/angle parameters:\n",
      "1: Concoord default parameters\n",
      "2: Engh-Huber parameters\n",
      "Selected parameter set 1\n",
      "copying /Users/gbayarri/miniconda3/envs/biobb_wf_flexdyn/share/concoord/lib/BONDS.DAT.noeh to BONDS.DAT in current working directory\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:33:35,734 [MainThread  ] [INFO ]  remark: distance between atom    1148 and    1111 (2.61) smaller than VdWsum (3.00)    O  NH1 0.043\n",
      "remark: distance between atom     347 and     341 (3.08) smaller than VdWsum (3.50)    C  NH1 0.072\n",
      "remark: distance between atom     544 and     538 (3.11) smaller than VdWsum (3.50)    C  NH1 0.042\n",
      "remark: distance between atom     545 and     539 (2.97) smaller than VdWsum (3.38)   OC   CA 0.061\n",
      "remark: distance between atom     599 and     595 (3.22) smaller than VdWsum (3.65) CR1E    C 0.080\n",
      "remark: distance between atom     638 and     634 (3.22) smaller than VdWsum (3.65) CR1E    C 0.078\n",
      "remark: distance between atom     859 and     853 (3.14) smaller than VdWsum (3.50)    C  NH1 0.012\n",
      "remark: distance between atom    1202 and    1196 (3.19) smaller than VdWsum (3.58) CH2E  NH1 0.037\n",
      "remark: distance between atom    1324 and    1319 (3.27) smaller than VdWsum (3.65) CR1E    C 0.030\n",
      "remark: distance between atom    1401 and    1397 (3.23) smaller than VdWsum (3.65) CR1E    C 0.068\n",
      "remark: distance between atom    1529 and    1525 (2.88) smaller than VdWsum (3.36)   OC    C 0.128\n",
      "remark: distance between atom      25 and      11 (3.29) smaller than VdWsum (3.83) CH2E    C 0.191\n",
      "remark: distance between atom      58 and      55 (3.00) smaller than VdWsum (3.36)    C    O 0.001\n",
      "remark: distance between atom      59 and      54 (2.98) smaller than VdWsum (3.36)    O    C 0.027\n",
      "remark: distance between atom      63 and      59 (2.84) smaller than VdWsum (3.36)    C    O 0.167\n",
      "remark: distance between atom      67 and      60 (3.44) smaller than VdWsum (3.80) CH2P CH3E 0.006\n",
      "remark: distance between atom      83 and      80 (2.97) smaller than VdWsum (3.36)    C    O 0.036\n",
      "remark: distance between atom      92 and      84 (2.96) smaller than VdWsum (3.36)    C    O 0.046\n",
      "remark: distance between atom     112 and     104 (2.98) smaller than VdWsum (3.36)    C    O 0.024\n",
      "remark: distance between atom     117 and     113 (2.96) smaller than VdWsum (3.36)    C    O 0.043\n",
      "remark: distance between atom     137 and     127 (2.83) smaller than VdWsum (3.36)    C    O 0.171\n",
      "remark: distance between atom     145 and     138 (2.88) smaller than VdWsum (3.36)    C    O 0.124\n",
      "remark: distance between atom     160 and     156 (3.20) smaller than VdWsum (3.58)  NH1 CH2E 0.026\n",
      "remark: distance between atom     183 and     172 (2.96) smaller than VdWsum (3.36)    C    O 0.048\n",
      "remark: distance between atom     225 and     220 (2.96) smaller than VdWsum (3.36)    C    O 0.047\n",
      "remark: distance between atom     232 and     226 (2.93) smaller than VdWsum (3.36)    C    O 0.071\n",
      "remark: distance between atom     236 and     233 (2.92) smaller than VdWsum (3.36)    C    O 0.085\n",
      "remark: distance between atom     244 and     237 (2.91) smaller than VdWsum (3.36)    C    O 0.100\n",
      "remark: distance between atom     260 and     253 (2.93) smaller than VdWsum (3.36)    C    O 0.071\n",
      "remark: distance between atom     271 and     261 (2.89) smaller than VdWsum (3.36)    C    O 0.119\n",
      "remark: distance between atom     276 and     272 (2.85) smaller than VdWsum (3.36)    C    O 0.154\n",
      "remark: distance between atom     281 and     277 (2.87) smaller than VdWsum (3.36)    C    O 0.139\n",
      "remark: distance between atom     303 and     298 (3.00) smaller than VdWsum (3.36)    C    O 0.006\n",
      "remark: distance between atom     313 and     308 (2.88) smaller than VdWsum (3.36)    C    O 0.124\n",
      "remark: distance between atom     334 and     331 (2.85) smaller than VdWsum (3.36)    C    O 0.160\n",
      "remark: distance between atom     341 and     330 (2.98) smaller than VdWsum (3.50)  NH1    C 0.171\n",
      "remark: distance between atom     350 and     334 (3.13) smaller than VdWsum (3.50)  NH1    C 0.025\n",
      "remark: distance between atom     357 and     353 (2.72) smaller than VdWsum (3.36)    C    O 0.285\n",
      "remark: distance between atom     382 and     375 (2.93) smaller than VdWsum (3.36)    C    O 0.078\n",
      "remark: distance between atom     390 and     383 (2.95) smaller than VdWsum (3.36)    C    O 0.057\n",
      "remark: distance between atom     431 and     425 (2.85) smaller than VdWsum (3.36)    C    O 0.153\n",
      "remark: distance between atom     438 and     432 (2.85) smaller than VdWsum (3.36)    C    O 0.160\n",
      "remark: distance between atom     446 and     439 (2.87) smaller than VdWsum (3.36)    C    O 0.135\n",
      "remark: distance between atom     455 and     447 (2.99) smaller than VdWsum (3.36)    C    O 0.015\n",
      "remark: distance between atom     463 and     456 (2.98) smaller than VdWsum (3.36)    C    O 0.027\n",
      "remark: distance between atom     470 and     464 (2.87) smaller than VdWsum (3.36)    C    O 0.131\n",
      "remark: distance between atom     478 and     471 (3.00) smaller than VdWsum (3.36)    C    O 0.001\n",
      "remark: distance between atom     498 and     492 (2.92) smaller than VdWsum (3.36)    C    O 0.088\n",
      "remark: distance between atom     507 and     499 (2.83) smaller than VdWsum (3.36)    C    O 0.171\n",
      "remark: distance between atom     516 and     508 (2.98) smaller than VdWsum (3.36)    C    O 0.025\n",
      "remark: distance between atom     527 and     517 (2.78) smaller than VdWsum (3.36)    C    O 0.227\n",
      "remark: distance between atom     533 and     516 (2.99) smaller than VdWsum (3.50)  NH1    C 0.160\n",
      "remark: distance between atom     549 and     541 (2.90) smaller than VdWsum (3.36)    C    O 0.102\n",
      "remark: distance between atom     556 and     540 (3.11) smaller than VdWsum (3.50)  NH1    C 0.041\n",
      "remark: distance between atom     568 and     528 (3.03) smaller than VdWsum (3.38) CH1E    O 0.003\n",
      "remark: distance between atom     569 and     216 (3.17) smaller than VdWsum (3.68)    S CH3E 0.151\n",
      "remark: distance between atom     575 and     550 (2.87) smaller than VdWsum (3.43) CH2E    O 0.213\n",
      "remark: distance between atom     576 and     550 (2.96) smaller than VdWsum (3.43) CH2E    O 0.125\n",
      "remark: distance between atom     597 and     592 (3.01) smaller than VdWsum (3.43) CH2E    O 0.071\n",
      "remark: distance between atom     645 and     633 (3.35) smaller than VdWsum (3.78)    C   CA 0.070\n",
      "remark: distance between atom     667 and     652 (3.35) smaller than VdWsum (3.78) CH3E    C 0.079\n",
      "remark: distance between atom     670 and     664 (2.98) smaller than VdWsum (3.36)    C    O 0.026\n",
      "remark: distance between atom     678 and     671 (2.83) smaller than VdWsum (3.36)    C    O 0.175\n",
      "remark: distance between atom     682 and     668 (3.10) smaller than VdWsum (3.53) CH2P  NH1 0.074\n",
      "remark: distance between atom     685 and     679 (3.00) smaller than VdWsum (3.36)    C    O 0.004\n",
      "remark: distance between atom     699 and     695 (2.95) smaller than VdWsum (3.36)    C    O 0.053\n",
      "remark: distance between atom     707 and     700 (2.89) smaller than VdWsum (3.36)    C    O 0.111\n",
      "remark: distance between atom     720 and     713 (2.95) smaller than VdWsum (3.36)    C    O 0.057\n",
      "remark: distance between atom     722 and     700 (3.01) smaller than VdWsum (3.43) CH2E    O 0.071\n",
      "remark: distance between atom     729 and     721 (2.82) smaller than VdWsum (3.36)    C    O 0.182\n",
      "remark: distance between atom     736 and     720 (3.10) smaller than VdWsum (3.50)  NH1    C 0.050\n",
      "remark: distance between atom     821 and     106 (3.32) smaller than VdWsum (3.73) CR1E CH2E 0.061\n",
      "remark: distance between atom     837 and     825 (3.36) smaller than VdWsum (3.78) CH3E    C 0.062\n",
      "remark: distance between atom     847 and     841 (2.94) smaller than VdWsum (3.36)    C    O 0.069\n",
      "remark: distance between atom     872 and     865 (2.97) smaller than VdWsum (3.36)    C    O 0.032\n",
      "remark: distance between atom     880 and     873 (2.81) smaller than VdWsum (3.36)    C    O 0.195\n",
      "remark: distance between atom     895 and     888 (2.84) smaller than VdWsum (3.36)    C    O 0.162\n",
      "remark: distance between atom     906 and     896 (2.96) smaller than VdWsum (3.36)    C    O 0.046\n",
      "remark: distance between atom     934 and     928 (3.20) smaller than VdWsum (3.58)  NH1 CH2E 0.027\n",
      "remark: distance between atom     939 and     925 (3.44) smaller than VdWsum (3.83) CH2E    C 0.037\n",
      "remark: distance between atom     950 and     936 (3.33) smaller than VdWsum (3.78) CH3E    C 0.092\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:33:35,738 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_1a70d2a6-0f10-4266-bad8-f7d021f2fb5c', '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/978a6a32-0bc5-4100-a6fa-4610d563497e.stdin']\n",
      "2026-02-12 09:33:35,739 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_flexdyn.flexdyn.concoord_dist import concoord_dist\n",
    "\n",
    "concoord_dist_pdb = pdbCode + \"_dist.pdb\"\n",
    "concoord_dist_gro = pdbCode + \"_dist.gro\"\n",
    "concoord_dist_dat = pdbCode + \"_dist.dat\"\n",
    "\n",
    "# set prefix\n",
    "prefix = '/usr/local' if 'google.colab' in sys.modules else os.getenv('CONDA_PREFIX') \n",
    "concoord_lib = prefix +\"/share/concoord/lib\"\n",
    "\n",
    "# uncomment if executing in google colab\n",
    "# %env CONCOORD_OVERWRITE='1'\n",
    "# %env CONCOORDLIB=$concoord_lib\n",
    "\n",
    "prop = {\n",
    "    'retain_hydrogens' : False,\n",
    "    'cutoff' : 4.0,\n",
    "    'env_vars_dict' : {\n",
    "        'CONCOORD_OVERWRITE' : '1',\n",
    "        'CONCOORDLIB' : concoord_lib\n",
    "    }\n",
    "}\n",
    "\n",
    "concoord_dist(  input_structure_path=monomer,\n",
    "                output_pdb_path=concoord_dist_pdb,\n",
    "                output_gro_path=concoord_dist_gro,\n",
    "                output_dat_path=concoord_dist_dat,\n",
    "                properties=prop)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:33:44,289 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_64e1ff25-3315-496b-8128-635a80d018a9\n",
      "2026-02-12 09:33:44,292 [MainThread  ] [INFO ]  Copy to stage: 1ake_dist.pdb --> sandbox_64e1ff25-3315-496b-8128-635a80d018a9\n",
      "2026-02-12 09:33:44,294 [MainThread  ] [INFO ]  Copy to stage: 1ake_dist.dat --> sandbox_64e1ff25-3315-496b-8128-635a80d018a9\n",
      "2026-02-12 09:33:44,298 [MainThread  ] [INFO ]  Launching command (it may take a while): cd  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_64e1ff25-3315-496b-8128-635a80d018a9 ; disco -p 1ake_dist.pdb -d 1ake_dist.dat -or 1ake_disco_rmsd.dat -of 1ake_disco_bfactor.pdb -on 1ake_disco_traj.pdb -n 300\n",
      "2026-02-12 09:34:08,638 [MainThread  ] [INFO ]  Command 'cd  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_64e1ff25-3315-496b-8128-635a80d018a9 ; disco -p 1ake_dist.pdb -d 1ake_dist.dat -or 1ake_disco_rmsd.dat -of 1ake_disco_bfactor.pdb -on 1ake_disco_traj.pdb -n 300' finalized with exit code 0\n",
      "2026-02-12 09:34:08,639 [MainThread  ] [INFO ]  Determining system size...\n",
      "-------------------------------------------------\n",
      "disco - version 2.1.2 - build date: Thu Jan 24 10:29:30 CET 2013                                \n",
      "Process started with PID:    64960\n",
      "on host: OROZCO66.local\n",
      "number of threads =    8\n",
      "in directory: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_64e1ff25-3315-496b-8128-635a80d018a9\n",
      "on: 12/02/2026 at: 09:33:44\n",
      "-------------------------------------------------\n",
      "command line options:\n",
      "input coordinates...................: 1ake_dist.pdb                           \n",
      "distance constraints................: 1ake_dist.dat                           \n",
      "nr of structures to be generated....:     300\n",
      "maximum nr of iterations............:    2500\n",
      "maximum nr of trials................:      25\n",
      "maximal sum of violations...........:    1.00000\n",
      "output NMR-PDB file.................: 1ake_disco_traj.pdb                     \n",
      "output RMSD values..................: 1ake_disco_rmsd.dat                     \n",
      "output atomic fluctuations..........: 1ake_disco_bfactor.pdb                  \n",
      "chirality corrections will be performed on the fly\n",
      "Nr. of rounds of bound smoothing....:       0\n",
      "Damping factor (converted)..........:    1.00000\n",
      "processing input...\n",
      "number of atoms.....................:    1656\n",
      "number of residues..................:     214\n",
      "number of chiral centers to check...:     244\n",
      "number of planar groups to check....:     288\n",
      "sorting distances...\n",
      "number of read distances............:   57659\n",
      "number of angles....................:    2264\n",
      "total number of read distances......:   57659\n",
      "Nr of input violations:     103\n",
      "Sum of input violations: 0.23770E-03 nm\n",
      "Sum of input bumps: 0.00000E+00 nm\n",
      "perc. of total nr of distances......:    4.21\n",
      "Uncertainty sum/average:  0.21533E+05 0.37345E+00 nm\n",
      "Parallel operation, number of threads =    8\n",
      "Override by setting the environment variable OMP_NUM_THREADS\n",
      "generated structure.................:       1 with RMSD  1.26 A\n",
      "Nr. of violations...................:     445\n",
      "Sum of violations...................:    0.77914\n",
      "generated structure.................:       2 with RMSD  1.66 A\n",
      "Nr. of violations...................:     445\n",
      "Sum of violations...................:    0.83521\n",
      "generated structure.................:       3 with RMSD  2.35 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.81737\n",
      "generated structure.................:       4 with RMSD  2.58 A\n",
      "Nr. of violations...................:     472\n",
      "Sum of violations...................:    0.80287\n",
      "generated structure.................:       5 with RMSD  2.79 A\n",
      "Nr. of violations...................:     456\n",
      "Sum of violations...................:    0.81950\n",
      "generated structure.................:       6 with RMSD  1.74 A\n",
      "Nr. of violations...................:     426\n",
      "Sum of violations...................:    0.78850\n",
      "generated structure.................:       7 with RMSD  2.28 A\n",
      "Nr. of violations...................:     402\n",
      "Sum of violations...................:    0.80719\n",
      "generated structure.................:       8 with RMSD  1.16 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.73198\n",
      "generated structure.................:       9 with RMSD  1.20 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.84933\n",
      "generated structure.................:      10 with RMSD  1.48 A\n",
      "Nr. of violations...................:     452\n",
      "Sum of violations...................:    0.70433\n",
      "generated structure.................:      11 with RMSD  1.76 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.78108\n",
      "generated structure.................:      12 with RMSD  1.44 A\n",
      "Nr. of violations...................:     458\n",
      "Sum of violations...................:    0.79691\n",
      "generated structure.................:      13 with RMSD  2.03 A\n",
      "Nr. of violations...................:     426\n",
      "Sum of violations...................:    0.75311\n",
      "generated structure.................:      14 with RMSD  1.26 A\n",
      "Nr. of violations...................:     442\n",
      "Sum of violations...................:    0.77813\n",
      "generated structure.................:      15 with RMSD  1.31 A\n",
      "Nr. of violations...................:     460\n",
      "Sum of violations...................:    0.76014\n",
      "generated structure.................:      16 with RMSD  1.60 A\n",
      "Nr. of violations...................:     450\n",
      "Sum of violations...................:    0.76064\n",
      "generated structure.................:      17 with RMSD  1.31 A\n",
      "Nr. of violations...................:     477\n",
      "Sum of violations...................:    0.83720\n",
      "generated structure.................:      18 with RMSD  1.45 A\n",
      "Nr. of violations...................:     382\n",
      "Sum of violations...................:    0.75116\n",
      "generated structure.................:      19 with RMSD  1.78 A\n",
      "Nr. of violations...................:     430\n",
      "Sum of violations...................:    0.75219\n",
      "generated structure.................:      20 with RMSD  1.82 A\n",
      "Nr. of violations...................:     415\n",
      "Sum of violations...................:    0.73787\n",
      "generated structure.................:      21 with RMSD  1.32 A\n",
      "Nr. of violations...................:     418\n",
      "Sum of violations...................:    0.79625\n",
      "generated structure.................:      22 with RMSD  2.03 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.75057\n",
      "generated structure.................:      23 with RMSD  1.33 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.88986\n",
      "generated structure.................:      24 with RMSD  1.18 A\n",
      "Nr. of violations...................:     414\n",
      "Sum of violations...................:    0.81259\n",
      "generated structure.................:      25 with RMSD  1.91 A\n",
      "Nr. of violations...................:     473\n",
      "Sum of violations...................:    0.81918\n",
      "generated structure.................:      26 with RMSD  1.70 A\n",
      "Nr. of violations...................:     488\n",
      "Sum of violations...................:    0.87121\n",
      "generated structure.................:      27 with RMSD  1.27 A\n",
      "Nr. of violations...................:     457\n",
      "Sum of violations...................:    0.85637\n",
      "generated structure.................:      28 with RMSD  1.33 A\n",
      "Nr. of violations...................:     431\n",
      "Sum of violations...................:    0.75089\n",
      "generated structure.................:      29 with RMSD  2.12 A\n",
      "Nr. of violations...................:     428\n",
      "Sum of violations...................:    0.77227\n",
      "Giving up, best violation sum was...: 394.66013\n",
      "structure rejected (no convergence after    715 steps)\n",
      "generated structure.................:      30 with RMSD  2.10 A\n",
      "Nr. of violations...................:     424\n",
      "Sum of violations...................:    0.73377\n",
      "generated structure.................:      31 with RMSD  1.64 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.79151\n",
      "generated structure.................:      32 with RMSD  1.60 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.75840\n",
      "generated structure.................:      33 with RMSD  1.92 A\n",
      "Nr. of violations...................:     443\n",
      "Sum of violations...................:    0.84566\n",
      "generated structure.................:      34 with RMSD  2.03 A\n",
      "Nr. of violations...................:     457\n",
      "Sum of violations...................:    0.81023\n",
      "generated structure.................:      35 with RMSD  1.17 A\n",
      "Nr. of violations...................:     417\n",
      "Sum of violations...................:    0.71904\n",
      "generated structure.................:      36 with RMSD  2.35 A\n",
      "Nr. of violations...................:     480\n",
      "Sum of violations...................:    0.82693\n",
      "generated structure.................:      37 with RMSD  1.37 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.73601\n",
      "generated structure.................:      38 with RMSD  2.19 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.82114\n",
      "generated structure.................:      39 with RMSD  2.21 A\n",
      "Nr. of violations...................:     464\n",
      "Sum of violations...................:    0.86806\n",
      "generated structure.................:      40 with RMSD  1.23 A\n",
      "Nr. of violations...................:     424\n",
      "Sum of violations...................:    0.80524\n",
      "generated structure.................:      41 with RMSD  2.00 A\n",
      "Nr. of violations...................:     463\n",
      "Sum of violations...................:    0.78100\n",
      "generated structure.................:      42 with RMSD  1.55 A\n",
      "Nr. of violations...................:     442\n",
      "Sum of violations...................:    0.77941\n",
      "generated structure.................:      43 with RMSD  1.18 A\n",
      "Nr. of violations...................:     443\n",
      "Sum of violations...................:    0.80713\n",
      "generated structure.................:      44 with RMSD  2.48 A\n",
      "Nr. of violations...................:     468\n",
      "Sum of violations...................:    0.84219\n",
      "generated structure.................:      45 with RMSD  1.37 A\n",
      "Nr. of violations...................:     412\n",
      "Sum of violations...................:    0.78971\n",
      "generated structure.................:      46 with RMSD  1.84 A\n",
      "Nr. of violations...................:     445\n",
      "Sum of violations...................:    0.85174\n",
      "generated structure.................:      47 with RMSD  2.00 A\n",
      "Nr. of violations...................:     411\n",
      "Sum of violations...................:    0.77044\n",
      "generated structure.................:      48 with RMSD  1.69 A\n",
      "Nr. of violations...................:     413\n",
      "Sum of violations...................:    0.78282\n",
      "generated structure.................:      49 with RMSD  1.55 A\n",
      "Nr. of violations...................:     453\n",
      "Sum of violations...................:    0.70918\n",
      "generated structure.................:      50 with RMSD  1.73 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.78243\n",
      "generated structure.................:      51 with RMSD  2.11 A\n",
      "Nr. of violations...................:     407\n",
      "Sum of violations...................:    0.74696\n",
      "generated structure.................:      52 with RMSD  1.84 A\n",
      "Nr. of violations...................:     451\n",
      "Sum of violations...................:    0.80478\n",
      "generated structure.................:      53 with RMSD  1.85 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.81970\n",
      "generated structure.................:      54 with RMSD  1.49 A\n",
      "Nr. of violations...................:     426\n",
      "Sum of violations...................:    0.84884\n",
      "generated structure.................:      55 with RMSD  2.32 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.80005\n",
      "generated structure.................:      56 with RMSD  2.25 A\n",
      "Nr. of violations...................:     405\n",
      "Sum of violations...................:    0.81745\n",
      "Giving up, best violation sum was...: 405.46317\n",
      "structure rejected (no convergence after    974 steps)\n",
      "generated structure.................:      57 with RMSD  2.00 A\n",
      "Nr. of violations...................:     420\n",
      "Sum of violations...................:    0.74049\n",
      "generated structure.................:      58 with RMSD  2.26 A\n",
      "Nr. of violations...................:     438\n",
      "Sum of violations...................:    0.76536\n",
      "generated structure.................:      59 with RMSD  1.41 A\n",
      "Nr. of violations...................:     443\n",
      "Sum of violations...................:    0.78243\n",
      "generated structure.................:      60 with RMSD  1.15 A\n",
      "Nr. of violations...................:     427\n",
      "Sum of violations...................:    0.71395\n",
      "generated structure.................:      61 with RMSD  2.18 A\n",
      "Nr. of violations...................:     440\n",
      "Sum of violations...................:    0.80047\n",
      "generated structure.................:      62 with RMSD  1.52 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.79915\n",
      "generated structure.................:      63 with RMSD  2.02 A\n",
      "Nr. of violations...................:     451\n",
      "Sum of violations...................:    0.83417\n",
      "generated structure.................:      64 with RMSD  1.77 A\n",
      "Nr. of violations...................:     440\n",
      "Sum of violations...................:    0.80364\n",
      "generated structure.................:      65 with RMSD  0.94 A\n",
      "Nr. of violations...................:     401\n",
      "Sum of violations...................:    0.74946\n",
      "generated structure.................:      66 with RMSD  1.77 A\n",
      "Nr. of violations...................:     401\n",
      "Sum of violations...................:    0.76736\n",
      "Giving up, best violation sum was...: 379.22742\n",
      "structure rejected (no convergence after   1636 steps)\n",
      "generated structure.................:      67 with RMSD  1.75 A\n",
      "Nr. of violations...................:     467\n",
      "Sum of violations...................:    0.86649\n",
      "generated structure.................:      68 with RMSD  1.80 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.81444\n",
      "generated structure.................:      69 with RMSD  1.93 A\n",
      "Nr. of violations...................:     418\n",
      "Sum of violations...................:    0.70731\n",
      "generated structure.................:      70 with RMSD  1.16 A\n",
      "Nr. of violations...................:     468\n",
      "Sum of violations...................:    0.80347\n",
      "generated structure.................:      71 with RMSD  1.75 A\n",
      "Nr. of violations...................:     430\n",
      "Sum of violations...................:    0.75053\n",
      "generated structure.................:      72 with RMSD  1.37 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.77665\n",
      "generated structure.................:      73 with RMSD  1.36 A\n",
      "Nr. of violations...................:     441\n",
      "Sum of violations...................:    0.79506\n",
      "generated structure.................:      74 with RMSD  1.05 A\n",
      "Nr. of violations...................:     434\n",
      "Sum of violations...................:    0.84277\n",
      "generated structure.................:      75 with RMSD  2.25 A\n",
      "Nr. of violations...................:     395\n",
      "Sum of violations...................:    0.73667\n",
      "generated structure.................:      76 with RMSD  1.83 A\n",
      "Nr. of violations...................:     388\n",
      "Sum of violations...................:    0.73053\n",
      "generated structure.................:      77 with RMSD  2.16 A\n",
      "Nr. of violations...................:     434\n",
      "Sum of violations...................:    0.76272\n",
      "generated structure.................:      78 with RMSD  1.95 A\n",
      "Nr. of violations...................:     455\n",
      "Sum of violations...................:    0.75213\n",
      "generated structure.................:      79 with RMSD  1.77 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.77273\n",
      "generated structure.................:      80 with RMSD  1.98 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.81415\n",
      "generated structure.................:      81 with RMSD  1.91 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.82862\n",
      "generated structure.................:      82 with RMSD  1.90 A\n",
      "Nr. of violations...................:     400\n",
      "Sum of violations...................:    0.77742\n",
      "generated structure.................:      83 with RMSD  1.34 A\n",
      "Nr. of violations...................:     452\n",
      "Sum of violations...................:    0.86090\n",
      "generated structure.................:      84 with RMSD  1.67 A\n",
      "Nr. of violations...................:     427\n",
      "Sum of violations...................:    0.62225\n",
      "generated structure.................:      85 with RMSD  1.45 A\n",
      "Nr. of violations...................:     422\n",
      "Sum of violations...................:    0.81951\n",
      "generated structure.................:      86 with RMSD  1.30 A\n",
      "Nr. of violations...................:     431\n",
      "Sum of violations...................:    0.77878\n",
      "generated structure.................:      87 with RMSD  1.15 A\n",
      "Nr. of violations...................:     401\n",
      "Sum of violations...................:    0.63825\n",
      "generated structure.................:      88 with RMSD  1.73 A\n",
      "Nr. of violations...................:     425\n",
      "Sum of violations...................:    0.76669\n",
      "generated structure.................:      89 with RMSD  1.85 A\n",
      "Nr. of violations...................:     450\n",
      "Sum of violations...................:    0.81414\n",
      "generated structure.................:      90 with RMSD  1.56 A\n",
      "Nr. of violations...................:     425\n",
      "Sum of violations...................:    0.75452\n",
      "generated structure.................:      91 with RMSD  1.58 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.80829\n",
      "generated structure.................:      92 with RMSD  1.79 A\n",
      "Nr. of violations...................:     454\n",
      "Sum of violations...................:    0.82367\n",
      "generated structure.................:      93 with RMSD  2.24 A\n",
      "Nr. of violations...................:     424\n",
      "Sum of violations...................:    0.82394\n",
      "generated structure.................:      94 with RMSD  2.26 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.69473\n",
      "generated structure.................:      95 with RMSD  1.62 A\n",
      "Nr. of violations...................:     457\n",
      "Sum of violations...................:    0.84095\n",
      "generated structure.................:      96 with RMSD  1.70 A\n",
      "Nr. of violations...................:     468\n",
      "Sum of violations...................:    0.86515\n",
      "generated structure.................:      97 with RMSD  1.64 A\n",
      "Nr. of violations...................:     392\n",
      "Sum of violations...................:    0.66514\n",
      "generated structure.................:      98 with RMSD  1.08 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.84951\n",
      "generated structure.................:      99 with RMSD  1.07 A\n",
      "Nr. of violations...................:     461\n",
      "Sum of violations...................:    0.88843\n",
      "generated structure.................:     100 with RMSD  1.53 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.80280\n",
      "generated structure.................:     101 with RMSD  1.77 A\n",
      "Nr. of violations...................:     413\n",
      "Sum of violations...................:    0.79804\n",
      "generated structure.................:     102 with RMSD  1.60 A\n",
      "Nr. of violations...................:     460\n",
      "Sum of violations...................:    0.86239\n",
      "generated structure.................:     103 with RMSD  1.72 A\n",
      "Nr. of violations...................:     417\n",
      "Sum of violations...................:    0.81042\n",
      "generated structure.................:     104 with RMSD  1.99 A\n",
      "Nr. of violations...................:     390\n",
      "Sum of violations...................:    0.69173\n",
      "generated structure.................:     105 with RMSD  1.95 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.77477\n",
      "generated structure.................:     106 with RMSD  1.28 A\n",
      "Nr. of violations...................:     388\n",
      "Sum of violations...................:    0.68779\n",
      "generated structure.................:     107 with RMSD  2.09 A\n",
      "Nr. of violations...................:     432\n",
      "Sum of violations...................:    0.83301\n",
      "generated structure.................:     108 with RMSD  1.35 A\n",
      "Nr. of violations...................:     441\n",
      "Sum of violations...................:    0.87827\n",
      "generated structure.................:     109 with RMSD  1.90 A\n",
      "Nr. of violations...................:     418\n",
      "Sum of violations...................:    0.73530\n",
      "generated structure.................:     110 with RMSD  1.41 A\n",
      "Nr. of violations...................:     401\n",
      "Sum of violations...................:    0.82109\n",
      "generated structure.................:     111 with RMSD  1.24 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.76595\n",
      "generated structure.................:     112 with RMSD  1.49 A\n",
      "Nr. of violations...................:     410\n",
      "Sum of violations...................:    0.85274\n",
      "generated structure.................:     113 with RMSD  1.26 A\n",
      "Nr. of violations...................:     388\n",
      "Sum of violations...................:    0.79097\n",
      "generated structure.................:     114 with RMSD  1.50 A\n",
      "Nr. of violations...................:     461\n",
      "Sum of violations...................:    0.84563\n",
      "generated structure.................:     115 with RMSD  1.74 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.81274\n",
      "generated structure.................:     116 with RMSD  2.32 A\n",
      "Nr. of violations...................:     493\n",
      "Sum of violations...................:    0.85627\n",
      "generated structure.................:     117 with RMSD  1.28 A\n",
      "Nr. of violations...................:     435\n",
      "Sum of violations...................:    0.80293\n",
      "generated structure.................:     118 with RMSD  1.89 A\n",
      "Nr. of violations...................:     428\n",
      "Sum of violations...................:    0.85597\n",
      "generated structure.................:     119 with RMSD  1.58 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.81332\n",
      "generated structure.................:     120 with RMSD  1.51 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.81704\n",
      "generated structure.................:     121 with RMSD  1.60 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.86431\n",
      "generated structure.................:     122 with RMSD  1.93 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.82159\n",
      "generated structure.................:     123 with RMSD  1.56 A\n",
      "Nr. of violations...................:     437\n",
      "Sum of violations...................:    0.80259\n",
      "generated structure.................:     124 with RMSD  1.64 A\n",
      "Nr. of violations...................:     402\n",
      "Sum of violations...................:    0.73752\n",
      "generated structure.................:     125 with RMSD  1.53 A\n",
      "Nr. of violations...................:     427\n",
      "Sum of violations...................:    0.76509\n",
      "generated structure.................:     126 with RMSD  1.93 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.83611\n",
      "generated structure.................:     127 with RMSD  2.15 A\n",
      "Nr. of violations...................:     455\n",
      "Sum of violations...................:    0.85932\n",
      "generated structure.................:     128 with RMSD  1.76 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.76132\n",
      "generated structure.................:     129 with RMSD  2.19 A\n",
      "Nr. of violations...................:     467\n",
      "Sum of violations...................:    0.82781\n",
      "generated structure.................:     130 with RMSD  1.38 A\n",
      "Nr. of violations...................:     461\n",
      "Sum of violations...................:    0.79990\n",
      "generated structure.................:     131 with RMSD  1.72 A\n",
      "Nr. of violations...................:     430\n",
      "Sum of violations...................:    0.77136\n",
      "generated structure.................:     132 with RMSD  1.11 A\n",
      "Nr. of violations...................:     419\n",
      "Sum of violations...................:    0.79730\n",
      "generated structure.................:     133 with RMSD  1.28 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.84604\n",
      "generated structure.................:     134 with RMSD  1.45 A\n",
      "Nr. of violations...................:     431\n",
      "Sum of violations...................:    0.65415\n",
      "generated structure.................:     135 with RMSD  2.04 A\n",
      "Nr. of violations...................:     455\n",
      "Sum of violations...................:    0.82268\n",
      "generated structure.................:     136 with RMSD  2.02 A\n",
      "Nr. of violations...................:     421\n",
      "Sum of violations...................:    0.79853\n",
      "generated structure.................:     137 with RMSD  2.43 A\n",
      "Nr. of violations...................:     451\n",
      "Sum of violations...................:    0.82648\n",
      "generated structure.................:     138 with RMSD  1.72 A\n",
      "Nr. of violations...................:     406\n",
      "Sum of violations...................:    0.76392\n",
      "generated structure.................:     139 with RMSD  1.56 A\n",
      "Nr. of violations...................:     446\n",
      "Sum of violations...................:    0.82260\n",
      "generated structure.................:     140 with RMSD  1.62 A\n",
      "Nr. of violations...................:     430\n",
      "Sum of violations...................:    0.80992\n",
      "generated structure.................:     141 with RMSD  1.26 A\n",
      "Nr. of violations...................:     438\n",
      "Sum of violations...................:    0.78323\n",
      "generated structure.................:     142 with RMSD  2.16 A\n",
      "Nr. of violations...................:     485\n",
      "Sum of violations...................:    0.85733\n",
      "generated structure.................:     143 with RMSD  1.04 A\n",
      "Nr. of violations...................:     428\n",
      "Sum of violations...................:    0.76001\n",
      "generated structure.................:     144 with RMSD  1.19 A\n",
      "Nr. of violations...................:     426\n",
      "Sum of violations...................:    0.71557\n",
      "generated structure.................:     145 with RMSD  1.91 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.84036\n",
      "generated structure.................:     146 with RMSD  1.75 A\n",
      "Nr. of violations...................:     432\n",
      "Sum of violations...................:    0.82122\n",
      "generated structure.................:     147 with RMSD  1.80 A\n",
      "Nr. of violations...................:     455\n",
      "Sum of violations...................:    0.86479\n",
      "generated structure.................:     148 with RMSD  1.06 A\n",
      "Nr. of violations...................:     443\n",
      "Sum of violations...................:    0.74946\n",
      "generated structure.................:     149 with RMSD  1.86 A\n",
      "Nr. of violations...................:     418\n",
      "Sum of violations...................:    0.74781\n",
      "generated structure.................:     150 with RMSD  1.57 A\n",
      "Nr. of violations...................:     421\n",
      "Sum of violations...................:    0.80316\n",
      "generated structure.................:     151 with RMSD  2.16 A\n",
      "Nr. of violations...................:     484\n",
      "Sum of violations...................:    0.84464\n",
      "generated structure.................:     152 with RMSD  1.23 A\n",
      "Nr. of violations...................:     405\n",
      "Sum of violations...................:    0.79281\n",
      "generated structure.................:     153 with RMSD  1.44 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.86843\n",
      "generated structure.................:     154 with RMSD  1.40 A\n",
      "Nr. of violations...................:     475\n",
      "Sum of violations...................:    0.92210\n",
      "generated structure.................:     155 with RMSD  2.52 A\n",
      "Nr. of violations...................:     410\n",
      "Sum of violations...................:    0.66370\n",
      "generated structure.................:     156 with RMSD  1.66 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.86779\n",
      "generated structure.................:     157 with RMSD  1.69 A\n",
      "Nr. of violations...................:     464\n",
      "Sum of violations...................:    0.85108\n",
      "generated structure.................:     158 with RMSD  1.78 A\n",
      "Nr. of violations...................:     456\n",
      "Sum of violations...................:    0.82770\n",
      "Giving up, best violation sum was...: 378.40967\n",
      "structure rejected (no convergence after   1766 steps)\n",
      "generated structure.................:     159 with RMSD  1.92 A\n",
      "Nr. of violations...................:     435\n",
      "Sum of violations...................:    0.83764\n",
      "generated structure.................:     160 with RMSD  2.08 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.81105\n",
      "generated structure.................:     161 with RMSD  2.23 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.74361\n",
      "generated structure.................:     162 with RMSD  1.91 A\n",
      "Nr. of violations...................:     454\n",
      "Sum of violations...................:    0.78904\n",
      "Giving up, best violation sum was...: 388.46039\n",
      "structure rejected (no convergence after   2295 steps)\n",
      "generated structure.................:     163 with RMSD  2.20 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.79704\n",
      "generated structure.................:     164 with RMSD  1.42 A\n",
      "Nr. of violations...................:     416\n",
      "Sum of violations...................:    0.79973\n",
      "generated structure.................:     165 with RMSD  1.50 A\n",
      "Nr. of violations...................:     450\n",
      "Sum of violations...................:    0.86583\n",
      "generated structure.................:     166 with RMSD  1.89 A\n",
      "Nr. of violations...................:     474\n",
      "Sum of violations...................:    0.87985\n",
      "generated structure.................:     167 with RMSD  1.64 A\n",
      "Nr. of violations...................:     441\n",
      "Sum of violations...................:    0.83988\n",
      "generated structure.................:     168 with RMSD  1.57 A\n",
      "Nr. of violations...................:     440\n",
      "Sum of violations...................:    0.75085\n",
      "generated structure.................:     169 with RMSD  1.69 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.78987\n",
      "generated structure.................:     170 with RMSD  2.30 A\n",
      "Nr. of violations...................:     430\n",
      "Sum of violations...................:    0.81805\n",
      "generated structure.................:     171 with RMSD  2.29 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.82056\n",
      "generated structure.................:     172 with RMSD  1.87 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.79761\n",
      "generated structure.................:     173 with RMSD  2.14 A\n",
      "Nr. of violations...................:     425\n",
      "Sum of violations...................:    0.80026\n",
      "generated structure.................:     174 with RMSD  2.35 A\n",
      "Nr. of violations...................:     432\n",
      "Sum of violations...................:    0.81016\n",
      "generated structure.................:     175 with RMSD  1.62 A\n",
      "Nr. of violations...................:     410\n",
      "Sum of violations...................:    0.71789\n",
      "generated structure.................:     176 with RMSD  1.82 A\n",
      "Nr. of violations...................:     446\n",
      "Sum of violations...................:    0.84161\n",
      "generated structure.................:     177 with RMSD  1.92 A\n",
      "Nr. of violations...................:     397\n",
      "Sum of violations...................:    0.66128\n",
      "generated structure.................:     178 with RMSD  1.57 A\n",
      "Nr. of violations...................:     458\n",
      "Sum of violations...................:    0.79058\n",
      "generated structure.................:     179 with RMSD  1.17 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.84305\n",
      "generated structure.................:     180 with RMSD  1.70 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.78540\n",
      "generated structure.................:     181 with RMSD  1.19 A\n",
      "Nr. of violations...................:     396\n",
      "Sum of violations...................:    0.77060\n",
      "generated structure.................:     182 with RMSD  1.06 A\n",
      "Nr. of violations...................:     404\n",
      "Sum of violations...................:    0.77424\n",
      "generated structure.................:     183 with RMSD  1.68 A\n",
      "Nr. of violations...................:     440\n",
      "Sum of violations...................:    0.78021\n",
      "generated structure.................:     184 with RMSD  1.40 A\n",
      "Nr. of violations...................:     415\n",
      "Sum of violations...................:    0.72233\n",
      "generated structure.................:     185 with RMSD  1.44 A\n",
      "Nr. of violations...................:     388\n",
      "Sum of violations...................:    0.69168\n",
      "generated structure.................:     186 with RMSD  1.91 A\n",
      "Nr. of violations...................:     456\n",
      "Sum of violations...................:    0.77674\n",
      "generated structure.................:     187 with RMSD  1.64 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.79761\n",
      "generated structure.................:     188 with RMSD  1.64 A\n",
      "Nr. of violations...................:     415\n",
      "Sum of violations...................:    0.73851\n",
      "generated structure.................:     189 with RMSD  1.41 A\n",
      "Nr. of violations...................:     435\n",
      "Sum of violations...................:    0.73183\n",
      "generated structure.................:     190 with RMSD  1.81 A\n",
      "Nr. of violations...................:     401\n",
      "Sum of violations...................:    0.81859\n",
      "generated structure.................:     191 with RMSD  2.11 A\n",
      "Nr. of violations...................:     467\n",
      "Sum of violations...................:    0.81975\n",
      "generated structure.................:     192 with RMSD  1.75 A\n",
      "Nr. of violations...................:     431\n",
      "Sum of violations...................:    0.78769\n",
      "generated structure.................:     193 with RMSD  1.25 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.83550\n",
      "generated structure.................:     194 with RMSD  1.41 A\n",
      "Nr. of violations...................:     466\n",
      "Sum of violations...................:    0.73543\n",
      "generated structure.................:     195 with RMSD  2.34 A\n",
      "Nr. of violations...................:     468\n",
      "Sum of violations...................:    0.84648\n",
      "generated structure.................:     196 with RMSD  2.36 A\n",
      "Nr. of violations...................:     442\n",
      "Sum of violations...................:    0.70654\n",
      "generated structure.................:     197 with RMSD  2.06 A\n",
      "Nr. of violations...................:     427\n",
      "Sum of violations...................:    0.77269\n",
      "generated structure.................:     198 with RMSD  1.57 A\n",
      "Nr. of violations...................:     464\n",
      "Sum of violations...................:    0.84103\n",
      "generated structure.................:     199 with RMSD  1.80 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.75065\n",
      "generated structure.................:     200 with RMSD  2.22 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.82891\n",
      "generated structure.................:     201 with RMSD  2.16 A\n",
      "Nr. of violations...................:     446\n",
      "Sum of violations...................:    0.81087\n",
      "generated structure.................:     202 with RMSD  1.97 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.84389\n",
      "generated structure.................:     203 with RMSD  1.27 A\n",
      "Nr. of violations...................:     458\n",
      "Sum of violations...................:    0.75954\n",
      "generated structure.................:     204 with RMSD  1.29 A\n",
      "Nr. of violations...................:     405\n",
      "Sum of violations...................:    0.72249\n",
      "generated structure.................:     205 with RMSD  1.88 A\n",
      "Nr. of violations...................:     425\n",
      "Sum of violations...................:    0.85520\n",
      "generated structure.................:     206 with RMSD  1.34 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.75661\n",
      "generated structure.................:     207 with RMSD  1.86 A\n",
      "Nr. of violations...................:     432\n",
      "Sum of violations...................:    0.69821\n",
      "generated structure.................:     208 with RMSD  1.92 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.70572\n",
      "generated structure.................:     209 with RMSD  1.82 A\n",
      "Nr. of violations...................:     452\n",
      "Sum of violations...................:    0.85225\n",
      "generated structure.................:     210 with RMSD  1.87 A\n",
      "Nr. of violations...................:     442\n",
      "Sum of violations...................:    0.77785\n",
      "generated structure.................:     211 with RMSD  1.72 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.82849\n",
      "generated structure.................:     212 with RMSD  1.52 A\n",
      "Nr. of violations...................:     420\n",
      "Sum of violations...................:    0.78782\n",
      "generated structure.................:     213 with RMSD  1.25 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.78382\n",
      "generated structure.................:     214 with RMSD  1.95 A\n",
      "Nr. of violations...................:     448\n",
      "Sum of violations...................:    0.85475\n",
      "generated structure.................:     215 with RMSD  1.52 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.78349\n",
      "generated structure.................:     216 with RMSD  1.24 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.67162\n",
      "generated structure.................:     217 with RMSD  2.44 A\n",
      "Nr. of violations...................:     427\n",
      "Sum of violations...................:    0.78111\n",
      "generated structure.................:     218 with RMSD  1.22 A\n",
      "Nr. of violations...................:     418\n",
      "Sum of violations...................:    0.79938\n",
      "generated structure.................:     219 with RMSD  1.42 A\n",
      "Nr. of violations...................:     456\n",
      "Sum of violations...................:    0.79912\n",
      "generated structure.................:     220 with RMSD  2.13 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.82946\n",
      "generated structure.................:     221 with RMSD  1.22 A\n",
      "Nr. of violations...................:     437\n",
      "Sum of violations...................:    0.74726\n",
      "generated structure.................:     222 with RMSD  1.88 A\n",
      "Nr. of violations...................:     454\n",
      "Sum of violations...................:    0.88064\n",
      "generated structure.................:     223 with RMSD  2.47 A\n",
      "Nr. of violations...................:     440\n",
      "Sum of violations...................:    0.81076\n",
      "generated structure.................:     224 with RMSD  1.95 A\n",
      "Nr. of violations...................:     456\n",
      "Sum of violations...................:    0.71212\n",
      "generated structure.................:     225 with RMSD  1.56 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.80217\n",
      "generated structure.................:     226 with RMSD  2.68 A\n",
      "Nr. of violations...................:     441\n",
      "Sum of violations...................:    0.78654\n",
      "generated structure.................:     227 with RMSD  2.36 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.85631\n",
      "generated structure.................:     228 with RMSD  1.89 A\n",
      "Nr. of violations...................:     472\n",
      "Sum of violations...................:    0.82845\n",
      "generated structure.................:     229 with RMSD  1.71 A\n",
      "Nr. of violations...................:     417\n",
      "Sum of violations...................:    0.77636\n",
      "generated structure.................:     230 with RMSD  2.10 A\n",
      "Nr. of violations...................:     463\n",
      "Sum of violations...................:    0.81072\n",
      "generated structure.................:     231 with RMSD  1.75 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.80301\n",
      "generated structure.................:     232 with RMSD  1.80 A\n",
      "Nr. of violations...................:     461\n",
      "Sum of violations...................:    0.83796\n",
      "generated structure.................:     233 with RMSD  1.04 A\n",
      "Nr. of violations...................:     428\n",
      "Sum of violations...................:    0.74988\n",
      "generated structure.................:     234 with RMSD  1.18 A\n",
      "Nr. of violations...................:     457\n",
      "Sum of violations...................:    0.79409\n",
      "generated structure.................:     235 with RMSD  1.80 A\n",
      "Nr. of violations...................:     459\n",
      "Sum of violations...................:    0.72997\n",
      "generated structure.................:     236 with RMSD  1.96 A\n",
      "Nr. of violations...................:     446\n",
      "Sum of violations...................:    0.83733\n",
      "generated structure.................:     237 with RMSD  1.54 A\n",
      "Nr. of violations...................:     452\n",
      "Sum of violations...................:    0.86958\n",
      "generated structure.................:     238 with RMSD  1.57 A\n",
      "Nr. of violations...................:     435\n",
      "Sum of violations...................:    0.87363\n",
      "generated structure.................:     239 with RMSD  1.06 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.73844\n",
      "generated structure.................:     240 with RMSD  1.62 A\n",
      "Nr. of violations...................:     424\n",
      "Sum of violations...................:    0.74941\n",
      "generated structure.................:     241 with RMSD  1.54 A\n",
      "Nr. of violations...................:     405\n",
      "Sum of violations...................:    0.84070\n",
      "generated structure.................:     242 with RMSD  1.18 A\n",
      "Nr. of violations...................:     469\n",
      "Sum of violations...................:    0.85301\n",
      "generated structure.................:     243 with RMSD  2.20 A\n",
      "Nr. of violations...................:     446\n",
      "Sum of violations...................:    0.83043\n",
      "generated structure.................:     244 with RMSD  2.26 A\n",
      "Nr. of violations...................:     420\n",
      "Sum of violations...................:    0.68792\n",
      "generated structure.................:     245 with RMSD  1.46 A\n",
      "Nr. of violations...................:     419\n",
      "Sum of violations...................:    0.69562\n",
      "generated structure.................:     246 with RMSD  1.72 A\n",
      "Nr. of violations...................:     405\n",
      "Sum of violations...................:    0.75248\n",
      "generated structure.................:     247 with RMSD  1.29 A\n",
      "Nr. of violations...................:     442\n",
      "Sum of violations...................:    0.83564\n",
      "generated structure.................:     248 with RMSD  1.11 A\n",
      "Nr. of violations...................:     465\n",
      "Sum of violations...................:    0.82468\n",
      "generated structure.................:     249 with RMSD  1.10 A\n",
      "Nr. of violations...................:     431\n",
      "Sum of violations...................:    0.70294\n",
      "generated structure.................:     250 with RMSD  1.27 A\n",
      "Nr. of violations...................:     469\n",
      "Sum of violations...................:    0.84855\n",
      "generated structure.................:     251 with RMSD  1.73 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.87796\n",
      "generated structure.................:     252 with RMSD  2.31 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.76117\n",
      "generated structure.................:     253 with RMSD  1.14 A\n",
      "Nr. of violations...................:     391\n",
      "Sum of violations...................:    0.75216\n",
      "generated structure.................:     254 with RMSD  1.70 A\n",
      "Nr. of violations...................:     437\n",
      "Sum of violations...................:    0.72178\n",
      "generated structure.................:     255 with RMSD  2.00 A\n",
      "Nr. of violations...................:     391\n",
      "Sum of violations...................:    0.73860\n",
      "generated structure.................:     256 with RMSD  1.24 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.82788\n",
      "generated structure.................:     257 with RMSD  1.17 A\n",
      "Nr. of violations...................:     399\n",
      "Sum of violations...................:    0.79120\n",
      "generated structure.................:     258 with RMSD  1.57 A\n",
      "Nr. of violations...................:     449\n",
      "Sum of violations...................:    0.79780\n",
      "generated structure.................:     259 with RMSD  2.06 A\n",
      "Nr. of violations...................:     422\n",
      "Sum of violations...................:    0.76580\n",
      "generated structure.................:     260 with RMSD  1.95 A\n",
      "Nr. of violations...................:     407\n",
      "Sum of violations...................:    0.80857\n",
      "generated structure.................:     261 with RMSD  1.20 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.79705\n",
      "generated structure.................:     262 with RMSD  1.54 A\n",
      "Nr. of violations...................:     428\n",
      "Sum of violations...................:    0.79842\n",
      "generated structure.................:     263 with RMSD  2.13 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.84571\n",
      "generated structure.................:     264 with RMSD  1.91 A\n",
      "Nr. of violations...................:     470\n",
      "Sum of violations...................:    0.82364\n",
      "generated structure.................:     265 with RMSD  2.08 A\n",
      "Nr. of violations...................:     423\n",
      "Sum of violations...................:    0.74504\n",
      "generated structure.................:     266 with RMSD  2.18 A\n",
      "Nr. of violations...................:     478\n",
      "Sum of violations...................:    0.88243\n",
      "generated structure.................:     267 with RMSD  2.12 A\n",
      "Nr. of violations...................:     421\n",
      "Sum of violations...................:    0.70797\n",
      "generated structure.................:     268 with RMSD  1.44 A\n",
      "Nr. of violations...................:     473\n",
      "Sum of violations...................:    0.89621\n",
      "generated structure.................:     269 with RMSD  1.53 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.78594\n",
      "generated structure.................:     270 with RMSD  2.26 A\n",
      "Nr. of violations...................:     460\n",
      "Sum of violations...................:    0.80791\n",
      "generated structure.................:     271 with RMSD  1.25 A\n",
      "Nr. of violations...................:     420\n",
      "Sum of violations...................:    0.79014\n",
      "generated structure.................:     272 with RMSD  1.35 A\n",
      "Nr. of violations...................:     406\n",
      "Sum of violations...................:    0.73133\n",
      "generated structure.................:     273 with RMSD  2.15 A\n",
      "Nr. of violations...................:     421\n",
      "Sum of violations...................:    0.72698\n",
      "generated structure.................:     274 with RMSD  1.40 A\n",
      "Nr. of violations...................:     429\n",
      "Sum of violations...................:    0.72157\n",
      "generated structure.................:     275 with RMSD  1.68 A\n",
      "Nr. of violations...................:     445\n",
      "Sum of violations...................:    0.85546\n",
      "generated structure.................:     276 with RMSD  1.94 A\n",
      "Nr. of violations...................:     463\n",
      "Sum of violations...................:    0.82971\n",
      "generated structure.................:     277 with RMSD  1.42 A\n",
      "Nr. of violations...................:     445\n",
      "Sum of violations...................:    0.82529\n",
      "generated structure.................:     278 with RMSD  1.58 A\n",
      "Nr. of violations...................:     447\n",
      "Sum of violations...................:    0.83412\n",
      "generated structure.................:     279 with RMSD  2.04 A\n",
      "Nr. of violations...................:     458\n",
      "Sum of violations...................:    0.76733\n",
      "generated structure.................:     280 with RMSD  1.39 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.81198\n",
      "generated structure.................:     281 with RMSD  1.37 A\n",
      "Nr. of violations...................:     442\n",
      "Sum of violations...................:    0.77385\n",
      "generated structure.................:     282 with RMSD  2.28 A\n",
      "Nr. of violations...................:     436\n",
      "Sum of violations...................:    0.83739\n",
      "generated structure.................:     283 with RMSD  1.47 A\n",
      "Nr. of violations...................:     454\n",
      "Sum of violations...................:    0.84751\n",
      "generated structure.................:     284 with RMSD  1.74 A\n",
      "Nr. of violations...................:     439\n",
      "Sum of violations...................:    0.79327\n",
      "generated structure.................:     285 with RMSD  1.81 A\n",
      "Nr. of violations...................:     420\n",
      "Sum of violations...................:    0.73633\n",
      "generated structure.................:     286 with RMSD  1.88 A\n",
      "Nr. of violations...................:     413\n",
      "Sum of violations...................:    0.72573\n",
      "generated structure.................:     287 with RMSD  1.88 A\n",
      "Nr. of violations...................:     433\n",
      "Sum of violations...................:    0.80026\n",
      "Giving up, best violation sum was...: 384.77966\n",
      "structure rejected (no convergence after   2500 steps)\n",
      "generated structure.................:     288 with RMSD  1.53 A\n",
      "Nr. of violations...................:     455\n",
      "Sum of violations...................:    0.85647\n",
      "generated structure.................:     289 with RMSD  1.94 A\n",
      "Nr. of violations...................:     418\n",
      "Sum of violations...................:    0.75314\n",
      "generated structure.................:     290 with RMSD  1.85 A\n",
      "Nr. of violations...................:     457\n",
      "Sum of violations...................:    0.86458\n",
      "generated structure.................:     291 with RMSD  1.16 A\n",
      "Nr. of violations...................:     402\n",
      "Sum of violations...................:    0.74855\n",
      "generated structure.................:     292 with RMSD  2.25 A\n",
      "Nr. of violations...................:     459\n",
      "Sum of violations...................:    0.85515\n",
      "generated structure.................:     293 with RMSD  1.90 A\n",
      "Nr. of violations...................:     425\n",
      "Sum of violations...................:    0.77876\n",
      "Giving up, best violation sum was...: 391.93079\n",
      "structure rejected (no convergence after   1097 steps)\n",
      "generated structure.................:     294 with RMSD  1.41 A\n",
      "Nr. of violations...................:     421\n",
      "Sum of violations...................:    0.76174\n",
      "generated structure.................:     295 with RMSD  1.89 A\n",
      "Nr. of violations...................:     408\n",
      "Sum of violations...................:    0.71793\n",
      "generated structure.................:     296 with RMSD  1.60 A\n",
      "Nr. of violations...................:     409\n",
      "Sum of violations...................:    0.70929\n",
      "generated structure.................:     297 with RMSD  1.81 A\n",
      "Nr. of violations...................:     435\n",
      "Sum of violations...................:    0.89404\n",
      "generated structure.................:     298 with RMSD  1.71 A\n",
      "Nr. of violations...................:     444\n",
      "Sum of violations...................:    0.81328\n",
      "generated structure.................:     299 with RMSD  0.95 A\n",
      "Nr. of violations...................:     397\n",
      "Sum of violations...................:    0.71966\n",
      "Giving up, best violation sum was...: 392.52573\n",
      "structure rejected (no convergence after   1558 steps)\n",
      "generated structure.................:     300 with RMSD  1.29 A\n",
      "Nr. of violations...................:     438\n",
      "Sum of violations...................:    0.78583\n",
      "\n",
      "2026-02-12 09:34:08,652 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_64e1ff25-3315-496b-8128-635a80d018a9']\n",
      "2026-02-12 09:34:08,653 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_flexdyn.flexdyn.concoord_disco import concoord_disco\n",
    "\n",
    "concoord_disco_pdb = pdbCode + \"_disco_traj.pdb\"\n",
    "concoord_disco_rmsd = pdbCode + \"_disco_rmsd.dat\"\n",
    "concoord_disco_bfactor = pdbCode + \"_disco_bfactor.pdb\"\n",
    "\n",
    "# set prefix\n",
    "prefix = '/usr/local' if 'google.colab' in sys.modules else os.getenv('CONDA_PREFIX') \n",
    "concoord_lib = prefix +\"/share/concoord/lib\"\n",
    "\n",
    "# uncomment if executing in google colab\n",
    "# %env CONCOORD_OVERWRITE='1'\n",
    "# %env CONCOORDLIB=$concoord_lib\n",
    "\n",
    "prop = {\n",
    "    'vdw' : 4,\n",
    "    'num_structs' : num_frames,\n",
    "    'env_vars_dict' : {\n",
    "        'CONCOORD_OVERWRITE' : '1',\n",
    "        'CONCOORDLIB' : concoord_lib\n",
    "    }\n",
    "}\n",
    "\n",
    "concoord_disco(     input_pdb_path=concoord_dist_pdb,\n",
    "                    input_dat_path=concoord_dist_dat,\n",
    "                    output_traj_path=concoord_disco_pdb,\n",
    "                    output_rmsd_path=concoord_disco_rmsd,\n",
    "                    output_bfactor_path=concoord_disco_bfactor,\n",
    "                    properties=prop)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "<div class=\"alert alert-info\">Visualizing a multi-model PDB with > 100 frames is highly dangerous. Please use the trajectory visualization below.</div>"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Show protein (if num_frames <= 100)\n",
    "\n",
    "if (num_frames <= 100):\n",
    "    view = nglview.show_structure_file(concoord_disco_pdb, default_representation=False)\n",
    "    view.add_representation(repr_type='line', selection='all', color='modelindex')\n",
    "    view.center()\n",
    "    view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "    view\n",
    "else:\n",
    "    #print(\"Visualizing a multi-model PDB with > 100 frames is highly dangerous. Please use the trajectory visualization below.\")\n",
    "    display(Markdown('<div class=\"alert alert-info\">Visualizing a multi-model PDB with > 100 frames is highly dangerous. Please use the trajectory visualization below.</div>'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "83c6cb2472434a3a8d654722e7c5100d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "    view = nglview.show_structure_file(concoord_disco_pdb, default_representation=False)\n",
    "    view.clear_representations()\n",
    "    view.add_representation(repr_type='backbone', selection='all', color='modelindex')\n",
    "    view.center()\n",
    "    view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "    view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:35:56,176 [MainThread  ] [INFO ]  Module: biobb_analysis.ambertools.cpptraj_rms Version: 5.2.0\n",
      "2026-02-12 09:35:56,178 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97\n",
      "2026-02-12 09:35:56,182 [MainThread  ] [INFO ]  Copy to stage: 1ake_dist.pdb --> sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97\n",
      "2026-02-12 09:35:56,184 [MainThread  ] [INFO ]  Copy to stage: 1ake_disco_traj.pdb --> sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97\n",
      "2026-02-12 09:35:56,194 [MainThread  ] [INFO ]  Copy to stage: 1ake_monomer.pdb --> sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97\n",
      "2026-02-12 09:35:56,196 [MainThread  ] [INFO ]  Launching command (it may take a while): cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/41e65ccb-88e6-4466-839e-2f2927db8bb1instructions.in\n",
      "2026-02-12 09:35:56,582 [MainThread  ] [INFO ]  Command 'cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/41e65ccb-88e6-4466-839e-2f2927db8bb1instructions.in' finalized with exit code 0\n",
      "2026-02-12 09:35:56,583 [MainThread  ] [INFO ]  \n",
      "CPPTRAJ: Trajectory Analysis. V6.18.1 (AmberTools)\n",
      "    ___  ___  ___  ___\n",
      "     | \\/ | \\/ | \\/ | \n",
      "    _|_/\\_|_/\\_|_/\\_|_\n",
      "\n",
      "| Date/time: 02/12/26 09:35:56\n",
      "| Available memory: 3.349 GB\n",
      "\n",
      "INPUT: Reading input from '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/41e65ccb-88e6-4466-839e-2f2927db8bb1instructions.in'\n",
      "  [parm /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_dist.pdb]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_dist.pdb' as PDB File\n",
      "\tReading bond info from CONECT records.\n",
      "\tNot reading bond info from LINK records.\n",
      "\tDetermining bond info from distances.\n",
      "Warning: Determining bond length parameters from element types for '1ake_dist.pdb'.\n",
      "Warning: 1 molecules have non-contiguous segments of atoms.\n",
      "\t       1        2 segments: 1-1353 (1353)  1356-1433 (78) \n",
      "Warning: The 'fixatomorder' command can be used to reorder the topology and any\n",
      "Warning:  associated coordinates.\n",
      "  [trajin /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_disco_traj.pdb 1 -1 1]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_disco_traj.pdb' as PDB\n",
      "  [center !@H*,1H*,2H*,3H* origin]\n",
      "    CENTER: Centering coordinates using geometric center of atoms in mask (!@H*,1H*,2H*,3H*) to\n",
      "\tcoordinate origin.\n",
      "  [autoimage]\n",
      "    AUTOIMAGE: To box center based on center of mass, anchor is first molecule.\n",
      "  [rms first !@H*,1H*,2H*,3H*]\n",
      "    RMSD: (!@H*,1H*,2H*,3H*), reference is first frame (!@H*,1H*,2H*,3H*).\n",
      "\tBest-fit RMSD will be calculated, coords will be rotated and translated.\n",
      "  [strip :WAT,HOH,SOL,TIP3,TP3,SOD,CLA,Na+,Cl-,NA,CL,K+,K]\n",
      "    STRIP: Stripping atoms in mask [:WAT,HOH,SOL,TIP3,TP3,SOD,CLA,Na+,Cl-,NA,CL,K+,K]\n",
      "  [strip !@CA]\n",
      "    STRIP: Stripping atoms in mask [!@CA]\n",
      "  [parm /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_monomer.pdb noconect [exp]]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_monomer.pdb' as PDB File\n",
      "\tNot reading bond info from CONECT records.\n",
      "\tNot reading bond info from LINK records.\n",
      "Warning: Atom 1288 CD in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1289 CD in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1290 NE in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1291 NE in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1292 CZ in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1293 CZ in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1294 NH1 in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1295 NH1 in res ARG 167 A has alternate location specifier B\n",
      "Warning: Atom 1296 NH2 in res ARG 167 A has alternate location specifier A\n",
      "Warning: Atom 1297 NH2 in res ARG 167 A has alternate location specifier B\n",
      "\tDetermining bond info from distances.\n",
      "Warning: Determining bond length parameters from element types for '1ake_monomer.pdb'.\n",
      "  [reference /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_monomer.pdb !:WAT,HOH,SOL,TIP3,TP3,SOD,CLA,Na+,Cl-,NA,CL,K+,K parm [exp]]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_monomer.pdb' as PDB\n",
      "Warning: PDB has 10 records wih alternate atom location IDs.\n",
      "\tStripped ref parm: 1661 atoms, 214 res, box: None, 1 mol\n",
      "\tSetting active reference for distance-based masks: '1ake_monomer.pdb'\n",
      "Warning: Active reference has 1661 atoms, parm '1ake_dist.pdb' has only 1656.\n",
      "Warning: Truncating reference coords for this parm (distance-based masks only).\n",
      "  [rms reference @CA out /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_concoord_rmsd.dat]\n",
      "    RMSD: (@CA), reference is \"1ake_monomer.pdb:1\" (@CA).\n",
      "\tBest-fit RMSD will be calculated, coords will be rotated and translated.\n",
      "---------- RUN BEGIN -------------------------------------------------\n",
      "\n",
      "PARAMETER FILES (2 total):\n",
      " 0: 1ake_dist.pdb, 1656 atoms, 214 res, box: None, 4 mol\n",
      " 1: [exp] 1ake_monomer.pdb, 1661 atoms, 214 res, box: None, 1 mol\n",
      "\n",
      "INPUT TRAJECTORIES (1 total):\n",
      " 0: '1ake_disco_traj.pdb' is a PDB file, Parm 1ake_dist.pdb (reading 300 of 300)\n",
      "  Coordinate processing will occur on 300 frames.\n",
      "\n",
      "REFERENCE FRAMES (1 total):\n",
      "    0: 1ake_monomer.pdb:1\n",
      "\tActive reference frame for distance-based masks is '1ake_monomer.pdb:1'\n",
      "\n",
      "BEGIN TRAJECTORY PROCESSING:\n",
      ".....................................................\n",
      "ACTION SETUP FOR PARM '1ake_dist.pdb' (6 actions):\n",
      "  0: [center !@H*,1H*,2H*,3H* origin]\n",
      "\tMask [!@H*,1H*,2H*,3H*] corresponds to 1656 atoms.\n",
      "  1: [autoimage]\n",
      "Warning: Topology 1ake_dist.pdb does not contain box information.\n",
      "Warning: Setup incomplete for [autoimage]: Skipping\n",
      "  2: [rms first !@H*,1H*,2H*,3H*]\n",
      "\tTarget mask: [!@H*,1H*,2H*,3H*](1656)\n",
      "\tReference topology: 1ake_dist.pdb\n",
      "\tReference mask: [!@H*,1H*,2H*,3H*](1656)\n",
      "  3: [strip :WAT,HOH,SOL,TIP3,TP3,SOD,CLA,Na+,Cl-,NA,CL,K+,K]\n",
      "\tStripping 0 atoms.\n",
      "Warning: No atoms to strip. Skipping.\n",
      "Warning: Setup incomplete for [strip]: Skipping\n",
      "  4: [strip !@CA]\n",
      "\tStripping 1442 atoms.\n",
      "\tStripped topology: 214 atoms, 214 res, box: None, 214 mol\n",
      "  5: [rms reference @CA out /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97/1ake_concoord_rmsd.dat]\n",
      "\tTarget mask: [@CA](214)\n",
      "\tReference topology: 1ake_monomer.pdb\n",
      "\tReference mask: [@CA](214)\n",
      "----- 1ake_disco_traj.pdb (1-300, 1) -----\n",
      " 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Complete.\n",
      "\n",
      "Read 300 frames and processed 300 frames.\n",
      "TIME: Avg. throughput= 2813.8102 frames / second.\n",
      "\n",
      "ACTION OUTPUT:\n",
      "TIME: Analyses took 0.0000 seconds.\n",
      "\n",
      "DATASETS (2 total):\n",
      "\tRMSD_00001 \"RMSD_00001\" (double, rms), size is 300 (2.400 kB)\n",
      "\tRMSD_00004 \"RMSD_00004\" (double, rms), size is 300 (2.400 kB)\n",
      "    Total data set memory usage is at least 4.800 kB\n",
      "\n",
      "DATAFILES (1 total):\n",
      "  1ake_concoord_rmsd.dat (Standard Data File):  RMSD_00004\n",
      "\n",
      "RUN TIMING:\n",
      "TIME:\t\tInit               : 0.0000 s (  0.02%)\n",
      "TIME:\t\tTrajectory Process : 0.1066 s ( 99.48%)\n",
      "TIME:\t\tAction Post        : 0.0000 s (  0.00%)\n",
      "TIME:\t\tAnalysis           : 0.0000 s (  0.00%)\n",
      "TIME:\t\tData File Write    : 0.0005 s (  0.47%)\n",
      "TIME:\t\tOther              : 0.0000 s (  0.00%)\n",
      "TIME:\tRun Total 0.1072 s\n",
      "---------- RUN END ---------------------------------------------------\n",
      "TIME: Total execution time: 0.3158 seconds.\n",
      "--------------------------------------------------------------------------------\n",
      "To cite CPPTRAJ use:\n",
      "Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for\n",
      "  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.\n",
      "  Theory Comput., 2013, 9 (7), pp 3084-3095.\n",
      "\n",
      "\n",
      "2026-02-12 09:35:56,585 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_8d0840ad-724c-40dd-b874-f6b4309a2e97']\n",
      "2026-02-12 09:35:56,586 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "concoord_rmsd = pdbCode + \"_concoord_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "cpptraj_rms(input_top_path=concoord_dist_pdb,\n",
    "            input_traj_path=concoord_disco_pdb,\n",
    "            output_cpptraj_path=concoord_rmsd,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "autobinx": false,
         "type": "histogram",
         "x": {
          "bdata": "jSjtDb4w9D/pt68D54z6P4qO5PIf0gJA6SYxCKycBEBLyAc9m1UGQGkAb4EExfs/qz5XW7E/AkDjx5i7lpDyPyBj7lpCPvM/AU2EDU+v9z+u2F92Tx78P4JzRpT2Bvc/FvvL7slDAEDtDb4wmSr0P0YldQKaCPU/3NeBc0aU+T+cM6K0N/j0P5Axdy0hH/c/rIvbaABv/D8fhetRuB79P0cDeAskKPU/FvvL7slDAEBfmEwVjEr1P7AD54wo7fI/jNtoAG+B/j+RD3o2qz77Pxlz1xLyQfQ/UdobfGEy9T8hH/RsVv0AQEvIBz2b1QBAPQrXo3A9+j91ApoIG575PyzUmuYdp/4/3bWEfNAzAEC8BRIUP8byP9O84xQdyQJApSxDHOvi9T9uNIC3QIIBQESLbOf7qQFAqz5XW7G/8z/HuriNBvD/P6g1zTtO0fg/mG4Sg8DK8j+bVZ+rrdgDQOxRuB6F6/U/TKYKRiV1/T9a9bnaiv3/P5ZDi2zn+/o/I9v5fmq8+D9oImx4eqX7PwmKH2Pu2gBA78nDQq1p/T/V52or9pf9P8rDQq1p3vc/guLHmLuWAkCSy39Iv/0BQHicoiO5/P8/6SYxCKwcAkCC4seYu5b2PzEIrBxaZPI/dy0hH/RsAUCh1jTvOEX4P/7UeOkmMQBAIGPuWkI+/D8yVTAqqRPuP662Yn/ZPfw/Io51cRsN/D8qqRPQRNj8P5tVn6ut2P4/OdbFbTSA8j+30QDeAgn8P/VKWYY41vU/AU2EDU+v9T/Qs1n1udrwP36MuWsJ+QFAKjqSy39I/T8dWmQ7308BQGiR7Xw/Nf8/MCqpE9BE/D924JwRpb3/PxQ/xty1hP4/Vp+rrdhf/j/VCWgibHj1Pxzr4jYawPo/C7Wmeccp9z9cIEHxY8z0P1yPwvUoXPI/Ke0NvjCZ+z9GlPYGX5j9PzVeukkMAvk/HVpkO99P+T/+ZffkYaH8P2B2Tx4W6gFAsVBrmnccAkAU0ETY8PT5P4BIv30dOPs/aJHtfD81+j+8lpAPejbxP6MBvAUSFPE/o5I6AU2E+D/GbTSAt0D8Pybkg57Nqvk/zhlR2ht8+z/arPpcbcX/Pwu1pnnHKf8/sHJoke189D+Uh4Va07wAQLFQa5p3nPU/Rdjw9EpZ/j+6SQwCK4f2P6wcWmQ73/M/RGlv8IXJ9z9/arx0kxj0P/JBz2bV5/c/hJ7Nqs/V+z+6SQwCK4cCQNjw9EpZhvQ/WKg1zTtO/j96Nqs+V1v5P7ivA+eMKPg/uycPC7Wm+T9ZF7fRAN7+P3S1FfvL7vg/5x2n6Egu+j/5oGez6nP4P0Jg5dAi2/4/bAn5oGczAUAYldQJaCL8P+zAOSNKewFAdZMYBFYO9j+3Yn/ZPXn7P42XbhKDwPE/2PD0SlmG9D8wTKYKRiX3P0Jg5dAiWwBARwN4CyQoAEAGgZVDi2wDQBQ/xty1hPs/L26jAbwF+T9CYOXQItv5P1TjpZvEIPQ/Dwu1pnlHAUBxGw3gLZDwPxueXinLEPM/HhZqTfOO/j9xrIvbaAD8P+SDns2qz/w/3nGKjuTy8D/Q1VbsL7v9Pylcj8L1KPk/tRX7y+5JAUDT3uALk6nzPxSuR+F6FPc/hA1Pr5Rl9j98YTJVMCoEQOm3rwPnjPo/Qj7o2az6+j+coiO5/If8P5p3nKIjuf4/93XgnBGlAEDHSzeJQeABQM/3U+Olm/4/3NeBc0aUAUApyxDHurj2P4JzRpT2Bvg/GXPXEvJB/j/ChqdXyjL6P1fsL7snD/k/BOeMKO0N+z/yQc9m1WcCQJjdk4eFWgJAgZVDi2zn/T89LNSa5h0BQLwFEhQ/xgJAMZkqGJXU+T8E54wo7Q39PzSitDf4wv4/2V92Tx4W+T8VHcnlP6TyP/rt68A5I/s/8BZIUPwY8z9CPujZrPrwP5F++zpwzvo/HThnRGlv9j+mCkYldQL3P2sr9pfdk/4/KVyPwvUo+j92cRsN4C36PwdfmEwVjPY/OGdEaW/w/D8X2c73U+MAQC9uowG8Bfw/9ihcj8L18z+MuWsJ+aD2P4qw4emVsgJAnzws1JrmAkBQjZduEoMAQJoIG55eKfk/RGlv8IXJ/D9LWYY41sUBQGaIY13cRgFA0gDeAgmK/z99PzVeukn0P4nS3uALk/Q/PE7RkVz+/T9Wfa62Yn/1P+58PzVeuv0/zO7Jw0Kt/j+dEaW9wRf9P6MjufyH9P0/j8L1KFyP+z9Wn6ut2F/4Py9uowG8BfQ/W7G/7J48/z83GsBbIEH4P18HzhlR2vM/WvW52op9A0B4CyQofozzP+Olm8QgsPY/+Q/pt68DAUBkO99PjZfzP5Axdy0hH/4/+8vuycPCA0DMXUvIBz3/P49TdCSX//g/1xLyQc9mBUBWn6ut2N8CQDMzMzMzM/4/Y3/ZPXlY+z+DL0ymCsYAQGTMXUvIB/w/si5uowG8/D9sCfmgZ7PwPxPyQc9m1fI/Dwu1pnnH/D/QRNjw9Er/PzPEsS5uo/g/t9EA3gIJ+T+28/3UeOnwP2Dl0CLb+fk/V1uxv+ye+D/yQc9m1efyP02EDU+vlAFAp3nHKToSAkANcayL22j3P7pJDAIrh/s/irDh6ZWy9D+rPldbsb/xP6RwPQrXo/E/Vp+rrdhf9D/dtYR80LP7P/mgZ7PqcwJA/tR46SYx8j9L6gQ0ETb7P0YldQKaCABA+THmriXk8z8m5IOezaryP2uad5yiI/k/SS7/If12AEA2PL1SliH/Pz0K16NwPfM/qFfKMsSx+D+J0t7gCxMBQNXnaiv2l/4/sb/snjysAEAAkX77OnABQPmgZ7Pq8wBAseHplbIM9z/wp8ZLN4n4P8/3U+OlGwJAOUVHcvkP9D+jkjoBTYT1P7gehetROAFA78nDQq1p9j+KjuTyH9L6PzlFR3L5D/8/g8DKoUW29j+RD3o2qz75P0vIBz2bVQBALUMc6+I29j9MN4lBYOX1PzAqqRPQRAJAxEKtad5x9z//If32deD7P+lILv8h/fw/JXUCmggb/j+x4emVsgz+PyUGgZVDi/g/YOXQItv5/j/CFyZTBaP9Px4Wak3zjvI/TRWMSuoEAkArGJXUCWj+P/aX3ZOHhfY/5x2n6Egu/j8v3SQGgZX5P1r1udqK/fw/FmpN845T+z9DrWnecYruP+C+DpwzovQ/",
          "dtype": "f8"
         },
         "xbins": {
          "size": 0.04
         }
        }
       ],
       "layout": {
        "height": 600,
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermap": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermap"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "RMSd variance"
        },
        "xaxis": {
         "title": {
          "text": "RMSd (Ångstroms)"
         }
        },
        "yaxis": {
         "title": {
          "text": "Population"
         }
        }
       }
      },
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABFQAAAJYCAYAAABM77IRAAAQAElEQVR4AezdB4AU1f048O/RBETFii12Y9QYy1+jMbHEGluMxoa9R7FEDXZFLDEaELGAXdHEirFE7KKosWIS/UXTjBoMqNg7SvN/b/DOA+64trs3u/tB3+zszJv33vfzBu7uezOznb7yhwABAgQIECBAgAABAgQIEKh0AfEVWKBT+EOAAAECBAgQIECAAAECBHInYEAE8i0goZLv+TE6AgQIECBAgAABAgTKRcA4CRCoKgEJlaqabsESIECAAAECBAgQ+EbAGgECBAi0XUBCpe12jiRAgAABAgQIECitgN4IECBAgEBuBCRUcjMVBkKAAAECBAhUnoCICBAgQIAAgUoVkFCp1JkVFwECBAgQaIuAYwgQIECAAAECBFokIKHSIiaVCBAgQCCvAsZFgAABAgQIECBAoCMEJFQ6Ql2fBAhUs4DYCRAgQIAAAQIECBCoAAEJlQqYRCEQKK6A1gkQIECAAAECBAgQIEBgVgEJlVlFvC9/AREQIECAAAECBAgQIECAAIEiC0ioFBm4Jc2rQ4AAAQIECBAgQIAAAQIECJSXQFsSKuUVodESIECAAAECBAgQIECAAAECbRFwzBwEJFTmgGMXAQIECBAgQIAAAQIECJSTgLESKJ2AhErprPVEgAABAgQIECBAgACBmQW8I0CgbAUkVMp26gycAAECBAgQIECAQOkF9EiAAAECMwQkVGY4WBIgQIAAAQIECFSmgKgIECBAgEBRBCRUisKqUQIECBAgQIBAWwUcR4AAAQIECJSDgIRKOcySMRIgQIAAgTwLGBsBAgQIECBAoAoFJFSqcNKFTIAAgWoXED8BAgQIECBAgACB9gpIqLRX0PEECBAovoAeCBAgQIAAAQIECBDImYCESs4mxHAIVIaAKAgQIECAAAECBAgQIFDZAhIqlT2/omupgHoECBAgQIAAAQIECBAgQKAVAhIqrcDKU1VjIUCAAAECBAgQIECAAAECBDpOoFQJlY6LUM8ECBAgQIAAAQIECBAgQIBAqQSqph8JlaqZaoESIECAAAECBAgQIECAwOwCthBom4CEStvcHEWAAAECBAgQIECAAIGOEdArAQK5EJBQycU0GAQBAgQIECBAgACByhUQGQECBCpRQEKlEmdVTAQIECBAgAABAu0RcCwBAgQIEGhWQEKlWSIVCBAgQIAAAQJ5FzA+AgQIECBAoNQCEiqlFtcfAQIECBAgEMGAAAECBAgQIFDmAhIqZT6Bhk+AAAECpRHQCwECBAgQIECAAIGGAhIqDTWsEyBAoHIEREKAAAECBAgQIECAQBEFJFSKiKtpAgRaI6AuAQIECBAgQIAAAQIEykdAQqV85spI8yZgPAQIECBAgAABAgQIECBQtQISKlU09UIlQIAAAQIECBAgQIAAAQIECiOQ54RKYSLsgFa++uqreO+Dj+O119+MDz76JKZP/6rFo0jHXH3TPfHvV8e3+JhSVPxy8pT49LNJMXXatFJ0pw8CBAgQIECAAAECBAgQqB6Bsoy0KhIq62x1SKy68b4zlZ/uc1JcecPdMbk2UdBw5l4Z90Z9vetG3t9wV/36cWdeWl9n2rTp9dsnfTE5Lrnuzvj+1ofGhjscGdvufWL8aPsjYrVN9ouDjx0cDz72XH3dplb+898Jcd6lt8Q/Xh7XVJUO2X7W0N/FutscGk899/cO6V+nBAgQIECAAAECBAgQyI+AkRCIqIqESt1E77HjZrH7DpvGVpusGylxcv7lI+PogcPqdmevXzW4muSK60fFlClTs+11izfeejfuHv103dv4qva/ujeX1iZTLr769ug+V9fYcesN44TDd8/6W3nFpeOJsS/GH+5+tK5q2b1+e7klY8P1Vo8Fes9TdmM3YAIECBAgQIAAAQJVLwCAAIGCC3QqeIs5bTAlAk46cs84+Zd7xeABh8bDI8+Pnj26x5gnn48JtUmSWYed9r3/4Sdx3yPPzrTrpjsfnul93Zv/vDYhu+IlJR4euOm8OPO4/WOvnbbI+rv1itNjyMDDYrE+C9VVn+013SY028YcbUixXHLO0bHqSsvkaFSGQoAAAQIECBAgUKkC4iJAgEDeBaomoTLrRPRZeP74yY+/n21OyZBspcFi1+1/nF2Nka5SqUt2pGeI3HjHw7H9lj+MdNVJg+rxr1f/l71dd61Vokf3btl6w8WWG68Tpx2zT8NNMWXqtBg+4o7Yes/j47s/3i/SbUgtvYpl6BW3xqEnnB/pipmZGq19c9s9j2X7/vbP12rfRQwaflPs+ovTY4OfHZHdqrRl32Nj8KU3Z895ySp8vTh32I1xzMDh2XNSbr/38Tj5nCvjyFMvjHfe+zDueuDJrM3/vfH217Vb3266Kii1n27BSuXU314dn3z6eX17dSvpap7DThqajXeTnY+OI06+IB549Lm63dlrSoTtf/S5kdpJJdUZN35its+CAAECBAgQINABArokQIAAgSoTqNqESprnSV98mV5ihWWXyF4bLrp36xb77PKT7NagPz37t2zXH2uTCp9P+iK78iTb0GCx1ndXzN498sRf482J72Xrc1qkJM1hJ54fw2oTKimpkG6n6dF9rnj8mRl9zenYtG/xPgvGY0+/EKMeeiq9rS/pAbgXX3N7PPfCv2KFZWbE9ccHnohXX38zVlphqdhy4xlJpGtuujf61SZkpjZ4yOxf/u/fcf+YZ2OPfmfFKedeFXfc96cY/fhfsofRpmRF6i8lleo6a227KWGU2l9uqcWyJlLi57e1yZ7szdeLNK70vJmUMPnud5aLPgsvEA/Xmh592sVf14hIdVLC5Zm//iM2WHe1WHrJPlmdlJh6+90P6+tZIUCAAAECBOYkYB8BAgQIECDQHoGqTai88PdX4t6Hn8l+GF9i0cZvxdl5240z26tuvCe7auPKG0bFD9ZedbarU1KlxWoTHGut9u0Y/+Y7sdmuv8qu9Eg/+D/13EuNXoWRrrhIV2Kk9tItQul2mpsvOy3OOeng1FyzZcuvr675w92PzfQpQs+/9HJMfOeD2GGrDeqvlLl8UP94atSwuHLwsTFkYL+49/rfxiY/XDNe/Ndr8d/X35qtr5RouuisI+ORW4dmdZvyaW27h+z903juvssjxXnfDb/NbrlKSZVpXz/Y9/UJE7MrZ5ZcbOF46ObzIpncOPzUePCmwbHdFutn40y+6eqadIXQE3deXBvPYZFuqTrr+AOy/SNuvjd7tSBAgACBChQQEgECBAgQIEAgRwJVk1BJz0NJV12ccPblsdcRZ8fu/c7Mbump+0G8sTmZb965Y48dN4+xz/8z0i02KVGx365bNVY125aezZISFelNuhIj/eB/YP9Bsd62/SLdOpQ+BSjtSyXtT699t9+0PvGR3ndv5HahtH3WMt88c8fPfvKjLIHz1xdfrt995/1PZOs7br1B9poWKfnQqaZTvPLfCfHoUy/Enff/KWo61aRd2fHZSoNFSlBs8qO1YpGFesdSSywS3bp1bbD3m9XWtJueSXPE/jvWx7rg/PNmV5ek1t7/8OP0Eg89/ufs9bD9fhYpQZW9qV0sXpvwqks01dXZb7etYp5ePSMlY1JJ462tGnW3OaV1hQABAh0toH8CBAgQIECAAIHKFaiahEqawvRckLseeDL+8rd/p7dxxeBjI11Vkr1pYrHnzzfL9qSrTZZfevFYf+1Vs/eNLdJzWS769S/jsdsvjOG/OTp+eeDPs0/GSXVTQua0QVen1az88z+vZ69rr7FS9tqWxc+32TA7LN2ak1Y+n/Rl3Drq0fjuSsvGd1ZYKm3KSkpCbLTjkfHTfU+Ofieen93OM/rxv2T7pn/1VfZat0iJj6YSKHV16l4fqk2AtLTdumMavs43b6/sbd0nKb329dUyq9aOP9vRyCLdepQ2p4+u/t6m+0ddWX+7w9LmmPDWO9mrBQECbRJwEAECBAgQIECAAAECLRSomoTKAr3niZfGjMjKoXtvn/GkK1ZSEiJ708RiqSX6xOYbrp3tPWD3baKmZsaVHdmGJhbp6ouNfrB6HLzndtltK38c8eusZvq45bqrVN5576NIY0pXmmQ727BY87srRro9Jt028/mkL+KRJ/+atbLLT3+cvaZFuuXol6deFF98OSWOO6xv/P7ik7NbeU45aq+0u82lEO12msUy3WqUBjRXE1fEpH2ffTYpvUS62uX0/vvFrKX/Ibtl+y2qRUCcBAgQIECAAAECBAgQ6BiBqkmoNOTtt+/PYtMN1op/vDwuTvrNFTM9g6Rhvbr1dAvKYfvtEFt9/dySuu0NX6c2eLhrw+1pffllloh11vhOWo03vr6CIj2Y9f0PP4m6qzOyna1c1NTURN8dNs2OGv34XyJdgZPebLnxOuklK48983/Z65CBh8U+O28ZKQmTbuXpPle3bHtbF8Vod9lvLZoN538TvvkkoWxDg8UyX9dZY9UVYqdtN5qtbL3pug1q53DVkAgQIECAAAECBAgQIECgIgSqMqHSqVNN/ObEgyPdwvPgY8/FxdfcNsfJXHHZJaPfPts3+SyRdPDvb30w+p9xSW3C5N30dqaSPvXnpX/9N9u21JIzkgbfWXHGLTkPPjbjuSFpZ/qEnrp66X1Lyrab/SCrdvE1t0e6aiQ9SLfX3D2ybWnxznsfppfo2rVz9poWKfmTkklpvbnS1P5itFt3q8+1I+/Lno3SsO/0ST/p/eq1iZT0muKdMnVaWq0v6SqdZFC/wQoBAgQIECBAgAABAgQIECiSQKUlVFrMNHfP7jH8nKOzT5q57Hd3RXq2SosPbqJi+tSgzXfrH4eecH5cfPXtceUNd2fr6VN/0g/7Zxy7f3TtMiOxsfsOm2WtHHvmJTFo+E0xfMQdscsvBmYPr812tHCx0ALzRboiJX36TTqk4cNo0/t1Vp/xjJbTBl2TjWlYbeJl54NOi+tveyjtbnMpRrsb/WD1WHfNlePxZ/4WB/UfFCNHjYmb7nw4e4jwESdfkI31h+t8N/uEovQg3p8fcGpcN/L+SJ909OsLfhdb9j02brxzdFbPggABAgQIECBAgAABAgQqWqDDg6uahErXrl1mw07PH7mkNqmSdpxw9uWRrg6pqUnvIupeo5k/NVGT1UifMnNA362z56I89vQLccl1d8b5l4+MtJ6uhLnorCOj7iGy6YCVlv9WDDr10LQaI265L4bVJlRS0iV9qlDa2NL+U90dtprxcNrUz/dWWT5tqi8/33ajqPs0oDSm4dfeGd27zxV1z1lpaT919Tp1mnHKFKLdTp1m2NV0mtFmTU1NnH/G4bH7DpvGM3/9RwwcPCLOPP+6SA/w3WPHGQmoFNigAYfG4fvvEG++/X6cO+zGGDDo6rjh9tHZp/5stsH/S1UUAgQIECBAgAABAgQI5EzAcCpNYMZPspUW1SzxjL330nh45PmzbJ3xdu3VV8oeVPvSmBGx6krLRHreSVpPz0yZUaPx5a1XnJ4d17nzDMKlllgkPYo0wAAAEABJREFUjvnFLvH4HRfF06OGx13X/SbuvObXMfbey+KP154dKeEya0tbb7pu/PWBK+K2q86M0SOHxD2/PzdOOnKPrN2fbvHDWas3+X6DdVfLjkn9zFqpS+fO8esTDsweRHvLZQOzfm4cfmqcdsw+2TE/Xn/N+kNuvuy02vFeWv++4UrySC4pEZS2F6Ldk3+5VzaGxRZZIDWZlfSQ3rT9hdFXxQM3DY4Ha8tTo4bVuuyZ7U+L9PyXQ/fePp6955IY84ehmXMyT36tcUttKQQIECBAgAABAgQINCFgMwECcxSYkQ2YYxU7WyswT6+ekR46u8KyS0TPHnPN8fBu3bpGSlIsuvA3SYU5HtDGnelBtClhVOh+itVuStgssehCsXhtSeuNhV1TUxMLL9g7knMyb6yObQQIECBAgAABAtUjIFICBAiUUqBTKTvTFwECBAgQIECAAAEC9QJWCBAgQKCMBSRUynjyDJ0AAQIECBAgUFoBvREgQIAAAQJ1AhIqdRJeCRAgQIAAgcoTEBEBAgQIECBAoEgCEipFgtUsAQIECBBoi4BjCBAgQIAAAQIEykNAQqU85skoCRAgkFcB4yJAgAABAgQIECBQlQISKlU57YImUM0CYidAgAABAgQIECBAgED7BSRU2m+oBQLFFdA6AQIECBAgQIAAAQIECOROQEIld1NS/gMSAQECBAgQIECAAAECBAgQqHQBCZWISp9j8REgQIAAAQIECBAgQIAAAQIRBTWQUCkop8YIECBAgAABAgQIECBAgEChBLSTZwEJlTzPjrERIECAAAECBAgQIECgnASMlUAVCUioVNFkC5UAAQIECBAgQIAAgZkFvCNAgEBbBSRU2irnOAIECBAgQIAAAQKlF9AjAQIECOREQEIlJxNhGAQIECBAgACByhQQFQECBAgQqEwBCZXKnFdRESBAgAABAm0VcBwBAgQIECBAoAUCEiotQFKFAAECBAjkWcDYCBAgQIAAAQIESi8goVJ6cz0SIECg2gXET4AAAQIECBAgQKDsBSRUyn4KBUCAQPEF9ECAAAECBAgQIECAAIGZBSRUZvbwjkBlCIiCAAECBAgQIECAAAECBIoqIKFSVF6Nt1RAPQIECBAgQIAAAQIECBAgUE4CEiptmy1HESBAgAABAgQIECBAgAABApUv0GSEEipN0thBgAABAgQIECBAgAABAgTKTcB4SyUgoVIqaf0QIECAAAECBAgQIECAwOwCthAoUwEJlTKdOMMmQIAAAQIECBAgQKBjBPRKgACBJCChkhQUAgQIECBAgAABApUrIDICBAgQKIKAhEoRUDVJgAABAgQIECDQHgHHEiBAgACB/AtIqOR/joyQAAECBAgQyLuA8REgQIAAAQJVJyChUnVTLmACBAgQIBDBgAABAgQIECBAoH0CEirt83M0AQIECJRGQC8ECBAgQIAAAQIEciUgoZKr6TAYAgQqR0AkBAgQIECAAAECBAhUsoCESiXPrtgItEZAXQIECBAgQIAAAQIECBBosYCESoupVMybgPEQIECAAAECBAgQIECAAIGOEpBQKZ28nggQIECAAAECBAgQIECAAIEKEZhDQqVCIhQGAQIECBAgQIAAAQIECBAgMAcBu9oiIKHSFjXHECBAgAABAgQIECBAgEDHCeiZQA4EJFRyMAmGQIAAAQIECBAgQIBAZQuIjgCByhOQUKm8ORURAQIECBAgQIAAgfYKOJ4AAQIEmhGQUGkGyG4CBAgQIECAAIFyEDBGAgQIECBQWgEJldJ6640AAQIECBAgMEPAkgABAgQIEChrAQmVsp4+gydAgAABAqUT0BMBAgQIECBAgMA3AhIq31hYI0CAAIHKEhANAQIECBAgQIAAgaIJSKgUjVbDBAgQaK2A+gQIECBAgAABAgQIlIuAhEq5zJRxEsijgDERIECAAAECBAgQIECgSgUkVKp04qs1bHETIECAAAECBAgQIECAAIFCCEioFEKxeG1omQABAgQIECBAgAABAgQIEMihQIETKjmM0JAIECBAgAABAgQIECBAgACBAgtoTkLFOUCAAAECBAgQIECAAAEClS8gQgIFFpBQKTCo5ggQIECAAAECBAgQIFAIAW0QIJBvAQmVds7PG+9NCoWBc8A54BxwDjgHnAPOAeeAc2CS74v9bOAcKLNzoJ0/Dlf94RIqVX8KACBAgAABAgQIVKuAuAkQIECAQNsFJFTabudIAgQIECBAgEBpBfRGgAABAgQI5EZAQiU3U2EgBAgQIECg8gRERIAAAQIECBCoVAEJlUqdWXERIECAQFsEHEOAAAECBAgQIECgRQISKi1iUokAAQJ5FTAuAgQIECBAgAABAgQ6QkBCpSPU9UmgmgXEToAAAQIECBAgQIAAgQoQkFCpgEkUQnEFtE6AAAECBAgQIECAAAECBGYVkFCZVaT834uAAAECBAgQIECAAAECBAgQKLJADhIqRY5Q8wQIECBAgAABAgQIECBAgEAOBCprCBIqlTWfoiFAgAABAgQIECBAgACBQgloh8AcBCRU5oBjFwECBAgQIECAAAECBMpJwFgJECidgIRK6az1RIAAAQIECBAgQIDAzALeESBAoGwFJFTKduoMnAABAgQIECBAoPQCeiRAgAABAjMEJFRmOFgSIECAAAECBCpTQFQECBAgQIBAUQQkVIrCqlECBAgQIECgrQKOI0CAAAECBAiUg4CESjnMkjESIECAQJ4FjI0AAQIECBAgQKAKBSRUqnDShUyAQLULiJ8AAQIECBAgQIAAgfYKSKi0V9DxBAgUX0APBAgQIECAAAECBAgQyJmAhErOJsRwKkNAFAQIECBAgAABAgQIECBQ2QISKpU9vy2NTj0CBAgQIECAAAECBAgQIECgFQJlmlBpRYSqEiBAgAABAhUvMOCMLlHIUvFgAiRAgAABAmUjkN+BSqjkd26MjAABAgQIECBAgAABAgTKTcB4q0ZAQqVqplqgBAgQIECAAAECBAgQmF3AFgIE2iYgodI2N0cRIECAAAECBAgQINAxAnolQIBALgQkVHIxDQZBgAABAgQIECBQuQIiI0CAAIFKFJBQqcRZFRMBAgQIECBAoD0CjiVAgAABAgSaFZBQaZZIBQIECBAgQCDvAsZHgAABAgQIECi1gIRKqcX1R4AAAQIEIhgQIECAAAECBAiUuYCESplPoOETIECgNAJ6IUCAAAECBAgQIECgoYCESkMN6wQIVI6ASAgQIECAAAECBAgQIFBEAQmVIuJqmkBrBNQlQIAAAQIECBAgQIAAgfIRkFApn7nK20iNhwABAgQIECBAgAABAgQIVK1AFSVUqnaOBU6AAAECBAgQIECAAAECBKpIoDShSqiUxlkvBAgQIECAAAECBAgQIECgcQFby1JAQqUsp82gCRAgQIAAAQIECBAg0HECeiZAIEJCxVlAgAABAgQIECBAgEClC4iPAAECBReQUCk4qQYJECBAgAABAgQItFfA8QQIECCQdwEJlbzPkPERIECAAAECBMpBwBgJECBAgECVCUioNJjwqdOmxfTpXzXYMvPqlClTY8Jb78bkyVNm3uEdAQIECBAgUHYCBkyAAAECBAgQaI+AhMrXepO+mBw77n9q3DP66a+3fPPy2utvxl5HnB1rbH5gbLFb/7jt3se/2WmNAAECBAiURkAvBAgQIECAAAECORKQUKmdjMGX3hxr/+TgeGXcG7XvZv5/4jsfxLZ7nxh9Fp4/fnfRSfHcfZfHlhuvM3Ml7wgQIECgEQGbCBAgQIAAAQIECFSugIRK7dwe2HebGD1ySJY0qX070//X3nJfLNB7njjn5INjrdW+HT26d4v555tnpjreECBQIQLCIECAAAECBAgQIECAQAsFJFRqoXrP1ysWXXiB6NqlS8z650/P/i0W77NQ9D/9ktj1F6fHwMEj4q133p+1mvcEOkRApwQIECBAgAABAgQIECDQMQISKs24p9uA5u7ZPTb90Vqxf9+t4sV/vRb7H31upAfUpkMXX7BHKC02YOV8cQ44B5wDzoGinAPpa3Ihi6/tvrY7B5wDzgHnQDWcA4X82lmNbUmozHHWZ+zcY8fNY7st1o8tN/5+DDr1kBg3fmK8+vqb2c433psUCgPngHPAOeAccA507DmQfVEu4MJ8dux88ufvHHAOOAdKcw4U8EtnBTTV+hAkVJoxW3nFpeP1CRPra02fPj1bnzxlavZqQYAAAQIECBAgQIAAAQIESi6gww4XkFCpnYKp06bV38IzZerU+vXaXbH1puvG1TfdExPeejc++uSz+N2tD2YPqV1hmSXSboUAAQIECBAgQIAAAQIEWiCgCoFKE5BQqZ3R4868LNbY/MAY/+Y7ccq5V2Xrr319S8+eO24e6661SmyxW/9Yf7vD4rFnXohhZx+VfdpP7aH+J0CAAAECBAgQIECgMgVERYAAgTkKSKjU8gwZ2C9eGjNiprLsUotF+tOtW9cYPODQeGrU8Hjo5vNi9C1D4nurLJ92KQQIECBAgAABAgRyJGAoBAgQIFBKAQmVFmrP26tnLNZnwaipqWnhEaoRIECAAAECBAjMUcBOAgQIECBQxgISKmU8eYZOgAABAgQIlFZAbwQIECBAgACBOgEJlToJrwQIECBAoPIERESAAAECBAgQIFAkAQmVIsFqlgABAgTaIuAYAgQIECBAgAABAuUhIKFSHvNklAQI5FXAuAgQIECAAAECBAgQqEoBCZWqnHZBV7OA2AkQIECAAAECBAgQIECg/QISKu031EJxBbROgAABAgQIECBAgAABAgRyJyChUvAp0SABAgQIECBAgAABAgQIECBQ6QKdotIjFB8BAgQIECBAgAABAgQIECAQwaCgAq5QKSinxggQIECAAAECBAgQIECgUALaIZBnAQmVPM+OsREgQIAAAQIECBAgUE4CxkqAQBUJSKhU0WQLlQABAgQIECBAgMDMAt4RIECAQFsFJFTaKuc4AgQIECBAgACB0gvokQABAgQI5ERAQiUnE2EYBAgQIECAQGUKiIoAAQIECBCoTAEJlcqcV1ERIECAAIG2CjiOAAECBAgQIECgBQISKi1AUoUAAQIE8ixgbAQIECBAgAABAgRKLyChUnpzPRIgUO0C4idAgAABAgQIECBAoOwFJFTKfgoFQKD4AnogQIAAAQIECBAgQIAAgZkFJFRm9vCuMgREQYAAgbISGHBGlyh0KSsAgyVAgAABAgQIlKGAhEouJs0gCBAgQIAAAQIECBAgQIAAgXISaFtCpZwiNFYCBAgQIECAAAECBAgQIECgbQKOalJAQqVJGjsIECBAgAABAgQIECBAoNwEjJdAqQQkVEolrR8CBAgQIECAAAECBAjMLmALAQJlKiChUqYTZ9gECBAgQIAAAQIEOkZArwQIECCQBCRUkoJCgAABAgQIECBQuQIiI0CAAAECRRCQUCkCqiYJECBAgAABAsj3/WIAABAASURBVO0RcCwBAgQIECCQfwEJlfzPkRESIECAAIG8CxgfAQIECBAgQKDqBCRUqm7KBUyAAAECEQwIECBAgAABAgQItE9AQqV9fo4mQIBAaQT0QoAAAQIECBAgQIBArgQkVHI1HQZDoHIEREKAAAECBAgQIECAAIFKFpBQqeTZFVtrBNQlQIAAAQIECBAgQIAAAQItFpBQaTFV3ioaDwECBAgQIECAAAECBAgQINBRAqVLqHRUhPolQIAAAQIECBAgQIAAAQIESidQJT1JqFTJRAuTAAECBAgQIECAAAECBBoXsJVAWwQkVNqi5hgCBAgQIECAAAECBAh0nICeCRDIgYCESg4mwRAIECBAgAABAgQIVLaA6AgQIFB5AhIqDeZ06rRpMX36Vw22WCVAgAABAgQIEKhKAUETIECAAIFmBCRUvgaa9MXk2HH/U+Oe0U9/vWX2l/MvHxmrbrxvfPzp57PvtIUAAQIECBAg0IECuiZAgAABAgRKKyChUus9+NKbY+2fHByvjHuj9l3j/99+7+Nx5Q13N77TVgIECBAgQKC1AuoTIECAAAECBMpaQEKldvoO7LtNjB45JPosPH/tu9n/H/v8P+PsC6+PwQMOnX2nLQQIECBQJQLCJECAAAECBAgQIPCNgIRKrUXv+XrFogsvEF27dIlZ/4wbPzH6nTg0hp5xeKy47JKz7o5ONaHUBIOaYFATDGoiXwY1YTw1URYGs31xKcCGavv6VACymZqoNj/xlse/FebJPDkHnAOFPgdm+uLnTasFJFTmQPbRx5/FwccOjqMP3jl+uM53G625yPw9QmHgHCjMOcCxfB1POb1LFLJU27nQ6BeYdm5k2D7AavMTb/n++2vuzJ1zwDnQnnOgfV8tHd0JQdMCT//lpRj/5jvxvzfejt8OuzGuvHHGM1SGXnFr/OPlcdmBb70/KZSqNTD3zn/nwNfnQPYPYgEX1fbvagHp6ptiWE/RppVq8xOv72WcA84B50B1ngNt+iLpoHoBCZV6itlXVlhmifjlgT+P+efrFem2oHl79cwq9Z537ujWtUu2Xl4LoyVAgAABAgQIECBAgAABAgQKIZDvhEohImxBG1OnTYspU6ZmNadMnVq/vnxtQuXgPbeLurLLdj/O6uy761aR9mVvLAgQIECAAAECBAgQIECAAIH2CZTh0RIqtZN23JmXxRqbH5jd3nPKuVdl66+9/mbtHv8TIECAAAECBAgQIECAAIHZBWwhIKFSew4MGdgvXhozYqay7FKLxax/Vlh2iaxO3a0/s+73ngABAgQIECBAgAABAjkVMCwCBAosIKFSYFDNESBAgAABAgQIECBQCAFtECBAIN8CEir5nh+jI0CAAAECBAgQKBcB4yRAgACBqhKQUKmq6RYsAQIECBAgQOAbAWsECBAgQIBA2wUkVNpu50gCBAgQIECgtAJ6I0CAAAECBAjkRkBCJTdTYSAECBAgUHkCIiJAgAABAgQIEKhUAQmVSp1ZcREgQKAtAo4hQIAAAQIECBAgQKBFAhIqLWJSiQCBvAoYFwECBAgQIECAAAECBDpCQEKlI9T1Wc0CYidAgAABAgQIECBAgACBChCQUKmASSxuCFonQIAAAQIECBAgQIAAAQIEZhWovITKrBF6T4AAAQIECBAgQIAAAQIECFSeQAdHJKHSwROgewIECBAgQIAAAQIECBCoDgFRVpaAhEplzadoCBAgQIAAAQIECBAgUCgB7RAgMAcBCZU54NhFgAABAgQIECBAgEA5CRgrAQIESicgoVI6az0RIECAAAECBAgQmFnAOwIECBAoWwEJlbKdOgMnQIAAAQIECJReQI8ECBAgQIDADAEJlRkOlgQIECBAgEBlCoiKAAECBAgQIFAUAQmVorBqlAABAgQItFXAcQQIECBAgAABAuUgIKFSDrNkjAQIEMizgLERIECAAAECBAgQqEIBCZUqnHQhV5bAgDO6RKFLZQnNHo0tBAgQaE7Av6vNCdlPgAABAgQISKg4BwjkX8AICRAgQIAAAQIECBAgQCBnAhIqOZuQyhiOKAgQIECAAAECBAgQIECAQGULSKik+VUIECBAgAABAgQIECBAgACByhcoYIQSKgXE1BQBAgQIECBAgAABAgQIECikgLbyKyChkt+5MTICBAgQIECAAAECBAiUm4DxEqgaAQmVqplqgRIgQIAAAQIECBAgMLuALQQIEGibgIRK29wcRYAAAQIECBAgQKBjBPRKgAABArkQkFDJxTQYBAECBAgQIECgcgVERoAAAQIEKlFAQqUSZ1VMBAgQIECAQHsEHEuAAAECBAgQaFZAQqVZIhUIECBAgEDeBYyPAAECBAgQIECg1AISKqUW1x8BAgQIRDAgQIAAAQIECBAgUOYCEiplPoGGT4BAaQT0QoAAAQIECBAgQIAAgYYCEioNNawTqBwBkRAgQIAAAQIECBAgQIBAEQUkVIqIq+nWCKhLgAABAgQIECBAgAABAgTKR0BCpa1z5TgCBAgQIECAAAECBAgQIECg8gWaiFBCpQkYmwkQIECAAAECBAgQIECAQDkKGHNpBCRUSuOsFwIECBAgQIAAAQIECBBoXMBWAmUpIKHSYNqmTpsW06d/1WDLjNW0/c23348vJ0+ZscGSAAECBAgQIECAAIEqFhA6AQIEIiRUvj4LJn0xOXbc/9S4Z/TTX2+Z8XLF9aNi9U0PiM12OSbW2uKgOGbgsPjo489m7LQkQIAAAQIECBAgUA4CxkiAAAECBReQUKklHXzpzbH2Tw6OV8a9Uftu5v97z9crrhpyXDx33+Vx+9Vnxdjn/xm33/v4zJW8I0CAAAECBAgQKKiAxggQIECAQN4FJFRqZ+jAvtvE6JFDos/C89e+m/n/nbfdONZba5Xo0b1bfHu5JWPj9deMx55+YeZK3hEgQIAAAQLVLiB+AgQIECBAoMoEJFRqJzxdhbLowgtE1y5dat81/f+UqdPiibF/i1VXWra+UpfONaEw6MhzoP5kLOBKR8aj7/L8+1TA0y9rqtrOgyzoAi+aNyzPc62puArMV/Dmmhq37ZV1HppP8+kccA6U2zlQ8C94VdaghEorJvysodfFJ59Oir122qL+qAXmmSsUBh15DtSfjAVc6ch4OqLvk07rHIUsHRFDR/dZwNMva6rR+WjHPBXap9Djy4Iu8KLQMee9vQLzFby5vPsZn+9lnAPOAedAdZ4DBf+CV2UNSqi0cMKHj7gjbh31aFx9/vGxyEK96496+8MvQmHQkedA/clYwJWOjKcj+i4gXdZUXQzV9JoFnuNFoecix6HWD63QMee9vfrAc7qSdz/j872Mc8A54ByoznMgp182y2ZYEirNTFX6GOVBw2+Ka26+L0ZePjBW+843t/s0c6jdBMpNwHgJECBAgAABAgQIECBAoIUCEiq1UFOnTYspU6bWrkVMmTq1fj1tGDDo6hhxy30xZOBhMd+8vWLCW+9mJR2T9isdKaBvAgQIECBAgAABAgQIECDQMQISKrXux515Wayx+YEx/s134pRzr8rWX3v9zdo9kX1Mclo55PjzYovd+teXCW++mza3rqhNgAABAgQIECBAgAABAgQIVITAHBMqFRFhC4IYMrBfvDRmxExl2aUWi/Tn/hsHzbS9rt7SS/ZJuxUCBAgQIECAAAECBAgQIFD2AgJovYCESuvNHEGAAAECBAgQIECAAAECHSugdwIdLiCh0uFTYAAECBAgQIAAAQIECFS+gAgJEKg0AQmVSptR8RAgQIAAAQIECBAohIA2CBAgQGCOAhIqc+SxkwABAgQIECBAoFwEjJMAAQIECJRSQEKllNr6IkCAAAECBAh8I2CNAAECBAgQKGMBCZUynjxDJ0CAAAECpRXQGwECBAgQIECAQJ2AhEqdhFcCBAgQqDwBEREgQIAAAQIECBAokoCESpFgNUuAAIG2CDiGAAECBAgQIECAAIHyEJBQKY95MkoCeRUwLgIECBAgQIAAAQIECFSlgIRKVU57NQctdgIECBAgQIAAAQIECBAg0H4BCZX2Gxa3Ba0TIECAAAECBAgQIECAAAECuRMoeEIldxEaEAECBAgQIECAAAECBAgQIFBwgWpvUEKl2s8A8RMgQIAAAQIECBAgQKA6BERJoKACEioF5dQYAQIECBAgQIAAAQIECiWgHQIE8iwgoZLn2TE2AgQIECBAgAABAuUkYKwECBCoIgEJlSqabKESIECAAAECBAjMLOAdAQIECBBoq4CESlvlHEeAAAECBAgQKL2AHgkQIECAAIGcCEio5GQiDIMAAQIECFSmgKgIECBAgAABApUpIKFSmfMqKgIECMxRYMAZXaKQZY6dldtO4yVAgAABAgQIECDQAgEJlRYgqUKAAIE8CxgbAQIECBAgQIAAAQKlF5BQKb25HglUu4D4CRAgQIAAAQIECBAgUPYCEiplP4UCKL6AHggQIECAAAECBAgQIECAwMwCEioze1TGO1EQIECAAAECBAgQIECAAAECRRXIRUKlqBFqnAABAgQIECBAgAABAgQIEMiFQCUNQkKlkmZTLAQIECBAgAABAgQIECBQSAFtEWhSQEKlSRo7CBAgQIAAAQIECBAgUG4CxkuAQKkEJFRKJa0fAgQIECBAgAABAgRmF7CFAAECZSogoVKmE2fYBAgQIECAAAECHSOgVwIECBAgkAQkVJKCQoAAAQIECBCoXAGRESBAgAABAkUQkFApAqomCRAgQIAAgfYIOJYAAQIECBAgkH+BXCZUJr7zQdx2z2Mx9IpbZytffDk5/6pGSIAAAQLVJSBaAgQIECBAgACBqhPIXULl/jHPxiY7Hx2n/vbq+MPdj8YfH3hipjJlytSqmyQBEyBAoNAC2iNAgAABAgQIECBAoH0CuUuoXH3jvbHumivH2Hsvi8fvuCgeHnn+TGWeXj3bF7GjCRAoRwFjJkCAAAECBAgQIECAQK4EcpdQmfTFl7H2Gt+Jnj3myhWUwRBonYDaBAgQIECAAAECBAgQIFDJArlLqGz4g9XjybEvVrJ5PmMzKgIECBAgQIAAAQIECBAgQKDFArlLqKy47JLx1xdfjsGX3hzX3/bQbGXy5ClZcMVYTJ02LaZP/6rRpj/59PP44KNPGt1nIwECBAgQIECAAAECBAgQIFAcgby2mruEyiNP/DWzuuame+PsC38/W5lUpE/5mfTF5Nhx/1PjntFPZ/3XLT6f9EUccfIFsd62/eJH2x8RffudGe++/1Hdbq8ECBAgQIAAAQIECBAgQKChgPUqEchdQmXoGYfHS2NGNFnmm2fugk9Nuhpm7Z8cHK+Me2O2tm+4fXT8+9Xx8citQ+PpUcOjc6dOccGVf5itng0ECBAgQIAAAQIECBAoTwGjJkCgLQK5S6g0DOK9Dz6ON956N9KtOA23F3r9wL7bxOiRQ6LPwvPP1vR9jzwbO227USyyUO9InzC0106bx233PBbel6zUAAAQAElEQVRffdX4rUGzNWADAQIECBAgQIAAAQKFFdAaAQIEciCQy4TK7fc+Hhv87IjYcIcjY/Pd+sfqmx4Qpw+5Nj765LOikPWer1csuvAC0bVLl5j1z7jxE2OpJfrUb/7W4otk6x9/+nn2akGAAAECBAgQIECgOQH7CRAgQKDyBHKXUBn14FNxyrlXxfLLLBFHHbRTnHHs/rHNpuvFLX98JE749eUlvTIkXYWSnqHSfa5u9TM/V7eu2frnn3+RvXbt0qk2EaNw6LhzIIrwp9rms9CE5eBX6Jjz3l6h5yTv8abxFTrmvLeXYs5zybtfI+Pz/Y3v8ZwDzgHnQBWcA3n+2lkOY8tdQuWmOx+ONb+7YowYekIctMe28fNtNozfnnpIDOy/bzz29Avx5sT3SuZaU1MTPXt0jy8nf/PJQnXrPXt2z8bRe+6uoVS2wYkDOkUhS6HPl+xELPCi0GPMe3sF5iuLfxMKHXPe2yv0OZj3eNP4CvnvVjm0FZGizm8p9Dmovcr+3sP8ml/ngHOgVOdAfr9ylsfIcpdQGTf+rdjkR2vOpvej738v2zbhrXez11Itll6yT7w+YWJ9d/974+1sfd5ePbPXdz76MpTKNsgmuoCLQp8vBRxafVOFHmPe26sPvEAreY83ja9AoZZNMynmQpaCBK6RqhIo5Pmnrcr+vsP8ml/ngHOglOdAVX0xLkKwuUuorLrSMnHHvX+K6dNnfujrXQ88kYW/9JKLZq+FXKSH3k6ZMjVrcsrUqVG3njZsufE6MfKuMfH2ux/Gp59Nit/d+mDsuPWGUVNTk3YrBAgQqBoBgRIgQIAAAQIECBAg8I1A7hIqB+2xXfbxxRvteGSccPblkT7SeIf9T8k+qjglMtKn7Xwz/MKsHXfmZbHG5gfG+DffyZ7fktZfe/3NrPHdd9gsllt68fjxTkfFutscmiVbjth/x2yfBQECuRYwOAIECBAgQIAAAQIECBRNIHcJlf/3vW/H7y46KVZaYakY/fhf4pqb7o1p06bHCYfvHgOO3rsoEEMG9ouXxoyYqSy71GKR/szds3tccs7R8eRdw+LR2y6Imy87LfsI5bRPIVBYAa0RIECAAAECBAgQIECAQLkI5C6hkuDWWu3bceXgY2PsvZfGi49cE3+89uzYa6ctomvXLml3h5T55pk7Flpgvg7pO7edGhgBAgQIECBAgAABAgQIEKhSgVwkVKZOmxbp44nTxxTPOg81NTWzbmrzewcSIECAAAECBAgQIECAAAEClS9QighzkVD50zN/i3W2OiRen/B2HDNwWKy68b5Nlo8++awULvogQIAAAQIECBAgQIAAAQKlEtBPGQrkIqGy1JJ94hd7bRfptprttlg/e15KemZKY6X7XN3KkNmQCRAgQIAAAQIECBAgUEkCYiFAIBcJleWWWiyOPODn0Xu+XvG9lZeP7X/yo+yZKem5KXVl5+02jk1+tFZ068DnqDhdCBAgQIAAAQIECBAoUwHDJkCAQIEFcpFQaRjTwMHXxPW3PdhwU7Y+/o13Yovd+sdbb78f/hAgQIAAAQIECBCodAHxESBAgEC+BXKXUGmKa+6e3bNdH3/6efZqQYAAAQIECBAgkCsBgyFAgAABAlUlkJuEys13PhzX3HRvvDLujfjz//07W0/vU7n893fFUQMujgV6zxMrLLNEVU2QYAkQIECAAIFiCWiXAAECBAgQINB2gdwkVK6uTaYMvvTmGDd+Yjz13EuR1uvKFdffHUsv2ScGDTg0OnfOzZDbru5IAgQIECDQFgHHECBAgAABAgQI5EYgN9mJ+28cFC+NGRH77LxlDOy/b7ae3qcy9t5L47enHhLrrbVKbuAMhAABAgSaF1CDAAECBAgQIECAQKUK5CahUgd83GF9Y+dtN65765UAAQKlFNAXAQIECBAgQIAAAQIEWiSQu4TK5MlTYugVt8auvzg9tt7z+NnKJx5K26KJValaBMRJgAABAgQIECBAgAABAh0hkLuEylU33RNXXD8qFuuzYPY8lXXXWiU2+dFa8c57H0XPHt2jS5cuHeGkz0IJaIcAAQIECBAgQIAAAQIECFSAQO4SKmOeeD4O6Lt1nH3iQRnvfrv+JPofsmscd9huMfGd96Nr187Z9lIt9EOAAAECBAgQIECAAAECBAhUvkBrI8xdQuWd9z+MpZboEz17zJXF8sbE97LXdddcOd7/8JN4+dXx2XsLAgQIECBAgAABAgQIECBQxQJC72CB3CVUFug9b7z93ocZyw/WXjUefPS5bH3Cm+9mr13d8pM5WBAgQIAAAQIECBAgQKC8BIyWQGUJ5C6hstxSi8Xjz/xfprz9Fj+Mm+58OH66z0lxYP9B8e3llowVll0i22dBgAABAgQIECBAgACBogponAABAnMQyF1CZcAx+8TFv/5lNuTttlg/zjh2/1h6yT5x2H47xKXn/irbbkGAAAECBAgQIECAwOwCthAgQIBA6QRyl1DpNXePWHD+eaPuz8+32TAuqk2w9Ntn++iz8Px1m70SIECAAAECBAiUv4AICBAgQIBA2QrkIqHy+oSJ8dK//tuiMm3a9LLFNnAC5SIw4IwuUchSLnEbJwECBJoXUIMAAQIECBAgMEMgFwmVQcNvil1+MbBF5dPPJ80YuSUBAgQIECDQvIAaBAgQIECAAAECRRHIRULl2H59Y+TlA1tUevXsURQIjRIgQIBAPgSMggABAgQIECBAgEA5COQiobLUEovEKt9epkWlc+dcDLkc5tYYCRAojYBeCBAgQIAAAQIECBCoQoHcZSfuHzM2rrh+VJPly8lTqnCahEygkALaIkCAAAECBAgQIECAAIH2CuQuofLIk3+Ny38/arYy9IpbIxUJlfZOeRkeb8gECBAgQIAAAQIECBAgQCBnArlLqJxz0sEx9t5LZyvbbLpebLnx92PeXj1zRjj7cGwhQIAAAQIECBAgQIAAAQIEKlsgJVTKIsI9fr553D/m2Xj73Q/LYrwGSYAAAQIECBAgQIAAAQIEciZgOAUUKJuEytw9umdhv/u+hEoGYUGAAAECBAgQIECAAIGKFxAggfwK5C6h8vSf/x633/v4TOW6kffHsWdeEn0Wnj9WWn6p/GoaGQECBAgQIECAAAEC1S0gegIEqkagU94ivf62B+OUc6+aqZw77MZYdqnF4jcnHRydO+duyHkjNB4CBAgQIECAAAECLRZQkQABAgTaJpC77MSQ0w+PP99/+UzlhdFXxZCBh8W6a67ctigdRYAAAQIECBAgUCkC4iBAgAABArkQyF1CpWuXztF9rm4xV7eu8cUXk7PSpXPnXGAZBAECBAgQIECg9QKOIECAAAECBCpRIHcJlWnTpsdFV98W39/60Pjh9odnZZ2tDonLfndXfDl5SiXOgZgIECBAgEC+BIyGAAECBAgQIECgWYHcJVRuunN0XHrdH2PN764QRx+8cxx3WN9s/cKr/hC/ufD6ZgNSgQABAgSqT0DEBAgQIECAAAECBEotkLuEylU33hPfXWnZuHxQ/zhw921in523zNYP6Lt1jBw1Jj765LNSG+mPAAEChRbQHgECBAgQIECAAAECZS6Qu4TKwgv0jh+svepsrNtv+cNs21tvv5+9WhAgUEoBfREgQIAAAQIECBAgQIBAQ4HcJVS2/PE68cCjY2PqtGkNxxmffv5F9n6xPgtmr6VefD7pi/joY1fHlNq9zf05kAABAgQIECBAgAABAgQIFFEgdwmVSZO+jHHjJ8Z+R50bR556YX059dyrMoZTzr0y2zZw8IjsfbEXE9/5II44+YLYaMejYrNdfxV7HXF2/OPlcQXvVoMECBAgQIAAAQIECBAgQIBA+Qi0NaFStAg7de4UG663evSau0dMmTKtviyx2MLZ9vptU6cWbQwNGz7v0puzTxd68o8Xx1OjhsUy31o0Lrjy1oZVrBMgQIAAAQIECBAgQIAAgUoVEFcTArlLqBy69/ZxyTlHN1t+fcKBTYRU2M1vTHwvFl6wd3Tt2iW6dO4ca622Yvz71fGF7URrBAgQIECAAAECBAgQIFAgAc0QKI1A7hIqdWGn234eevzPcdcDT8ZfX3x5tmeq1NUr9uv+u20Vd9z3p+w2o0ee/Gtccf2o6LfPz4rdrfYJECBAgAABAgQIEKgWAXESIFCWArlLqEyZMjVO+s0VsfWex8cvT70oTjj78tjz8F/HT/c5qUOuDFlphaVi6SX7RKeaTnHcmZfFJ59+HmusukL9ZHfr2imUyjaon+wCrcxVe84UshRoWEVtppDxFqOtQgdfjDEWus1Cx5z39vjlfYYqf3yFPge11ykYdKwBf/7FPgf8jNWpJD9nVv5X4OJGmLuEyhU33B133v9EHL7/DvH7i0+Ou677TZzef79M4agBF5X8SpVjThsW226+fgw94/B4eOSQWGeNlaNvvzPrxzFP9y6hVLZBdvIVcDF37TlTyFLAoRWtqeNP7RSFLIX0S20VOvDUZiFLIe3q2ip0zHlvr5DzkdrKe7zGlz+BdN5UW6n79yYnrwX9OtRYTNU2v+LtEgyKa+BnrNL8jJW/r5jlNaJOeRvufQ8/E9tsul6kZ6ms+d0VY7mlFoudtt0oTjxij+zTf8b9762SDfmzz7+IF//1Wnxn+W9lfc7Tq2cc0Hfr+HzSF/Ha629m2977ZHIolW2QTXQBF+/XnjOFLAUcWtk0VUi/1FahA09tFrIUenzV2F4h5yO1VY2GYm6fQDpvWlcmR7nXb59Y+R1d7vNl/OX/d67S5tDPWKX5Gav8/rXN14hzl1D5cvKU7BabWZkWX3ShbNNHn3yWvZZiMXfP7rHkYgvHLXc9EqnfdDvSPaOfzsa3bG2ipxRj0AcBAgQIECgLAYMkQIAAAQIECFSZQO4SKmuutmKMuOX+eGXcG/HVV19l0/HBR5/EZdf9MVtfafmlstdSLdKtPt26dY31tzss1v/p4fHq62/GuSf/IvvEn1KNQT8ECBAgUHgBLRIgQIAAAQIECBBoj0DuEiq/PODnWTzpIbQb7nBk7LD/KfGj7Y+Iu0c/HacevXekq0ayCiVarLzi0nHhmUfGs/dcGg/dfF72cc6rrbxciXrXDQECBOoFrBAgQIAAAQIECBAgkCOB3CVUFuuzYDx0y3lx1EE7ZQ+AXXSRBWOvnbaIWy4bGLttv0mH0aVEznzzzt1h/euYQPkJGDEBAgQIECBAgAABAgQqVyBXCZX7x4yNsy/8fVx63R9jrdW+HUMG9suuCDnh8N1j1ZWWqdxZEFk+BIyCAAECBAgQIECAAAECBAi0UCA3CZWURDlm4LC4/raH4rqR98feR56dfXxyC+OoymqCJkCAAAECBAgQIECAAAECBDpGoJQJlSYj/HzSF3HR1bfFjltvGC+Mviqeu+/y2HLj78dZQ38X6VN/mjzQDgIECBAgQIAAAQIECBAgQCBvAlUxnlwkVN54670Me//dtso+PadH77SGWQAAEABJREFU925x+P47REq0THjr3WyfBQECBAgQIECAAAECBAgQKI6AVgm0XiAXCZWPP/08G/mCC8yXvabFogsvkF7infc+zF4tCBAgQIAAAQIECBAgQOBrAS8ECHS4QC4SKnUKD4wZG3ePfjorDz72XLb5ybEvZu/rtk+ZMjXbbkGAAAECBAgQIECAQPkIGCkBAgQqTSBXCZXTBl8Tx515aVZO+s0VmfWVN9ydva/b/vkXX2bbLQgQIECAAAECBAgUUUDTBAgQIEBgjgK5SKh8d6Vl4qFbhrSozDN3zzkGZCcBAgQIECBAoDoFRE2AAAECBAiUUiAXCZVu3brGYoss0KLSqVNNKX30RYAAAQIECBRLQLsECBAgQIAAgTIWyEVCpYz9DJ0AAQIEqkhAqAQIECBAgAABAgTqBCRU6iS8EiBAoPIERESAAAECBAgQIECAQJEEJFSKBKtZAgTaIuAYAgQIECBAgAABAgQIlIeAhEp5zJNR5lXAuAgQIECAAAECBAgQIECgKgUkVKps2oVLgAABAgQIECBAgAABAgQItF8g7wmV9keoBQIECBAgQIAAAQIECBAgQCDvAmU3PgmVspsyAyZAgAABAgQIECBAgACBjhcwgmoXkFCp9jNA/AQIECBAgAABAgQIVIeAKAkQKKiAhEpBOTVGgAABAgQIECBAgEChBLRDgACBPAtIqOR5dnI6tgFndIlClpyGaVg5Fijk+ZfaKnSoqc1ClkKPT3sECBAgUDSBgjZcyK8lqa2CDk5jBAgQIBASKk4CAgQIECBAgEDVCgicAAECBAgQaKuAhEpb5RxHgAABAgQIlF5AjwQIECBAgACBnAhIqORkIgyDAAECBCpTQFQECBAgQIAAAQKVKSChUpnzKioCBAi0VcBxBAgQIECAAAECBAi0QEBCpQVIqhAgkGcBYyNAgAABAgQIECBAgEDpBSRUSm+ux2oXED8BAgQIECBAgAABAgQIlL2AhErZT2HxA9ADAQIECBAgQIAAAQIECBAgMLNAJSZUZo7QOwIECBAgQIAAAQIECBAgQKASBTo0JgmVDuXXOQECBAgQIECAAAECBAhUj4BIK0lAQqWSZlMsBAgQIECAAAECBAgQKKSAtggQaFJAQqVJGjsIECBAgAABAgQIECg3AeMlQIBAqQQkVEolrR8CBAgQIECAAAECswvYQoAAAQJlKiChUqYTZ9gECBAgQIAAgY4R0CsBAgQIECCQBCRUkoJCgAABAgQIVK6AyAgQIECAAAECRRCQUCkCqiYJECBAgEB7BBxLgAABAgQIECCQfwEJlfzPkRESIEAg7wLGR4AAAQIECBAgQKDqBCRUWjHlU6ZMjQlvvRuTJ09pxVGqEiCQPwEjIkCAAAECBAgQIECAQPsEJFRa4Pfa62/GXkecHWtsfmBssVv/uO3ex1twlCoECiigKQIECBAgQIAAAQIECBDIlYCESjPTMfGdD2LbvU+MPgvPH7+76KR47r7LY8uN12nmKLsJECBAgAABAgQIECBAgACBShaQUJkxu00ur73lvlig9zxxzskHx1qrfTt6dO8W8883T5P17SBAgAABAgQIECBAgAABAgRyK1CwgUmoNEP5p2f/Fov3WSj6n35J7PqL02Pg4BHx1jvvN3OU3QQIECBAgAABAgQIECBAoBAC2sirgIRKMzPzyrg3Yu6e3WPTH60V+/fdKl7812ux/9HnRnpAbTq0e7dOUW0lxV3Ikne/Qsaa2ip0vKlNhQCBOQt079a59t/qwpU592YvgdkFCn0OlkN7sytU9pZizEmhxYoxxuK2WX3fZ3evlJ8txFH7fUd5nL+F/nem2tqTUGnBjO+x4+ax3Rbrx5Ybfz8GnXpIjBs/MV59/c3syB7dukS1lSjwn7z7FTjcOO6UTgUthR6f9ghUosBxp9TU/r0rXKlEIzEVV6DQ52Ax2utR+wNQIUtxRfPXeiHt6toqdJTFOG8K2WZd3N+8dinZ99l5/37U+JwLxToHwp92CUioNMO38opLx+sTJtbXmj59erY+ecrU7PWDTydHtZUs8AIu8u5XwFA1RYAAAQIEcivwwadTar+nKVzJbaBFGlih/VJ7cxhqRe5KMc9cqu/77A+q8GcLMXfseV6R/5iUMCgJlWawt9503bj6pntiwlvvxkeffBa/u/XB7CG1KyyzRDNH2k2AAAECBAgQIDBDwJIAAQIECFSegIRKM3O6546bx7prrRJb7NY/1t/usHjsmRdi2NlHZZ/208yhdhMgQIAAAQLlKmDcBAgQIECAAIFmBCRUmgHq1q1rDB5waDw1ang8dPN5MfqWIfG9VZZv5ii7CRAgQIBAaQX0RoAAAQIECBAgUFoBCZUWes/bq2cs1mfBqKmpaeERqhEgQIDAHATsIkCAAAECBAgQIFDWAhIqZT19Bk+AQOkE9ESAAAECBAgQIECAAIFvBCRUvrGwRqCyBERDgAABAgQIECBAgAABAkUTkFApGq2GWyugPgECBAgQIECAAAECBAgQKBcBCZW2z5QjCRAgQIAAAQIECBAgQIAAgcoXaDRCCZVGWWwkQIAAAQIECBAgQIAAAQLlKmDcpRCQUCmFsj4IECBAgAABAgQIECBAoGkBewiUoYCEShlOmiETIECAAAECBAgQINCxAnonQICAhIpzgAABAgQIECBAgEDlC4iQAAECBAosIKFSYFDNESBAgAABAgQIFEJAGwQIECBAIN8CEir5nh+jI0CAAAECBMpFwDgJECBAgACBqhKQUKmq6RYsAQIECBD4RsAaAQIECBAgQIBA2wUkVNpu50gCBAgQKK2A3ggQIECAAAECBAjkRkBCJTdTYSAECFSegIgIECBAgAABAgQIEKhUAQmVSp1ZcRFoi4BjCBAgQIAAAQIECBAgQKBFAhIqLWJSKa8CxkWAAAECBAgQIECAAAECBDpCQEKltOp6I0CAAAECBAgQIECAAAECBCpAoJmESgVEKAQCBAgQIECAAAECBAgQIECgGQG7WysgodJaMfUJECBAgAABAgQIECBAoOMFjIBABwtIqHTwBOi+8AIDzugShSyFH6EWCRAgQIAAAQIEqlFAzAQIVJaAhEplzadoCBAgQIAAAQIECBRKQDsECBAgMAcBCZU54NhFgAABAgQIECBQTgLGSoAAAQIESicgoVI6az0RIECAAAECBGYW8I4AAQIECBAoWwEJlbKdOgMnQIAAAQKlF9AjAQIECBAgQIDADAEJlRkOlgQIECBQmQKiIkCAAAECBAgQIFAUAQmVorBqlAABAm0VcBwBAgQIECBAgAABAuUgIKFSDrNkjATyLGBsBAgQIECAAAECBAgQqEIBCZUqnPRqD1n8BAgQIECAAAECBAgQIECgvQISKu0VLP7xeiBAgAABAgQIECBAgAABAgRyJlCEhErOIjQcAgQIECBAgAABAgQIECBAoAgC1d2khEp1z7/oCRAgQIAAAQIECBAgUD0CIiVQQAEJlQJiaooAAQIECBAgQIAAAQKFFNAWAQL5FZBQye/cGBkBAgQIECBAgACBchMwXgIECFSNgIRK1Uy1QAkQIECAAAECBGYXsIUAAQIECLRNQEKlbW6OIkCAAAECBAh0jIBeCRAgQIAAgVwISKjkYhoMggABAgQIVK6AyAgQIECAAAEClSggoVKJsyomAgQIEGiPgGMJECBAgAABAgQINCsgodIs0TcVzr98ZKy68b7x8aeff7PRGgECBDpcwAAIECBAgAABAgQIECi1gIRKC8Vvv/fxuPKGu1tYWzUCBOYoYCcBAgQIECBAgAABAgTKXEBCpQUTOPb5f8bZF14fgwcc2oLaqlSigJgIECBAgAABAgQIECBAgEBDAQmVhhqNrI8bPzH6nTg0hp5xeKy47JKN1MjlJoMiQIAAAQIECBAgQIAAAQIEiiiQk4RKESNsR9MfffxZHHzs4Dj64J3jh+t8t9GWes7VOaqtNArRjo2F9mvHUBxKgAABAgSqVsDX4/ZNfaH9UnvtG1H5HZ1iVqrvZwtz3rFz3jH/UlROrxIqc5jLp//yUox/85343xtvx2+H3RhX3jjjGSpDr7g1/vHyuOzIrl06RbWVKPCfQvsVeHiaI0CAAAECVSHQ/+SaKGSpCrQGQRbSrq6tBs1XxWqhvyfUXqeq+zmlaHNewT/zhT/tEpBQmQPfCsssEb888Ocx/3y9ondtmbdXz6x273nnjm5du2TrH302JaqtZIEXcFFovwIOTVMECBAgQIAAAQIlEij094TV3J7Yq+9ntLbOeYn+eldsNxIqc5ja5WsTKgfvuV3UlV22+3FWe99dt4q0L3tjQYAAAQIECBAgQIBAewQcS4AAgbIUkFApy2kzaAIECBAgQIAAgY4T0DMBAgQIEIiQUGnFWbDCskvES2NGRN2tP604VFUCBAgQIECAQMcJ6JkAAQIECBAouICESsFJNUiAAAECBAi0V8DxBAgQIECAAIG8C0io5H2GjI8AAQIEykHAGAkQIECAAAECBKpMQEKlyiZcuAQIEJghYEmAAAECBAgQIECAQHsEJFTao+dYAgRKJ6AnAgQIECBAgAABAgQI5EhAQiVHk2EolSUgGgIECBAgQIAAAQIECBCoXAEJlcqd29ZGpj4BAgQIECBAgAABAgQIECDQQoEyTqi0MELVCBAgQIAAAQIECBAgQIAAgTIWyOfQJVTyOS9GRYAAAQIECBAgQIAAAQLlKmDcVSEgoVIV0yxIAgQIECBAgAABAgQINC1gDwECrReQUGm9mSMIECBAgAABAgQIEOhYAb0TIECgwwUkVDp8CgyAAAECBAgQIECg8gVESIAAAQKVJiChUmkzKh4CBAgQIECAQCEEtEGAAAECBAjMUUBCZY48dhIgQIAAAQLlImCcBAgQIECAAIFSCkiolFJbXwQIECBA4BsBawQIECBAgAABAmUsIKFSxpNn6AQIECitgN4IECBAgAABAgQIEKgTkFCpk/BKgEDlCYiIAAECBAgQIECAAAECRRKQUCkSrGYJtEXAMQQIECBAgAABAgQIECBQHgISKuUxT3kdpXERIECAAAECBAgQIECAAIGqFKiyhEpVzrGgCRAgQIAAAQIECBAgQIBAlQkUP1wJleIb66EZgQFndIlClma6s5sAAQIECBAgQIAAAQL5EzCishOQUCm7KTNgAgQIECBAgAABAgQIdLyAERCodgEJlWo/A8RPgAABAgQIECBAoDoEREmAAIGCCkioFJRTYwQIECBAgAABAgQKJaAdAgQIEMizgIRKnmfH2AgQIECAAAEC5SRgrAQIECBAoIoEJFSqaLKFSoAAAQIECMws4B0BAgQIECBAoK0CEiptlXMcAQIECBAovYAeCRAgQIAAAQIEciIgoZKTiTAMAgQIVKaAqAgQIECAAAECBAhUpoCESmXOq6gIEGirgOMIECBAgAABAgQIECDQAgEJlRYgqUIgzwLGRoAAAWKg4rsAABAASURBVAIECBAgQIAAAQKlF5BQKb15tfcofgIECBAgQIAAAQIECBAgUPYCEirNTqEKBAgQIECAAAECBAgQIECAQOULtC5CCZXWealNgAABAgQIECBAgAABAgTyIWAUHSogodKh/DonQIAAAQIECBAgQIBA9QiIlEAlCUioVNJsioUAAQIECBAgQIAAgUIKaIsAAQJNCkioNEljBwECBAgQIECAAIFyEzBeAgQIECiVgIRKqaT1Q4AAAQIECBAgMLuALQQIECBAoEwFJFRaMHFTp02LN99+P76cPKUFtVUhQIAAAQIEKllAbAQIECBAgACBJCChkhTmUK64flSsvukBsdkux8RaWxwUxwwcFh99/NkcjrCLAAECBAjkSsBgCBAgQIAAAQIEiiAgodIMau/5esVVQ46L5+67PG6/+qwY+/w/4/Z7H2/mKLsJECBAoO0CjiRAgAABAgQIECCQfwEJlWbmaOdtN4711lolenTvFt9ebsnYeP0147GnX2jmKLsJEKgqAcESIECAAAECBAgQIFB1AhIqrZjyKVOnxRNj/xarrrRsK45SlUD+BIyIAAECBAgQIECAAAECBNonIKHSCr+zhl4Xn3w6KfbaaYv6o+bu3iWqrdQHX7oVPREgQIAAAQIECFS4QLV9Ty3e6vs5Ko9zXuH/rBQ9PAmVFhIPH3FH3Drq0bj6/ONjkYV61x/1q5MiZi9t39a5dkbyXuqDt0KAAAECBAgQIECgQAKF/h64kN+jp7YKPT7tRTDoeIMC/fWt2mZqf3yPiKoNv/nAp0//KgYNvymuufm+GHn5wFjtO8W93efjz6dG3kvzamoQIECAAAECBAgQaJ1Aob8Hbl3vzdcu9Pi0l/+feyp2jhr8zNn8ma/GnAQkVOakU7tvwKCrY8Qt98WQgYfFfPP2iglvvZuVqdOm1e71PwECBAgQIECAAAECBAgUU0DbBPIqIKHSzMyMff6fWY1Djj8vttitf32Z8Oa72XYLAgQIECBAgAABAgQINBCwSoBAlQhIqDQz0fffOCheGjNitrL0kn2aOdJuAgQIECBAgAABAuUgYIwECBAg0BYBCZW2qDmGAAECBAgQIECg4wT0TIAAAQIEciAgoZKDSTAEAgQIECBAoLIFREeAAAECBAhUnoCESuXNqYgIECBAgEB7BRxPgAABAgQIECDQjICESjNAdhMgQIBAOQgYIwECBAgQIECAAIHSCkiolNZbbwQIEJghYEmAAAECBAgQIECAQFkLSKiU9fQZPIHSCeiJAAECBAgQIECAAAECBL4RkFD5xsJaZQmIhgABAgQIECBAgAABAgQIFE1AQqVotK1tWH0CBAgQIECAAAECBAgQIECgXATanlAplwiNkwABAgQIECBAgAABAgQIEGi7gCMbFZBQaZTFRgIECBAgQIAAAQIECBAoVwHjJlAKAQmVUijrgwABAgQIECBAgAABAk0L2EOAQBkKSKiU4aQZMgECBAgQIECAAIGOFdA7AQIECEioOAcIECBAgAABAgQqX0CEBAgQIECgwAISKgUG1RwBAgQIECBAoBAC2iBAgAABAgTyLSChku/5MToCBAgQIFAuAsZJgAABAgQIEKgqAQmVqppuwRIgQIDANwLWCBAgQIAAAQIECLRdQEKl7XaOJECAQGkF9EaAAAECBAgQIECAQG4EJFRyMxUGQqDyBEREgAABAgQIECBAgACBShWQUKnUmRVXWwQcQ4AAAQIECBAgQIAAAQIEWiQgodIiprxWan5cA87oEoUuzfeqBgECBAgQIECAAIHWCeT9e9ZCj68a22vdGdF87UIbNt9jx9eoxpg7Xr3pEZQ2odL0OOwhQIAAAQIECBAgQIAAAQIEKkWgCuKQUKmCSRYiAQIECBAgQIAAAQIECMxZwF4CrRWQUGmtmPoECBAgQIAAAQIECBDoeAEjIECggwUkVDp4AnRPgAABAgQIECBAoDoEREmAAIHKEpBQqaz5FA0BAgQIECBAgEChBLRDgAABAgTmICChMgccuwgQIECAAAEC5SRgrAQIECBAgEDpBCRUSmetJwIECBAgQGBmAe8IECBAgAABAmUrIKFStlNn4AQIECBQegE9EiBAgAABAgQIEJghIKEyw8GSAAEClSkgKgIECBAgQIAAAQIEiiIgoVIUVo0SINBWAccRIECAAAECBAgQIECgHAQkVMphlowxzwLGRoAAAQIECBAgQIAAAQJVKCChUnWTLmACBAgQIECAAAECBAgQIECgvQL5T6i0N0LHEyBAgAABAgQIECBAgAABAvkXKLMRSqiU2YQZLgECBAgQIECAAAECBAjkQ8AoqltAQqW651/0BAgQIECAAAECBAhUj4BICRAooICESgExNUWAAAECBAgQIECAQCEFtEWAAIH8CkiotHBuPvn08/jgo09aWFs1AgQIECBAgACBqhQQNAECBAhUjYCESjNT/fmkL+KIky+I9bbtFz/a/ojo2+/MePf9j5o5ym4CBAgQIECAQHkIGCUBAgQIECDQNgEJlWbcbrh9dPz71fHxyK1D4+lRw6Nzp05xwZV/aOYouwkQIECAAIEiCWiWAAECBAgQIJALAQmVZqbhvkeejZ223SgWWah3zNOrZ+y10+Zx2z2PxVdffdXMkXYTIECAAIEkoBAgQIAAAQIECFSigIRKM7M6bvzEWGqJPvW1vrX4Itn6x59+nr1aECBAoOIEBESAAAECBAgQIECAQLMCEipzIEpXoaRnqHSfq1t9rbm6dc3WP//8i+y10IvFF+wRhSyFHp/2CORRwJgIECBAgAABAgTyL1DIn3NSW4WOOLWZ91LomLXXPgEJlTn41dTURM8e3ePLyVPqa9Wt9+zZPdt25QVdo5Ala7SAi0KOTVsFm+uCnjPmxbw4B5wDzgHngHPAOeAccA6UwzlQwB9zsqYKHXPWaM4X1RhznqdEQqWZ2Vl6yT7x+oSJ9bX+98bb2fq8vXpmr9WxECUBAgQIECBAgAABAgQIECDQUKAyEyoNI2zn+pYbrxMj7xoTb7/7YXz62aT43a0Pxo5bbxg1NTXtbNnhBAgQIECAAAECBAgQIECAQLsEOvBgCZVm8HffYbNYbunF48c7HRXrbnNoTJkyNY7Yf8dmjsrP7qnTpuVnMFU2kpbap3pvvv3+TLeWVRlVScJNz0RK1iXpTCctEpg+/auYNm16i+rWVfrs8y8i/X1Jx9Zt81o4geTamjl59/2P4osvJxduAFpqkcBHn3yW/aKnRZVVKqhAa+1bW7+gg62CxtK/WenfoeRcBeGWRYhTa3/2SF+n6x6T0NJBp5+xJrz1bkxu8KiFlh5bjvVKOea2zEl6juhHH39WymGWbV8SKs1M3dw9u8cl5xwdT941LB697YK4+bLTYpGFejdzVD52vz7h7Vh90wPijdp/nPIxouoZRUvtr7h+VDZHm+1yTKy1xUFxzMBh4R+v4pwnox58KrbYrX+rGj//8pGx6sb7hk/1ahVbiyqnBNfpQ0bEGedf26L6jz71Qmy95/Hx/a0PifT35T//ndCi41RquUBr5uSJsS9G335nxg77nxIb/OzIOPLUC7OrOFvem5ptEUg/OKa/B+tvd1j2i56f7nNS3PXAk21pyjGtFGitfWvrt3I4qtcKPPXcS9kvOzfa8ZeR/k7se9Q58eK/Xqvd0/z/6WtK+vqeXpuvrUZEtAihLd/Xvvb6m7HXEWfHGpsfmH2fdtu9j7eoL5VaJtDaOZn4zgdxxMkXxEY7HhWb7fqrbG7+8fK4lnVWpbUkVFo48fPNM3cstMB8Lazd8dXSN7pb7XFcxw+kCkfQGvve8/WKq4YcF8/dd3ncfvVZMfb5f8btvpAU9KxJz0Dasu+xccLZl7eq3TQPV95wd6uOUbllAvePeTY23OHIuHXUoy06YMyTz0e/E8+PLTZaJ0Zd95t44s6L41uLL9KiY1VqmUBr5iT9puvgYwfHhut9Lx67/cLaXzYMjf+8NiFuueuRlnWmVpsF0m/jf/aTH8XokUPi6VHD4yc//n5tUvK6mPSFq4TajNrCA1tr39r6LRyGag0EajrVxICj985+6fnwyPMj/RJ02DV3fF2j6Zd/vfK/6H/GJU1XsKfNAq39vjb98L7t3idGn4Xnj99ddFL2/fCWG6/T5v4dOLtAa+fkvEtvjnR10ZN/vDieGjUslvnWonHBlbfO3rAt9QISKvUUlbVywRlHxI3DT62soMokmtbY77ztxrHeWqtEj+7d4tvLLRkbr79mPPb0C2USaXkMc/FFF4prLzwxTv7lXi0ecEpsnX3h9TF4wKEtPkbFlgtssO7qMfKK02PbzX/Q7EHpqon0hXy7LdaPow7aKZZdarFI3xykvzPNHqxCiwVaMyeTJ0/N2l104QWy54mlT8NbY9UV4rXX38q2WxRPIF0he/Ce20Wyn6dXz/jplj+MdFn2P17+b/E61XIm0Fr71taPrBeL1gik75/S14b0S8/0A/lPNv5+9j3U1GnTmmzmnfc+jENPGBKnHbNPpH+7mqxoR5sEWvt97bW33BcL9J4nzjn54FhrtW9n3w/PP988berbQY0LtHZO3pj4Xiy8YO/o2rVLdOncuXZeVox/vzq+8cZtzQQ6ZUuLihNIX8j71H6zW3GBlUFAbbWfMnVaPDH2b7HqSsuWQZTlM8T0xSD98DH/fL1aNOhx4ydGvxOHxtAzDo8Vl12yRceo1DqBnj3myn4gnLtnj2YP/OCjT7Mv5J99NikOOf687NLT4SPuCM/taJauVRVaMyep7gF9t45Tzr0qBtf+Juvu0U/H48/8X+y+w6at6lPl9guMff6fWSPLfGux7LWUi2rvq7X2ra1f7b5tif+J516MlVdcOvshsLHj05Vch590Qey41YYtSug31oZtLRdoyfe1f3r2b7F4n4Wi/+mXxK6/OD0GDh4Rb73zfss7UbNVAi2Zk/132yruuO9P2a28jzz510i3DPXb52et6qfaKkuoVNuMize3AmcNvS4++XRS7LXTFrkdY6UPLD2/Jt3KcPTBO8cP1/lupYdbFvFN/PobqwXnny923Hqj2L72N/LX3HxfnHPRDWUx/kod5Pq1fz/SbxXTve/HnXlp9kNMuhosx/FW3NBefm18pCvpDt17++w3vBUXYI4Daq19a+vnOPTcDi09SyiVX/1il0bHmG7BOvmcK2OJxRaOfvv64bBRpAJvbMn3ta+MeyO7VWvTH60V+/fdKnsGzv5Hn5t9CEiBh6O5WoGWzMlKKywVSy/ZJzrVdIrjzrys9meTzyNdhVp7uP+bEJBQaQLGZgKlFEi/cb911KNx9fnHl81Dj0vpU6q+nv7LSzH+zXfif2+8Hb8ddmNceeOMZ6gMveLW+IcHcpVqGhrt58gDfx5bbLR27LTtRnHiEbtHuioi3Q4UFfsnv4GlT9M44JjfxuDT+sWws4+Ke68/N956+/04/byWPWA4v5GVz8jSJ2FrA+DLAAAQAElEQVT84rjzYpMfrRmH7rN9+Qy8AkbaWvvW1q8AopKHkB6SnZ6Tlm7j+cHaqzbaf3pI8P1jno15evWIwZfclH2NT7fLpWc/3T9mbKPH2Nh2gdZ8X7vHjptHunVry42/H4NOPSTSlcKvvv5m2zt3ZKMCLZ2TY04bFttuvn52pfbDI4fEOmusnD2Efk630jXaYRVtlFCposkWav4E0m9MBg2/KdJv3EdePjBW+86y+RtkFY1ohWWWiF/W/uCebg9Kz+mYt1fPLPre884d3bp2ydYtmhAo0ua6qx7G1ya56rqYOnVapG+Ev/qqbovXUgr8399fzbpbablvZa9LLdEnS3Sl236yDRZFFUgPAN7tkNNjg3W/F78+4cDo3Nm3ckUFb9B4a+1bW79BV1ZbKHB/bZIkXVl61vEHxC4//XGTR/Wau3v29X2JRRfKnsOVvsanyr3m7hHpNsa0rrRfoLXf16ZbtNKHB9T1PH369Gx18pSp2atF+wWmT/8qWvqzxmeff5FdJfSd5Wd8fU/P6jqg79bZ91zpitT2j6YyW/BVuDLnNdI9cpOnTMmiS+upZG8sii6QrJuyH3HLfdkzIOoGMWDQ1ZG2DRl4WMw3b69Iv8lKRRa4Tqj9r+kqhim1X5jTD+GptWx92rRI66kcM3B49hyItL58bUIlPfCxruyy3YxvzvbddatI+1Idpf0C06ZNzy7nnTZtWqR5SXOSvuDXtdxwTtLDBjdYd7W4+Jrbsy/or094O/5w92Ox+YZrR6dONXWHeG2nwJzm5JNPP4+dDjot7n34mayX5Zae8byO6297MJvHj2v3P/Doc7H26itl+y2KJ5A+nWT7/U6OH/y/VePA3beJie98kH3d+OCjT4rXqZYzgebsxz7/z+zvSfrtejqgufqpjtI+gTvvfyLS14sTDt89vr/mytnfhfQ9VEq4p5bT91fp43jTenoAbd3X9rrXtG3rTdbLkpOpjtJ+gea+r204J6m3rTddN66+6Z5s7tLVj7+79cHsFsb0C660X2m/QHNzkv4OpeehpZ7m7tk9llxs4UhXbqX5SN+f3TP66ewWoPShAKmOMruAhMrsJhWxZf3tDouf7D7jY5O33vP42GSnoyoirnIIYk7277z7YfzzP6/Xh5G+AUtv0sM2t9itf/b5++l1wpvvps0Ni/U2Crzy3zdijc0PjHQ5cPrhI62nh2nWNffa62/EG2/xrvMoxesf7n40m5N0m1t68Fmakzvue7y+61nn5JSj9o73P/wk1tnqkNhqj+Oy+61POarln9pU37CVJgXmNCcp2ZVuefvw40+z49NveIcM7BdP/fnvsf5PD49Ndz4m+2brtF/tk+23KJ7Aq+PeyBq/u/Yb3PQ1Pn29SOXcYTdm2y2KJ9Cc/aefT8puDa17YHZz9Ys30upp+YW/v5IFe87FN9R//5T+Ptz/9S08s37PlVW2KKrA2NrEYuqgqe9rZ52TPXfcPNZda5Vs/tL3z48980J2K2mP7t1SM0oBBJqbk1m/50ofytCtW9dI85G+xqfbr849+RdNPuy5AEMs+yYkVMp+ChsPYOy9l8ZLY0bUl8fvuKjxikXbWr0Nz8n+2H67Rdpfp3P/jYPq56jhfKWHQdXV8do+gRWWXWI243NOOri+0duvPiuGDDys/n3Dlbpj6279abjPetsF0mXZDc/3tL7j1hvWNzjrnKTfltx6xenxyK1DI/1blp41tNAC89XXt9J+gTnNyXzzzp39Her7s28+xSfd6/77i0+Oe35/TvzpzouyW0/Sp2m1fyRamJPAVpusm81F+jvTsDT8N21Ox9vXdoHm7H+8/prZ3Kz09aXyzdVv+0gcWScw4Oi9M/OGfxfS+g5bbZBVOXaW77myjQ0W6fuxjX6weoMtVtsr0Nz3tbPOSfrBffCAQ+OpUcPjoZvPi9G3DInvrbJ8e4fh+AYCzc3JrN9zpduwLjzzyHj2nkuzObnknKNjtZWXa9Ci1VkFJFTqRLwSIECAQK4FFlmod3YpcK4HWWWDW3jB3jFX7W+yqixs4RIgQIBAAQXSL64W67Ng1NS4lbeArO1qKt3+k36h0q5G8n5wgcYnoVIgSM0QIECAAAECBAgQIECAAIFiCGgznwISKvmcF6MiQIAAAQIECBAgQIBAuQoYN4GqEJBQqYppFiQBAgQIECBAgAABAk0L2EOAAIHWC0iotN7MEQQIECBAgAABAgQ6VkDvBAgQINDhAhIqHT4FBkCAAAECBAgQqHwBERIgQIAAgUoTkFCptBkVDwECBAgQIFAIAW0QIECAAAECBOYoIKEyRx47CRAgQIBAuQgYJwECBAgQIECAQCkFJFRKqa0vAgQIEPhGwBoBAgQIECBAgACBMhaQUCnjyTN0AgRKK6A3AgQIECBAgAABAgQI1AlIqNRJeCVQeQIiIkCAAAECBAgQIECAAIEiCUioFAlWs20RcAwBAgQIVIPAex98XA1hipEAAQIECBCocAEJlfZMsGMJECBAgACBFgv8+9Xx0bffmbHhDkfGTgedFi/+67UWH6siAQIECBAgQKBDBRrpXEKlERSbCBAgQIAAgcILXHrdnfHt5ZaMp0YNj3XW+E4Mu+aOwneiRQIECBAgQCATsCi+gIRK8Y31QIAAAQIECNQKvPSv/8Yaq64Q8/bqGWt+d8V47oV/1W71PwECBAgQyAQsCJSdgIRK2U2ZARMgQIBANQh88unn8fCf/lJfnhj7Yox/853ZQn99wttZnX+8PG62fRPf+SDb99cXX55p3/TpX8XTf/57XH/bQzHi5vvi7tFPx8uvjZ+pzqxvUlvpdp2W3Kbzt3++lt3ac9cDT87UzC4//XGcfeH18esLfhdnnn9t7L7DpjPt9yZi6rRpcczA4bVlWLbOhACBPAsYGwEC1S7QqdoBxE+AAAECBPIokJInR5xyYdSVg48dHFv2PTaO//VlMWXqtPohP/7M/2V10v4vJ0+p355Wzr9iZLbvoqtuS2+zMumLydH30DPigF/9NoZecWtcdePdcdyZl8bP9jslTh9ybVanscWXkyfH//39lfj000mN7a7flhICp583Ihaaf97YetP16renlQP6bh0rLLtE3HD76OjZo3scddBOaXNJy/Br74wNfnZESftsTWddOneOI/bfIe4fMzZu+eOY1hyqLoHmBdQgQIAAgYIKSKgUlFNjBAgQIECgsAKXD+ofL40ZEc/cfUkctt8OMerBp+Ke0U/P1sn7H36S7avbMeGtd2PWK0TSvuEj7sgeBnvx2b+Mp0YNi8fvuCieHjU8Tv7lXtG1S5dUpV3l5jsfjnS1zKlH7xOdO8/8bUYaY0rKLL1kn+xqm5Q0aldnbTj4q+nT23BUaQ9ZdqnF4qQj98yu5Hn73Q9L23nOejMcAgQIECCQZ4GZv9PJ80iNjQABAgQIVLFAr7l7xEF7bJsJ/Oe1Cdlrw8UeO24el//+rvrbRH7/hwfjuystGz9Ye9WG1eLZv/4jUkLjx+uvGelqiLRznl49s9tvTjpyj/Q2Kx99/FmcfM6Vsc5Wh2RlwKBrsu1zWqRbiYZe8YdsnIss1Hu2qg88OjbbduGZR2av94x+JnutW6RkzDEDh8XIUWNih/1PyfodNPymmDWpcP+YZ7P9q268b2y95/FxxMkXxJnnX1fXTPzxgSeyTxFKY09X9Zz0myvinfc+jHQ1z021CZ+U2Nnz8F9HKrf88ZFIV+2k9br9h500NIs9Nfju+x9lV/Bs8LMjsvGkK4TStrQvlTTmowZcHKndn+5zUlbnhLMvj48++SzS1TCp/012PjquvOHurJ90TCpNjTHtSyXdHpWu4vn9Hx5IbxUCBAgQIEAghwISKjmcFEMiQIAAAQKNCbw+YWK2efVVl89eGy722WXL7KqPB8Y8Fx989ElcN/L+OGSfn0anmpqG1WLTDdaKceMnxqXX/THenPjeTPvq3qTESLqF6I77/hRbb7puDDh671hi0YXqdjf5mq44+XzSF7XHrNdondvveTw2+eGa2W0/6fW2ex6Lr776qr5uuqrm/jFj45Jr74wtN/5+HLzntjHilvvihtsfqq/z4GPPRXrGSLeuXeOMY/ePvj/bNF7693/jX6/8L6vz1HMvxYlnXxGrrbxcnHPSwbHnzzeP51/6T2azeG0MK6+4dFZvm83Wi1RWXG7JmDp1aqTnzBxy/JAYds3tMVe3btGtW9fs1qq9jzw7Hnny+ei7w2ax325bxcN/+mukbXW3XaUxpzFdc9O9sd0W68e+tfNw1wNPxvrbHRb3PfxM7Lr9j2ObTX8Q518+Mp4Y+7es7zmNMatQu+japXPsvO1G8fdGno1Tu9v/BAgQIECAQA4EJFRyMAmGQIAAAQItFKjCai/UJgMe/tNfapMKo+Og/oOyq0t+8P9WnU0iJTy23/KHtYmSO+PGOx6O5ZdePDZab43Z6u249YZZUuWiq2+LzXb9VXZFxaEnnF/7w/6L9XXTD/7p4bPnnvyLOL3/flmi4Bd7bVe/v6mVV8bNuHLm27VJilnrvPLfCdmtRttt8cNs17abr58lOVKyI9vw9WKB3vPE3b87Nw7Z+6fZlS5777xl1F3Zkqqk58GkK2xuvuy0+Pk2G8ZeO20RDT3+8Z9xqVocuvf2WZxp/13X/ia7WieZrL7K8pH66FubiEklfdpQdkDtYtftN4mHR54fQwb2i9OO2ac2efKXLPk0eMCh0W+f7bMy6NRDsm2PPPHXqPuT2rtzxK+z8abbsjZYd7XM/w9XnhH777Z1/OqQXbL+04OF0zFzGmPaX1fWXG3FeOGlV+reeiVAgAABAgRyJiChkrMJMRwCBCpLQDQE2iswbMQd2YNl0yfjpGec/OHKM2Punt0bbTZdQfHKuDeyqyx+UZuQ6NRp5qtT0kELLTBfpFtu7rjmrEi3+Gyx0drx4j9fjXRFypDLbklV6q/2mPV2oWznHBb/fmV8lkhorMrdXz/3JSVD0pU2y3xr0azaXQ8+lb3WLdJtLj26d6t7G30Wnj9LYKQNkydPiRTfZhv8v/S20fLDdVbLtm+z1wlxyrlXxa2jHo30QN2uXZt/Psza31spGtZLfaXGUmIjvaZSt/6f2gRRep9KGnP3ub4Z80IL9I4e3eeaqa10C9SbE99N1aOlY1x04QUiXfEz6y1PWSMWBAgQIECAQIcLSKh0+BQYAIFcCRgMAQI5E6h7KO0xv9glu6Kj4e0vsw51xWWXzG6pSUmILTdeZ9bdM71PddNzV359woHx4M3nxTprfCe7siV9Ss+nn03K6s47z9zZa0sX6fadWR9Em45NbY68a0xajR0PODW22uP47DVtSLfHfPHl5LTaaEm3vtTtmDZ9xu1B3WuTFXXbZn1daflvxajrfhM7bLVB/PM/r8dpg6+JjXY8Kt6c+N6sVZt9nxI4qdJc3bqml6zUrU+b9s0nLWU7GiwaM6hpkNxq8Ri/vl0ruTZo3ioBAgQIvK+DuAAACXNJREFUECCQEwEJlZxMhGG0VcBxBAgQqA6BA/punSUJ0lUk6TkjTUV95nEHxA3DT61/4Oys9R596oWYNm3mT7pJV1f0nrdXdjVEeqTJ4n0WzA57+dUZzyXJ3rRg8e3aZMa/Xx0/W83nnv9XvP/hJ5Fupbnn9+dGXUnv0xUYaUyzHdTIhh7du2W366QH69YlGdLrx598Vl87xTbjU3L2iFuvOD1uv/qsLK6HHv9zVmeuubrFF19OydabW6SraVKd9MlE6TWV//v7q+klllxs4ey1LYvmxljX5sR33o909UtKkNVt80qAAAECBAjkR0BCpdRzoT8CBAgQINBGgfRw2LVW+3YcM3BYvPD3Vxptpfd8vSLdKhJN/Dnl3Ctjp4MGRPokndRGekBqemBqerDqbttvEumKkA1/sHp2dPpknwcefS5SAueYgcOzbXNarLDM4tnuultlsje1i7sefDJLhGy2wdrZM2BSoiKV9H6B3vNkn8pTW61F/x/Qd5sY+/w/sytc0rh3P+yseLjB80wuvOoPcfaFv4+UBHnjrXfjL3/7d9butxZfJHv9/hrfyRIst9/7eHZrU2MJoKxi7WLTDf5fltA4d9iNMebJ5yMlfn47/MZsW9pXW6VN/zc3xrpGUwyrfWfZurdeCRAgQIAAgZwJNJtQydl4DYcAAQIECFSVQKevb/tIQadPnrnwrCOyqyMO/NWgFt/G0qnzN1/uD9h9m+wKlYGDR8Tu/c6MA/sPyj7SNz3E9bjD+qZusoRM+oScf7w8Lo4+7eIsgZNuU8l2zv5YlmxzWiyx6IyrNu4d/Ux6m5VJX0yO9GlBP99mo5j1Vpj0Pm1PyYp0BUtNTWONz7wtPaT22H67RfqUn6f//Pf4wf9bJTZcb/XoNfeM58qstPxSce/Dz0Tf2tg2361/3HDbQ3HUQTvFxuuvkY1n1ZWWje22WD97vkq6/ej+Mc9GTc3MfWQVaxfz9uoZV553bLz/4cdx2ElDo9+J52frV9VuS/tqqzR6bE3UpF0zlU41naKmtqSNzY0x1Zk6bVqkpM93JVQSh0KAAAECJRLQTesEOrWuutoECBAgQIBAKQRWXnHpeGnMiJj1wbDzzzdP3H/joBh776WxWJ8FY48dN8vqNTWm9AyWKwcfW797311+En+89ux47r7L497rz41Hbh0aLz5yTRy+/w5R93yQVDklHf58/+XZ80j++sAV8esTDsz6WW+tVdLuRktKkKRkx7Uj74933/8oq9Oje7fsuJTUyDbMskjbU5zpSpWjD945i61hlVnjSw/aTTGkT/lJ5bB9d4hx49+KpZaY8ZDb9DHPj91+YaSSYkuxHrTHtvVNpjGmZFHye/S2C+Lw/XaoTcb0yMaYjq2v+PVK+lSg0bcMiYe+Lmn9e6ss//XeiMbGPLD/vpHGVl+pdmXoGYfHJeccXbsWkfpJ40ulsTGmSulhuu9/+EnsXju/6b1CgAABAo0K2EigQwUkVDqUX+cECBAgQKBjBFKiY6kl+kT69JmamtmvqEijSs9WSc8jSVfGpPctKXvsuHksudhCccb512ZXwrTkmNbUSbfwHHL8eTH82jtjxM33ZR8lPW78xNhlu43rm6mpqYkF5583i61+4ywr6dkk6ROPamoaj71h9ZqamlhskQWyUlPTfP2Gxza1XlPT9BjTpyCdd+ktccLhu8ecbt9qqm3bCRDIs4CxESBQSQISKpU0m2IhQIAAAQIdLJCewTKw/34x8e0P4u7RTxV8NOkKnXTrzKgHn4zb7nkses83T9x0yYBYYdklCt5XRzSYbvUZesWtsd5aK0ffHTbtiCHok8DMAt4RIECAQJMCEipN0thBgAABAgQItEUg3SaTbnn56RY/bMvhczwmXTGTbp1JnxSUbudJnxS02srLzfGYctrZpXPnGDLwsLjo179s8pOayimejhirPgkQIECAQKkEJFRKJa0fAgQIECBAgMDsArYQIECAAAECZSogoVKmE2fYBAgQIECgYwT0SoAAAQIECBAgkAQkVJKCQoAAAQKVKyAyAgQIECBAgAABAkUQkFApAqomCRAg0B4BxxIgQIAAAQIECBAgkH8BCZX8z5EREsi7gPERIECAAAECBAgQIECg6gQkVKpuygUcwYAAAQIECBAgQIAAAQIECLRPQEKlfX6lOVovBAgQIECAAAECBAgQIECAQK4EipJQyVWEBkOAAAECBAgQIECAAAECBAgURaCaG5VQqebZFzsBAgQIECBAgAABAgSqS0C0BAomIKFSMEoNESBAgAABAgQIECBAoNAC2iNAIK8CEip5nRnjIkCAAAECBAgQIFCOAsZMgACBKhGQUKmSiRYmAQIECBAgQIBA4wK2EiBAgACBtghIqLRFzTEECBAgQIAAgY4T0DMBAgQIECCQAwEJlRxMgiEQIECAAIHKFhAdAQIECBAgQKDyBCRUKm9ORUSAAAEC7RVwPAECBAgQIECAAIFmBCRUmgGymwABAuUgYIwECBAgQIAAAQIECJRWQEKltN56I0BghoAlAQIECBAgQIAAAQIEylpAQqWsp8/gSyegJwIECBAgQIAAAQIECBAg8I2AhMo3FpW1JhoCBAgQIECAAAECBAgQIECgaAK5SagULUINEyBAgAABAgQIECBAgAABArkRqJSBSKhUykyKgwABAgQIECBAgAABAgSKIaBNAo0KSKg0ymIjAQIECBAgQIAAAQIEylXAuAkQKIWAhEoplPVBgAABAgQIECBAgEDTAvYQIECgDAUkVMpw0gyZAAECBAgQIECgYwX0ToAAAQIEJFScAwQIECBAgACByhcQIQECBAgQIFBgAQmVAoNqjgABAgQIECiEgDYIECBAgAABAvkWkFDJ9/wYHQECBAiUi4BxEiBAgAABAgQIVJWAhEpVTbdgCRAg8I2ANQIECBAgQIAAAQIE2i4godJ2O0cSIFBaAb0RIECAAAECBAgQIEAgNwISKrmZCgOpPAERESBAgAABAgQIECBAgEClCkioVOrMtiUuxxAgQIAAAQIECBAgQIAAAQItEijrhEqLIlSJAAECBAgQIECAAAECBAgQKGuBPA5eQiWPs2JMBAgQIECAAAECBAgQIFDOAsZeBQISKlUwyUIkQIAAAQIECBAgQIDAnAXsJUCgtQISKq0VU58AAQIECBAgQIAAgY4XMAICBAh0sICESgdPgO4JECBAgAABAgSqQ0CUBAgQIFBZAv8fAAD//6MvlZYAAAAGSURBVAMAyquqAC5Xg+4AAAAASUVORK5CYII="
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(concoord_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_convert](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_convert) from **biobb_analysis.ambertools.cpptraj_convert**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026-02-12 09:35:59,673 [MainThread  ] [INFO ]  Module: biobb_analysis.ambertools.cpptraj_convert Version: 5.2.0\n",
      "2026-02-12 09:35:59,676 [MainThread  ] [INFO ]  Directory successfully created: /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7\n",
      "2026-02-12 09:35:59,679 [MainThread  ] [INFO ]  Copy to stage: 1ake_dist.pdb --> sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7\n",
      "2026-02-12 09:35:59,682 [MainThread  ] [INFO ]  Copy to stage: 1ake_disco_traj.pdb --> sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7\n",
      "2026-02-12 09:35:59,700 [MainThread  ] [INFO ]  Launching command (it may take a while): cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/ea7cb7eb-5737-407b-83cd-8b6d4cd1b053instructions.in\n",
      "2026-02-12 09:36:00,019 [MainThread  ] [INFO ]  Command 'cpptraj -i /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/ea7cb7eb-5737-407b-83cd-8b6d4cd1b053instructions.in' finalized with exit code 0\n",
      "2026-02-12 09:36:00,021 [MainThread  ] [INFO ]  \n",
      "CPPTRAJ: Trajectory Analysis. V6.18.1 (AmberTools)\n",
      "    ___  ___  ___  ___\n",
      "     | \\/ | \\/ | \\/ | \n",
      "    _|_/\\_|_/\\_|_/\\_|_\n",
      "\n",
      "| Date/time: 02/12/26 09:35:59\n",
      "| Available memory: 3.223 GB\n",
      "\n",
      "INPUT: Reading input from '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/ea7cb7eb-5737-407b-83cd-8b6d4cd1b053instructions.in'\n",
      "  [parm /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_dist.pdb]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_dist.pdb' as PDB File\n",
      "\tReading bond info from CONECT records.\n",
      "\tNot reading bond info from LINK records.\n",
      "\tDetermining bond info from distances.\n",
      "Warning: Determining bond length parameters from element types for '1ake_dist.pdb'.\n",
      "Warning: 1 molecules have non-contiguous segments of atoms.\n",
      "\t       1        2 segments: 1-1353 (1353)  1356-1433 (78) \n",
      "Warning: The 'fixatomorder' command can be used to reorder the topology and any\n",
      "Warning:  associated coordinates.\n",
      "  [trajin /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_disco_traj.pdb 1 -1 1]\n",
      "\tReading '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_disco_traj.pdb' as PDB\n",
      "  [strip !@CA]\n",
      "    STRIP: Stripping atoms in mask [!@CA]\n",
      "  [trajout /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_disco_traj.trr trr]\n",
      "\tWriting '/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_disco_traj.trr' as Gromacs TRX\n",
      "---------- RUN BEGIN -------------------------------------------------\n",
      "\n",
      "PARAMETER FILES (1 total):\n",
      " 0: 1ake_dist.pdb, 1656 atoms, 214 res, box: None, 4 mol\n",
      "\n",
      "INPUT TRAJECTORIES (1 total):\n",
      " 0: '1ake_disco_traj.pdb' is a PDB file, Parm 1ake_dist.pdb (reading 300 of 300)\n",
      "  Coordinate processing will occur on 300 frames.\n",
      "\n",
      "OUTPUT TRAJECTORIES (1 total):\n",
      "  '1ake_disco_traj.trr' (300 frames) is a GROMACS TRR file, big-endian, single precision\n",
      "\n",
      "BEGIN TRAJECTORY PROCESSING:\n",
      ".....................................................\n",
      "ACTION SETUP FOR PARM '1ake_dist.pdb' (1 actions):\n",
      "  0: [strip !@CA]\n",
      "\tStripping 1442 atoms.\n",
      "\tStripped topology: 214 atoms, 214 res, box: None, 214 mol\n",
      ".....................................................\n",
      "ACTIVE OUTPUT TRAJECTORIES (1):\n",
      "  /Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7/1ake_disco_traj.trr (coordinates)\n",
      "----- 1ake_disco_traj.pdb (1-300, 1) -----\n",
      " 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Complete.\n",
      "\n",
      "Read 300 frames and processed 300 frames.\n",
      "TIME: Avg. throughput= 2926.3154 frames / second.\n",
      "\n",
      "ACTION OUTPUT:\n",
      "TIME: Analyses took 0.0000 seconds.\n",
      "\n",
      "RUN TIMING:\n",
      "TIME:\t\tInit               : 0.0000 s (  0.02%)\n",
      "TIME:\t\tTrajectory Process : 0.1025 s ( 99.82%)\n",
      "TIME:\t\tAction Post        : 0.0000 s (  0.00%)\n",
      "TIME:\t\tAnalysis           : 0.0000 s (  0.00%)\n",
      "TIME:\t\tData File Write    : 0.0000 s (  0.00%)\n",
      "TIME:\t\tOther              : 0.0002 s (  0.00%)\n",
      "TIME:\tRun Total 0.1027 s\n",
      "---------- RUN END ---------------------------------------------------\n",
      "TIME: Total execution time: 0.2524 seconds.\n",
      "--------------------------------------------------------------------------------\n",
      "To cite CPPTRAJ use:\n",
      "Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for\n",
      "  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.\n",
      "  Theory Comput., 2013, 9 (7), pp 3084-3095.\n",
      "\n",
      "\n",
      "2026-02-12 09:36:00,023 [MainThread  ] [INFO ]  Removed: ['/Users/gbayarri/projects/BioBB/tutorials/biobb_wf_flexdyn/biobb_wf_flexdyn/notebooks/sandbox_3a85cf23-c452-4c2e-84a9-5e9a748cbcf7']\n",
      "2026-02-12 09:36:00,024 [MainThread  ] [INFO ]  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_convert import cpptraj_convert\n",
    "\n",
    "concoord_trr = pdbCode + \"_disco_traj.trr\"\n",
    "\n",
    "prop = {\n",
    "    'mask' : 'c-alpha',\n",
    "    'format': 'trr'\n",
    "}\n",
    "\n",
    "cpptraj_convert(input_top_path=concoord_dist_pdb,\n",
    "                input_traj_path=concoord_disco_pdb,\n",
    "                output_cpptraj_path=concoord_trr,\n",
    "                properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e757e199ce2341cea787d0caf2eb3335",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget(max_frame=299)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(concoord_trr, prot_ca), gui=True)\n",
    "view.center()\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"prody\"></a>\n",
    "***\n",
    "## PRODY\n",
    "\n",
    "**ProDy** is a free and open-source Python package for **protein structural dynamics analysis**. It is designed as a flexible and responsive API suitable for **interactive usage** and **application development**.\n",
    "\n",
    "* Structure analysis: **ProDy** has fast and flexible **PDB** and **DCD** file parsers, and powerful and customizable **atom selections** for contact identification, structure comparisons, and rapid implementation of new methods.\n",
    "\n",
    "\n",
    "* Dynamics analysis: \n",
    "    - **Principal component analysis** can be performed for: \n",
    "        - heterogeneous X-ray structures (missing residues, mutations)\n",
    "        - mixed structural datasets from Blast search\n",
    "        - NMR models and MD snapshots (essential dynamics analysis)\n",
    "        \n",
    "    * **Normal mode analysis** can be performed using: \n",
    "        - Anisotropic network model (ANM)\n",
    "        - Gaussian network model (GNM)\n",
    "        - ANM/GNM with distance and property dependent force constants\n",
    "\n",
    "\n",
    "References: <br>\n",
    "\n",
    "**PRODY website**: http://prody.csb.pitt.edu/\n",
    "\n",
    "**ProDy 2.0: Increased scale and scope after 10 years of protein dynamics modelling with Python.**<br>\n",
    "*Bioinformatics, 2021, 37(20):3657-3659.*<br>\n",
    "*Available at: https://doi.org/10.1093/bioinformatics/btab187*\n",
    "\n",
    "**ProDy: Protein Dynamics Inferred from Theory and Experiments.**<br>\n",
    "*Bioinformatics, 2011, 27(11):1575-1577.*<br>\n",
    "*Available at: https://doi.org/10.1093/bioinformatics/btr168*\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [prody_anm](https://biobb-flexdyn.readthedocs.io/en/latest/flexdyn.html#module-prody.prody_anm) from **biobb_flexdyn.flexdyn.prody_anm**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pkg_resources'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[32]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mbiobb_flexdyn\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mflexdyn\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mprody_anm\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m prody_anm\n\u001b[32m      3\u001b[39m prody_ensemble = pdbCode + \u001b[33m\"\u001b[39m\u001b[33m_prody_anm_traj.pdb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      5\u001b[39m prop = {\n\u001b[32m      6\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mselection\u001b[39m\u001b[33m'\u001b[39m : \u001b[33m'\u001b[39m\u001b[33mbackbone\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m      7\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mnum_structs\u001b[39m\u001b[33m'\u001b[39m : num_frames,\n\u001b[32m      8\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mrmsd\u001b[39m\u001b[33m'\u001b[39m : \u001b[32m2.0\u001b[39m\n\u001b[32m      9\u001b[39m }\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/biobb_flexdyn/__init__.py:1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m flexdyn\n\u001b[32m      2\u001b[39m name = \u001b[33m\"\u001b[39m\u001b[33mbiobb_flexdyn\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      3\u001b[39m __all__ = [\u001b[33m\"\u001b[39m\u001b[33mflexdyn\u001b[39m\u001b[33m\"\u001b[39m]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/biobb_flexdyn/flexdyn/__init__.py:3\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m concoord_dist\n\u001b[32m      2\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m concoord_disco\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m prody_anm\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m nolb_nma\n\u001b[32m      5\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m imod_imode\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/biobb_flexdyn/flexdyn/prody_anm.py:5\u001b[39m\n\u001b[32m      3\u001b[39m \u001b[33;03m\"\"\"Module containing the prody_anm class and the command line interface.\"\"\"\u001b[39;00m\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mtyping\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Optional\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mprody\u001b[39;00m  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[32m      6\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mbiobb_common\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mgeneric\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mbiobb_object\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m BiobbObject\n\u001b[32m      7\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mbiobb_common\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mtools\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mfile_utils\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m launchlogger\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/__init__.py:81\u001b[39m\n\u001b[32m     78\u001b[39m __all__.extend(kdtree.__all__)\n\u001b[32m     79\u001b[39m __all__.append(\u001b[33m'\u001b[39m\u001b[33mkdtree\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m81\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m atomic\n\u001b[32m     82\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01matomic\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m *\n\u001b[32m     83\u001b[39m __all__.extend(atomic.__all__)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/atomic/__init__.py:148\u001b[39m\n\u001b[32m    140\u001b[39m __all__ = [\u001b[33m'\u001b[39m\u001b[33mAtomic\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mAtomGroup\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m    141\u001b[39m            \u001b[33m'\u001b[39m\u001b[33mHierView\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mSegment\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mChain\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mResidue\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mAtom\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m    142\u001b[39m            \u001b[33m'\u001b[39m\u001b[33mAtomPointer\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mAtomSubset\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m    143\u001b[39m            \u001b[33m'\u001b[39m\u001b[33mSelection\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mAtomMap\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m    144\u001b[39m            \u001b[33m'\u001b[39m\u001b[33mBond\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mselect\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33matomgroup\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mhierview\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mchain\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mfields\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mflags\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m    146\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mfields\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m ATOMIC_FIELDS\n\u001b[32m--> \u001b[39m\u001b[32m148\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01matom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m *\n\u001b[32m    149\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mbond\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m *\n\u001b[32m    150\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mflags\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m *\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/atomic/atom.py:9\u001b[39m\n\u001b[32m      7\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mfields\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m ATOMIC_FIELDS, READONLY\n\u001b[32m      8\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mfields\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m wrapGetMethod, wrapSetMethod\n\u001b[32m----> \u001b[39m\u001b[32m9\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mpointer\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m AtomPointer\n\u001b[32m     10\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mbond\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Bond\n\u001b[32m     12\u001b[39m __all__ = [\u001b[33m'\u001b[39m\u001b[33mAtom\u001b[39m\u001b[33m'\u001b[39m]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/atomic/pointer.py:7\u001b[39m\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mnumbers\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Integral\n\u001b[32m      5\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mnumpy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;28mall\u001b[39m, array, concatenate, ones, unique, \u001b[38;5;28many\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m7\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01matomic\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Atomic\n\u001b[32m      8\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mbond\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Bond\n\u001b[32m      9\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mangle\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Angle\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/atomic/atomic.py:19\u001b[39m\n\u001b[32m     16\u001b[39m NOTALLNONE = \u001b[38;5;28mset\u001b[39m([\u001b[33m'\u001b[39m\u001b[33mnot\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mall\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mnone\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mindex\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33msequence\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mx\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33my\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mz\u001b[39m\u001b[33m'\u001b[39m])\n\u001b[32m     18\u001b[39m MODMAP = {}\n\u001b[32m---> \u001b[39m\u001b[32m19\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mopenData\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mmod_res_map.dat\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[32m     20\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m f:\n\u001b[32m     21\u001b[39m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/utilities/misctools.py:428\u001b[39m, in \u001b[36mopenData\u001b[39m\u001b[34m(filename, mode)\u001b[39m\n\u001b[32m    427\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mopenData\u001b[39m(filename, mode=\u001b[33m'\u001b[39m\u001b[33mr\u001b[39m\u001b[33m'\u001b[39m):\n\u001b[32m--> \u001b[39m\u001b[32m428\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(\u001b[43mgetDataPath\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m, mode)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/biobb_wf_flexdyn/lib/python3.11/site-packages/prody/utilities/misctools.py:424\u001b[39m, in \u001b[36mgetDataPath\u001b[39m\u001b[34m(filename)\u001b[39m\n\u001b[32m    423\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mgetDataPath\u001b[39m(filename):\n\u001b[32m--> \u001b[39m\u001b[32m424\u001b[39m     \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpkg_resources\u001b[39;00m\n\u001b[32m    425\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m pkg_resources.resource_filename(\u001b[33m'\u001b[39m\u001b[33mprody.utilities\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mdatafiles/\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[33m'\u001b[39m%filename)\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'pkg_resources'"
     ]
    }
   ],
   "source": [
    "from biobb_flexdyn.flexdyn.prody_anm import prody_anm\n",
    "\n",
    "prody_ensemble = pdbCode + \"_prody_anm_traj.pdb\"\n",
    "\n",
    "prop = {\n",
    "    'selection' : 'backbone',\n",
    "    'num_structs' : num_frames,\n",
    "    'rmsd' : 2.0\n",
    "}\n",
    "\n",
    "prody_anm(  input_pdb_path=monomer,\n",
    "            output_pdb_path=prody_ensemble,\n",
    "            properties=prop)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "prody_rmsd = pdbCode + \"_prody_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "cpptraj_rms(input_top_path=prody_ensemble,\n",
    "            input_traj_path=prody_ensemble,\n",
    "            output_cpptraj_path=prody_rmsd,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(prody_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_convert](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_convert) from **biobb_analysis.ambertools.cpptraj_convert**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_convert import cpptraj_convert\n",
    "\n",
    "prody_trr = pdbCode + \"_prody_anm_traj.trr\"\n",
    "\n",
    "prop = {\n",
    "    'mask' : 'c-alpha',\n",
    "    'format': 'trr'\n",
    "}\n",
    "\n",
    "cpptraj_convert(input_top_path=prot_backbone,\n",
    "                input_traj_path=prody_ensemble,\n",
    "                output_cpptraj_path=prody_trr,\n",
    "                properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(prody_trr, prot_ca), gui=True)\n",
    "view.center()\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"flexserv\"></a>\n",
    "***\n",
    "## FLEXSERV\n",
    "\n",
    "Despite recent advances in **experimental techniques**, the study of **molecular flexibility** is mostly a task for **theoretical methods**. The most powerful of them is **atomistic molecular dynamics** (MD), a rigorous method with solid physical foundations, which provides accurate representations of **protein flexibility** under **physiological-like environments**. Unfortunately, despite its power, MD is a **complex technique**, **computationally expensive** and their use requires a certain **degree of expertise**. The alternative to **atomistic MD** is the use of **coarse-grained** methods coupled to **simple potentials**. By using these techniques we assume a **lost of atomic detail** to gain formal and **computational simplicity** in the representation of near-native state **flexibility** of proteins. Unfortunately, despite its power, the practical use of coarse-grained methods is still limited, due mostly to the lack of **standardized protocols** for analysis and the existence of a myriad of different algorithms distributed in different websites.\n",
    "\n",
    "**FlexServ** webserver and associated **BioBB module** integrate three **coarse-grained** algorithms for the representation of **protein flexibility**: \n",
    "\n",
    "i) **Brownian dynamics** (BD) <br>\n",
    "ii) **Discrete dynamics** (DMD)<br>\n",
    "iii) **Normal mode analysis** (NMA) based on different types of elastic networks <br>\n",
    "\n",
    "This next cells of this tutorial show how to extract **molecular flexibility** (conformational ensemble) from a single, static structure **downloaded** from the **PDB database**, generating its **coarse-grained**, reduced representation ($C_{\\alpha}$) and running the previously mentioned **coarse-grained algorithms**.\n",
    "\n",
    "References: <br>\n",
    "\n",
    "**FLEXSERV website**: https://mmb.irbbarcelona.org/FlexServ/\n",
    "\n",
    "**FlexServ: an integrated tool for the analysis of protein flexibility.**<br>\n",
    "*Bioinformatics, Volume 25, Issue 13, 2009, Pages 1709–1710,.*<br>\n",
    "*Available at: https://doi.org/10.1093/bioinformatics/btp304*\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [bd_run](https://biobb-flexserv.readthedocs.io/en/latest/flexserv.html#module-flexserv.bd_run) from **biobb_flexserv.flexserv.bd_run**\n",
    " - [dmd_run](https://biobb-flexserv.readthedocs.io/en/latest/flexserv.html#module-flexserv.dmd_run) from **biobb_flexserv.flexserv.dmd_run**\n",
    " - [nma_run](https://biobb-flexserv.readthedocs.io/en/latest/flexserv.html#module-flexserv.nma_run) from **biobb_flexserv.flexserv.nma_run**\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"bd\"></a>\n",
    "***\n",
    "## Brownian Dynamics (BD)\n",
    "\n",
    "The **Brownian Dynamics** (BD) method introduces the protein in an **stochastic bath** that keeps the **temperature constant** and modulates the otherwise extreme oscillations of the residues. This bath is simulated with two terms accounting for a **velocity-dependent friction** and **stochastic forces** due to the **solvent environment**.\n",
    "**Velocity Verlet** algorithm is used to solve the **stochastic differential equation** (equation of motion) for **alpha-carbons** ($C\\alpha$). The **equation of motion** is integrated using **Verlet’s algorithm**\n",
    "The **potential energy** used to compute **forces** in the **equation of motion** assumes a **coarse-grained representation of the protein** ($C\\alpha$-only) and a **quasi-harmonic representation** of the **interactions** (similar to that suggested by [Kovacs et al. 2004](https://doi.org/10.1002/prot.20151)).\n",
    "\n",
    "The initial condition is a **native structure** (or MD-averaged conformation) that is supposed to be in the **minimal energy state**, from which the **relative vectors** are computed. Brownian Dynamics (BD) **simulation time scales** were equivalent to those considered in **Molecular Dynamics** (MD).\n",
    "\n",
    "Reference: <br>\n",
    "**Exploring the Suitability of Coarse-Grained Techniques for the Representation of Protein Dynamics.**<br>\n",
    "*Biophysical Journal, Volume 95, Issue 5, 1 September 2008, Pages 2127-2138*<br>\n",
    "*Available at: https://doi.org/10.1529/biophysj.107.119115*\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running Brownian Dynamics (BD)\n",
    "# Import module\n",
    "from biobb_flexserv.flexserv.bd_run import bd_run\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "\n",
    "bd_log = pdbCode + '_flexserv_bd_ensemble.log'\n",
    "bd_crd = pdbCode + '_flexserv_bd_ensemble.mdcrd'\n",
    "\n",
    "wfreq = 100\n",
    "time = num_frames * wfreq\n",
    "\n",
    "prop = {\n",
    "    'time': time,\n",
    "    'wfreq': wfreq\n",
    "}\n",
    "\n",
    "bd_run( \n",
    "     input_pdb_path=prot_ca,\n",
    "     output_crd_path=bd_crd,\n",
    "     output_log_path=bd_log,\n",
    "     properties=prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "flexserv_bd_rmsd = pdbCode + \"_flexserv_bd_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=bd_crd,\n",
    "            output_cpptraj_path=flexserv_bd_rmsd,\n",
    "            input_exp_path=monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(flexserv_bd_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Fitting and converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "flexserv_bd_rmsd = pdbCode + \"_flexserv_bd_rmsd.dat\" \n",
    "flexserv_bd_traj_fitted = pdbCode + \"_flexserv_bd_traj_fitted.trr\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=bd_crd,\n",
    "            output_cpptraj_path=flexserv_bd_rmsd,\n",
    "            output_traj_path=flexserv_bd_traj_fitted,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(flexserv_bd_traj_fitted, prot_ca), gui=True)\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"dmd\"></a>\n",
    "***\n",
    "## Discrete Molecular Dynamics (DMD)\n",
    "\n",
    "With the **Discrete Molecular Dynamics** (DMD) method, the proteins are modelled as a system of beads ($C\\alpha$ atoms) interacting through a **discontinuous potential** (square wells in the used tool). Outside the discontinuities, **potentials** are considered constant, thereby implying a **ballistic regime** for the particles (**constant potential, constant velocity**) in all conditions, except at such time as when the particles reach a **potential discontinuity** (this is called “an **event**” or “a **collision**”). At this time, the **velocities** of the colliding particles are modified by imposing **conservation of the linear momentum**, **angular momentum**, and **total energy**. Since the particles were constrained to move within a configurational space where the **potential energy** is constant (**infinite square wells**), the **kinetic energy** remains unchanged and therefore all collisions are assumed to be **elastic**.\n",
    "\n",
    "**DMD** has a major advantage over techniques like **MD** because, as it does not require the **integration of the equations of motion** at fixed **time steps**, the calculation progresses **from event to event**. In practice, the time between events decreases with **temperature** and **density** and depends on the **number of particles\n",
    "N**. The **equation of motion**, corresponding to **constant velocity**, is solved analytically.\n",
    "\n",
    "As the **integration of Newton’s equations** is no longer the rate limiting step, calculations can be extended for very **long simulation** periods and **large systems**, provided an efficient algorithm for **predicting collisions** is used.\n",
    "\n",
    "The **collision between particles** i and j is associated with a **transfer of linear momentum** in the direction of the vector $\\vec{r_{ij}}$. To calculate the **change in velocities**, the **velocity** of each **particle** is projected in the **direction of the vector** $\\vec{r_{ij}}$ so that the **conservation equations** become one-dimensional along the **interatomic coordinate**. \n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running Discrete Molecular Dynamics (DMD)\n",
    "# Import module\n",
    "from biobb_flexserv.flexserv.dmd_run import dmd_run\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "\n",
    "dmd_log = pdbCode + '_flexserv_dmd_ensemble.log'\n",
    "dmd_crd = pdbCode + '_flexserv_dmd_ensemble.mdcrd'\n",
    "\n",
    "prop = {\n",
    "    'frames': num_frames\n",
    "}\n",
    "dmd_run( \n",
    "     input_pdb_path=prot_ca,\n",
    "     output_crd_path=dmd_crd,\n",
    "     output_log_path=dmd_log,\n",
    "     properties=prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "flexserv_dmd_rmsd = pdbCode + \"_flexserv_dmd_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=dmd_crd,\n",
    "            output_cpptraj_path=flexserv_dmd_rmsd,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(flexserv_dmd_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Fitting and converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "flexserv_dmd_rmsd = pdbCode + \"_flexserv_dmd_rmsd.dat\"\n",
    "flexserv_dmd_traj_fitted = pdbCode + \"_flexserv_dmd_traj_fitted.trr\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=dmd_crd,\n",
    "            output_cpptraj_path=flexserv_dmd_rmsd,\n",
    "            output_traj_path=flexserv_dmd_traj_fitted,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(flexserv_dmd_traj_fitted, prot_ca), gui=True)\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"nma\"></a>\n",
    "***\n",
    "## Normal Mode Analysis (NMA)\n",
    "\n",
    "**Normal Mode Analysis** (NMA) can be defined as the **multidimensional** treatment of **coupled oscillators** from the analysis of **force-derivatives** in **equilibrium conformations**. This methodology assumes that the **displacement** of an **atom** from its **equilibrium position** is small and that the **potential energy** in the vicinity of the **equilibrium position** can be approximated as a **sum of terms** that are **quadratic** in the **atomic displacements**. In its purest form, it uses the same **all-atom force field** from a **MD simulation**, implying a prior **in vacuo minimization** ([Go and Scheraga 1976](https://pubs.acs.org/cgi-bin/abstract.cgi/mamobx/1976/9/i04/f-pdf/f_ma60052a001.pdf?sessid=6006l3); [Brooks III, Karplus et al. 1987](https://adsabs.harvard.edu/abs/1990PhT....43b.120B)). \n",
    "\n",
    "[Tirion (1996)](https://link.aps.org/doi/10.1103/PhysRevLett.77.1905) proposed a **simplified model** where the **interaction** between two atoms was described by **Hookean pairwise potential** where the distances are taken to be at the minimum, avoiding the minimization (referred as **Elastic Network Model** -ENM-). This idea being further extended to use **coarse-grained** ($C\\alpha$) protein representation by several research groups, as in the **Gaussian Network Model** -GNM- ([Bahar et al. 1997](https://linkinghub.elsevier.com/retrieve/pii/S1359027897000242); [Haliloglu et al. 1997](https://link.aps.org/doi/10.1103/PhysRevLett.79.3090)). The **GNM model** was later extended to a 3-D, vectorial **Anisotropic Network Model** -ANM-, which is the formalism implemented in the following cell ([Atilgan et al. 2001](https://www.biophysj.org/cgi/content/abstract/80/1/505)). Through the **diagonalization** of the **hessian matrix**, the **ANM** provides **eigenvalues** and **eigenvectors** that not only describe the **frequencies** and **shapes** of the **normal modes**, but also their **directions**.\n",
    "\n",
    "Once **NMA** is performed and the set of **eigenvectors/eigenvalues** is determined, ***Cartesian pseudo-trajectories*** at **physiologic temperature** can be obtained by activating **normal mode deformations** using a **Metropolis Monte Carlo algorithm**. The **displacements** obtained by such algorithm can then be projected to the **Cartesian space** to generate the ***pseudo-trajectories***.\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running Normal Mode Analysis (NMA)\n",
    "# Import module\n",
    "from biobb_flexserv.flexserv.nma_run import nma_run\n",
    "\n",
    "# uncomment if executing in google colab\n",
    "# %env CONDA_PREFIX=/usr/local\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "\n",
    "nma_log = pdbCode + '_flexserv_nma_ensemble.log'\n",
    "nma_crd = pdbCode + '_flexserv_nma_ensemble.mdcrd'\n",
    "\n",
    "prop = {\n",
    "    'frames' : num_frames\n",
    "}\n",
    "\n",
    "nma_run( \n",
    "     input_pdb_path=prot_ca,\n",
    "     output_crd_path=nma_crd,\n",
    "     output_log_path=nma_log,\n",
    "     properties=prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "flexserv_nma_rmsd = pdbCode + \"_flexserv_nma_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=nma_crd,\n",
    "            output_cpptraj_path=flexserv_nma_rmsd,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(flexserv_nma_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_convert](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_convert) from **biobb_analysis.ambertools.cpptraj_convert**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_convert import cpptraj_convert\n",
    "\n",
    "nma_trr = pdbCode + '_flexserv_nma_ensemble.trr'\n",
    "\n",
    "prop = {\n",
    "    'mask' : 'c-alpha',\n",
    "    'format': 'trr'\n",
    "}\n",
    "\n",
    "cpptraj_convert(input_top_path=prot_ca,\n",
    "                input_traj_path=nma_crd,\n",
    "                output_cpptraj_path=nma_trr,\n",
    "                properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(nma_trr, prot_ca), gui=True)\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"nolb\"></a>\n",
    "***\n",
    "## NOLB\n",
    "\n",
    "**NOn-Linear rigid Block NMA approach** (NOLB) is a conceptually simple and **computationally efficient** method for **non-linear normal mode analysis**.\n",
    "\n",
    "The key observation of the method is that the **angular velocity** of a **residue** can be interpreted as the result of an **implicit force**, such that the **motion of the residue** can be considered as a **pure rotation** about a **certain center**.\n",
    "\n",
    "References: <br>\n",
    "\n",
    "**NOLB website**: https://team.inria.fr/nano-d/software/nolb-normal-modes/\n",
    "\n",
    "**NOLB : Non-linear rigid block normal mode analysis method.**<br>\n",
    "*Journal of Chemical Theory and Computation, 2017, 13 (5), pp.2123-2134.*<br>\n",
    "*Available at: https://doi.org/10.1021/acs.jctc.7b00197*\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [nolb](https://biobb-flexdyn.readthedocs.io/en/latest/flexdyn.html#module-nolb.nolb) from **biobb_flexdyn.nolb.nolb**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexdyn.flexdyn.nolb_nma import nolb_nma\n",
    "\n",
    "nolb_pdb = pdbCode + '_nolb_ensemble.pdb'\n",
    "\n",
    "prop = {\n",
    "    'num_structs' : num_frames,\n",
    "    'rmsd' : 4\n",
    "}\n",
    "\n",
    "nolb_nma(   input_pdb_path=prot_ca,\n",
    "        output_pdb_path=nolb_pdb,\n",
    "        properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "nolb_rmsd = pdbCode + \"_nolb_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=nolb_pdb,\n",
    "            output_cpptraj_path=nolb_rmsd,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(nolb_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_convert](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_convert) from **biobb_analysis.ambertools.cpptraj_convert**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_convert import cpptraj_convert\n",
    "\n",
    "nolb_trr = pdbCode + '_nolb_ensemble.trr'\n",
    "\n",
    "prop = {\n",
    "    'mask' : 'c-alpha',\n",
    "    'format': 'trr'\n",
    "}\n",
    "\n",
    "cpptraj_convert(input_top_path=prot_ca,\n",
    "                input_traj_path=nolb_pdb,\n",
    "                output_cpptraj_path=nolb_trr,\n",
    "                properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(nolb_trr, prot_ca), gui=True)\n",
    "view.clear_representations()\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"imod\"></a>\n",
    "***\n",
    "## iMOD\n",
    "\n",
    "**iMOD** is an versatile toolkit to perform **Normal Mode Analysis (NMA)** in **internal coordinates (IC)** on both protein and nucleic acid atomic structures. **Vibrational analysis**, **motion animations**, **morphing trajectories** and **Monte-Carlo simulations** can be easily carried out at different scales of **resolution** using this toolkit. \n",
    "\n",
    "References: <br>\n",
    "\n",
    "**iMOD website**: https://chaconlab.org/multiscale-simulations/imod <br>\n",
    "**iMOD webserver**: https://imods.iqfr.csic.es/\n",
    "\n",
    "**iMod: multipurpose normal mode analysis in internal coordinates.**<br>\n",
    "*Bioinformatics, 2011, 27 (20): 2843-2850.*<br>\n",
    "*Available at: https://doi.org/10.1093/bioinformatics/btr497*\n",
    "\n",
    "**iMODS: Internal coordinates normal mode analysis server.**<br>\n",
    "*Nucleic acids research, 2014, 42:W271-6.*<br>\n",
    "*Available at: https://doi.org/10.1093/bioinformatics/btr497*\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [imod_imode](https://biobb-flexdyn.readthedocs.io/en/latest/flexdyn.html#module-flexdyn.imod_imode) from **biobb_flexdyn.flexdyn.imod_imode**\n",
    " - [imod_imc](https://biobb-flexdyn.readthedocs.io/en/latest/flexdyn.html#module-flexdyn.imod_imc) from **biobb_flexdyn.flexdyn.imod_imc**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_flexdyn.flexdyn.imod_imode import imod_imode\n",
    "\n",
    "imode_evecs = pdbCode + '_imode_evecs.dat'\n",
    "\n",
    "prop = {\n",
    "    'cg' : 2\n",
    "}\n",
    "\n",
    "imod_imode(  input_pdb_path=monomer,\n",
    "        output_dat_path=imode_evecs,\n",
    "        properties=prop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_flexdyn.flexdyn.imod_imc import imod_imc\n",
    "\n",
    "imc_pdb = pdbCode + '_imc.pdb'\n",
    "\n",
    "prop = {\n",
    "    'num_structs': num_frames,\n",
    "    'num_modes': 10,\n",
    "    'amplitude': 6.0\n",
    "}\n",
    "\n",
    "imod_imc(   input_pdb_path=monomer,\n",
    "            input_dat_path=imode_evecs,\n",
    "            output_traj_path=imc_pdb,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### RMSd distribution of the generated ensemble\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "imods_rmsd = pdbCode + \"_imods_rmsd.dat\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "\n",
    "cpptraj_rms(input_top_path=imc_pdb,\n",
    "            input_traj_path=imc_pdb,\n",
    "            output_cpptraj_path=imods_rmsd,\n",
    "            input_exp_path= monomer,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the RMSd distribution\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "df = pd.read_csv(imods_rmsd, header = 0, delimiter='\\s+')\n",
    "\n",
    "# Create a histogram plot\n",
    "fig = go.Figure(data=[go.Histogram(x=df['RMSD_00004'], xbins=dict(size=0.04), autobinx=False)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"RMSd variance\",\n",
    "                  xaxis_title=\"RMSd (Ångstroms)\",\n",
    "                  yaxis_title=\"Population\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_convert](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_convert) from **biobb_analysis.ambertools.cpptraj_convert**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_convert import cpptraj_convert\n",
    "\n",
    "imods_trr = pdbCode + '_imods_ensemble.trr'\n",
    "\n",
    "prop = {\n",
    "    'mask' : 'c-alpha',\n",
    "    'format': 'trr'\n",
    "}\n",
    "\n",
    "cpptraj_convert(input_top_path=imc_pdb,\n",
    "                input_traj_path=imc_pdb,\n",
    "                output_cpptraj_path=imods_trr,\n",
    "                properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(imods_trr, prot_ca), gui=True)\n",
    "view.clear_representations()\n",
    "view.add_representation(repr_type='ball+stick', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"metatraj\"></a>\n",
    "***\n",
    "## Meta-Trajectory & Clustering\n",
    "\n",
    "The generated **conformational ensembles** using the collection of **computational tools** can be used together  assembling a **meta-trajectory**. This **meta-trajectory** can then be **clusterized** to remove redundancies included in the structures. A meta-trajectory with results from different **computational methods** as different as the ones used in this **workflow** (Vibrational analysis, Coarse-grained molecular simulations, Atomistic intra-molecular interactions) can be helpful to enrich the **conformational sampling** and broadly populate the **conformational landscape**.    \n",
    "\n",
    "The next cells are generating a **meta-trajectory** from the set of **generated conformational ensembles**. Note that the ensembles used in this process need to be coherent on the **atomistic resolution** (they need to have the same **number of atoms**). As an example, if using **C-alpha only** structures, they all need to be **C-alpha only**. **Ensembles** used for the **meta-trajectory** creation can be easily modified in the next cell, commenting the lines of the unwanted files.  \n",
    "\n",
    "***\n",
    "\n",
    "### Building the meta-trajectory\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [trjcat](https://biobb-gromacs.readthedocs.io/en/latest/gromacs.html#module-gromacs.trjcat) from **biobb_gromacs.gromacs.trjcat**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "traj_zip = pdbCode + \"_concat_traj.zip\"\n",
    "\n",
    "with zipfile.ZipFile(traj_zip, 'w') as myzip:\n",
    "    myzip.write(concoord_trr)\n",
    "    myzip.write(prody_trr)\n",
    "    myzip.write(imods_trr)\n",
    "    #myzip.write(flexserv_bd_traj_fitted)\n",
    "    myzip.write(flexserv_dmd_traj_fitted)    \n",
    "    myzip.write(nma_trr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_gromacs.gromacs.trjcat import trjcat\n",
    "\n",
    "concat_trr = pdbCode + \"_concat_traj.trr\"\n",
    "\n",
    "trjcat(input_trj_zip_path=traj_zip,\n",
    "       output_trj_path=concat_trr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Clustering the meta-trajectory\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [make_ndx](https://biobb-gromacs.readthedocs.io/en/latest/gromacs.html#module-gromacs.make_ndx) from **biobb_gromacs.gromacs.make_ndx**\n",
    " - [gmx_cluster](https://biobb-analysis.readthedocs.io/en/latest/gromacs.html#module-gromacs.gmx_cluster) from **biobb_analysis.gromacs.gmx_cluster** \n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_gromacs.gromacs.make_ndx import make_ndx\n",
    "\n",
    "gmx_index_file = pdbCode + \"_gmx_ndx.ndx\"\n",
    "\n",
    "prop = { 'selection': 3 }\n",
    "\n",
    "make_ndx(input_structure_path=prot_ca,\n",
    "         output_ndx_path=gmx_index_file,\n",
    "         properties=prop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.gromacs.gmx_cluster import gmx_cluster\n",
    "\n",
    "cluster_concat_pdb = pdbCode + \"_concat_cluster.pdb\"\n",
    "\n",
    "prop = {\n",
    "    'fit_selection': 'System',\n",
    "    'output_selection': 'System',\n",
    "    'method': 'linkage',\n",
    "    'cutoff': 0.12 # (0.12 nm = 1.2 Ångstroms) \n",
    "    #'cutoff': 0.15 # (0.15 nm = 1.5 Ångstroms) \n",
    "}\n",
    "\n",
    "gmx_cluster(input_structure_path=prot_ca,\n",
    "            input_traj_path=concat_trr,\n",
    "            input_index_path=gmx_index_file,\n",
    "            output_pdb_path=cluster_concat_pdb,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (cluster)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show protein\n",
    "view = nglview.show_structure_file(cluster_concat_pdb, default_representation=False)\n",
    "view.add_representation(repr_type='tube', selection='all', color='modelindex')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"flexibility\"></a>\n",
    "***\n",
    "## Macromolecular Flexibility Analyses\n",
    "\n",
    "Once the final **ensemble** (meta-trajectory) is generated, it can be used to extract **flexibility/dynamic** patterns. A large variety of methods to **characterize flexibility** exist. In this **workflow**, the global movements of the structure will be analysed using **Essential Dynamics** (ED) routines (Amadei, et al., 1993), based in the well-known **Principal Components Analyisis** (PCA) statistical method.\n",
    "\n",
    "Other advanced tools allow the determination of **dynamic domains** and **hinge points** using a variety of techniques: i) exploration of the B-factor landscape after fitting with the gaussian RMSd method, ii) analysis of the force-constant profile (Sacquin-Mora and Lavery, 2006) and iii) clustering by inter-residue correlation (Navizet, et al., 2004). \n",
    "\n",
    "In all the analysis the resulting data is presented as a **json-formatted files**, and **2D plots** are generated when appropriate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"pca\"></a>\n",
    "***\n",
    "### Principal Component Analysis\n",
    "\n",
    "**Principal Component Analysis** is extensively used to characterize the most important **deformation modes** obtained by **diagonalization** of the **trajectory covariance matrix**. The **eigenvectors** represent the nature of the **essential deformation patterns**, while the **eigenvalues** can be transformed into the **frequencies** or **stiffness** of these movements. **Essential deformation movements** are ranked by importance and can be visualized and processed to obtain information (see Meyer, et al., 2006; Rueda, et al., 2007a for a detailed explanation), such as **B-Factor profiles**, the **collectivity index** (a measure of the collective nature of protein motions, Brüschweiler, 1995), the **variance profile**, the **dimensionality** (the number of movements defining a percentage of variance), or the **size of the essential space** (the number of modes with eigenvalues > 1 Å²). \n",
    "\n",
    "From the many available tools able to compute **PCA analysis** with **macromolecular data** (atomistic 3D coordinates), the **workflow** is using the **[pcasuite package](https://mmb.irbbarcelona.org/software/pcasuite/)**, integrated in the **BioBB library**. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Fitting and converting the generated ensemble (meta-trajectory) (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_rms](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_rms) from **biobb_analysis.ambertools.cpptraj_rms**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_rms import cpptraj_rms\n",
    "\n",
    "meta_traj_rmsd = pdbCode + \"_meta_traj_rmsd.dat\" \n",
    "meta_traj_fitted = pdbCode + \"_meta_traj_fitted.crd\"\n",
    "\n",
    "prop = {\n",
    "    'start': 1,\n",
    "    'end': -1,\n",
    "    'steps': 1,\n",
    "    'mask': 'c-alpha',\n",
    "    'reference': 'experimental'\n",
    "}\n",
    "cpptraj_rms(input_top_path=prot_ca,\n",
    "            input_traj_path=cluster_concat_pdb,\n",
    "            output_cpptraj_path=meta_traj_rmsd,\n",
    "            output_traj_path=meta_traj_fitted,\n",
    "            input_exp_path= prot_ca,\n",
    "            properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Computing the Principal Component Analysis (PCA)\n",
    "***\n",
    "\n",
    "**Building Block** used:\n",
    "\n",
    " - [pcz_zip](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#module-pcasuite.pcz_zip) from **biobb_flexserv.pcasuite.pcz_zip** \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_zip import pcz_zip\n",
    "\n",
    "concat_pcz = pdbCode + '_concat_ensemble.pcz'\n",
    "concat_pcz_gaussian = pdbCode + '_concat_ensemble_gaussian.pcz'\n",
    "\n",
    "# Classical RMSd fitting\n",
    "prop = {\n",
    "    'variance': 90,\n",
    "    'neigenv' : 10\n",
    "}\n",
    "\n",
    "pcz_zip( input_pdb_path=prot_ca,\n",
    "        input_crd_path=meta_traj_fitted,\n",
    "        output_pcz_path=concat_pcz,\n",
    "        properties=prop)\n",
    "\n",
    "# Gaussian (weighted) RMSd fitting\n",
    "prop = {\n",
    "    'variance': 90,\n",
    "    'neigenv' : 10,\n",
    "    'gauss_rmsd' : True\n",
    "}\n",
    "\n",
    "pcz_zip( input_pdb_path=prot_ca,\n",
    "        input_crd_path=meta_traj_fitted,\n",
    "        output_pcz_path=concat_pcz_gaussian,\n",
    "        properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Analysing the PCA report\n",
    "***\n",
    "\n",
    "The result of the **PCA analysis** is the generation of a set of **eigenvectors** (the modes or the principal components), which describe the nature of the **deformation movements** of the protein and a set of **eigenvalues**, which indicate the **stiffness** associated to every **mode**. \n",
    "\n",
    "The **eigenvectors** appear ranked after a **principal component analysis**, the first one is that explaining the **largest part of the variance** (as indicated by the associated **eigenvalue**). Since the **eigenvectors** represent a full-basis set, the **original Cartesian trajectory** can be always **projected** into the **eigenvectors space**, without lost of information. Furthermore, if a restricted set of **eigenvectors** is used, information is lost, but the level of error introduced in the simplification is always on user-control by considering the **annihilated variance** (the residual value between the **variance explained** by the set of the **eigenvectors** considered and the **total variance**).\n",
    "\n",
    "**Essential information** extracted from the **PCA analysis** can be retrieved from a **pcz file** with the **pcz_info building block**. Information such as **eigenvalues, number of eigenvectors, total variance, explained variance, quality or dimensionality** are included in the **PCA report**.  \n",
    "\n",
    "\n",
    "**Building Block** used:\n",
    " \n",
    " - [pcz_info](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#module-pcasuite.pcz_info) from **biobb_flexserv.pcasuite.pcz_info** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_info import pcz_info\n",
    "\n",
    "pcz_report = pdbCode + \"_pcz_report.json\"\n",
    "\n",
    "pcz_info( \n",
    "    input_pcz_path=concat_pcz,\n",
    "    output_json_path=pcz_report\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting Variance Profile\n",
    "***\n",
    "\n",
    "The following plot shows the **total variance** of the trajectory and how it is **distributed** along the different **eigenvectors**. **Variance** appears in Å² and **eigenvectors** are shown according to **eigenvalues** descending order, the first one being the **most important** one and the last that with the **lower contribution to variance**. This graph indicates the **size of the flexibility space** (the higher the variance, the higher the flexibility) and how it is distributed in different **modes**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pcz_report, 'r') as f:\n",
    "  pcz_info = json.load(f)\n",
    "print(json.dumps(pcz_info, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "# Plotting Variance Profile\n",
    "y = np.array(pcz_info['Eigen_Values'])\n",
    "x = list(range(1,len(y)+1))\n",
    "\n",
    "# Create a scatter plot\n",
    "fig = go.Figure(data=[go.Scatter(x=x, y=y)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Variance Profile\",\n",
    "                  xaxis_title=\"Principal Component\",\n",
    "                  yaxis_title=\"Variance\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting Dimensionality/quality profile\n",
    "***\n",
    "\n",
    "The following plot shows the **percentage** of **explained variance** for a given number of **eigenvectors** (quality) and the **dimensionality** of the sampled space. This graph indicates the **complexity of the flexibility space**, i.e. how many **modes** are necessary to explain the **entire flexibility** of the protein. \n",
    "\n",
    "Note that this plot and the previous one provide **physically-different** information and that proteins might display a very **complex pattern of flexibility** (leading to **large dimensionality**) and at the same time be **quite rigid** (**low variance**), or have a **large variance** which can be fully explained by a very **small number of modes**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "# Plotting Dimensionality/quality profile\n",
    "y = np.array(pcz_info['Eigen_Values_dimensionality_vs_total'])\n",
    "x = list(range(1,len(y)+1))\n",
    "\n",
    "# Create a scatter plot\n",
    "fig = go.Figure(data=[go.Scatter(x=x, y=y)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Dimensionality/Quality profile\",\n",
    "                  xaxis_title=\"Principal Component\",\n",
    "                  yaxis_title=\"Accumulated Quality (%)\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### PCA Eigenvectors & Eigenvalues \n",
    "***\n",
    "\n",
    "As stated above, the generated set of **eigenvectors** (the modes or the principal components) describe the nature of the **deformation movements** of the protein, whereas the **eigenvalues** indicate the **stiffness** associated to every **mode**. Inspection of the **atomic components** of **eigenvalues** associated to the most important **eigenvectors** helps to determine the **contribution** of different residues to the **key essential deformations** of the protein.\n",
    "\n",
    "The **pcz_evecs** building block returns the **atomic components** of the **eigenvalue** associated to a given **eigenvector**. \n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [pcz_evecs](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#module-pcasuite.pcz_evecs) from **biobb_flexserv.pcasuite.pcz_evecs**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_evecs import pcz_evecs\n",
    "\n",
    "pcz_evecs_report = pdbCode + \"_pcz_evecs.json\"\n",
    "\n",
    "prop = {\n",
    "    'eigenvector': 1\n",
    "}\n",
    "\n",
    "pcz_evecs( \n",
    "        input_pcz_path=concat_pcz,\n",
    "        output_json_path=pcz_evecs_report,\n",
    "        properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting Eigenvalue Residue Components \n",
    "\n",
    "Showing the **contribution** of different residues to the **key essential deformations** of the protein. The plot is displaying the **residue contributions** along the **first principal component**. The **eigenvector property** from the previous cell can be changed to explore the different **principal components**. \n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pcz_evecs_report, 'r') as f:\n",
    "  pcz_evecs_report_json = json.load(f)\n",
    "print(json.dumps(pcz_evecs_report_json, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "# Plotting Eigen Value Residue Components\n",
    "y = np.array(pcz_evecs_report_json['projs'])\n",
    "x = list(range(1,len(y)+1))\n",
    "\n",
    "# Create a bar plot\n",
    "fig = go.Figure(data=[go.Bar(x=x, y=y)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Eigen Value Residue Components\",\n",
    "                  xaxis_title=\"Residue Number\",\n",
    "                  yaxis_title=\"\\u00C5\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Animate Principal Components\n",
    "\n",
    "**Motions** described by the **eigenvectors** can be visualized by **projecting** the trajectory onto a given **eigenvector** and taking the **2 extreme projections** and **interpolating** between them to create an **animation**. This type of **visualization** is extremely popular as it allows a graphical an easy way to identify the **essential deformation movements** in macromolecules.\n",
    "\n",
    "The **pcz_animate** building block generates the **animation** of the **macromolecule** for a given **eigenvector**. \n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [pcz_animate](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#module-pcasuite.pcz_animate) from **biobb_flexserv.pcasuite.pcz_animate**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_animate import pcz_animate\n",
    "\n",
    "proj1 = pdbCode + \"_pcz_proj1.crd\"\n",
    "\n",
    "prop = {\n",
    "    'eigenvector': 1  # Try changing the eigenvector number!\n",
    "}\n",
    "\n",
    "pcz_animate( input_pcz_path=concat_pcz,\n",
    "        output_crd_path=proj1,\n",
    "        properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Converting the generated ensemble (for visualization purposes)\n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [cpptraj_convert](https://biobb-analysis.readthedocs.io/en/latest/ambertools.html#module-ambertools.cpptraj_convert) from **biobb_analysis.ambertools.cpptraj_convert**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from biobb_analysis.ambertools.cpptraj_convert import cpptraj_convert\n",
    "\n",
    "proj1_xtc = pdbCode + '_pcz_proj1.xtc'\n",
    "\n",
    "prop = {\n",
    "    'format': 'xtc'\n",
    "}\n",
    "\n",
    "cpptraj_convert(input_top_path=prot_ca,\n",
    "                input_traj_path=proj1,\n",
    "                output_cpptraj_path=proj1_xtc,\n",
    "                properties=prop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the generated ensemble (as a trajectory)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(proj1_xtc, prot_ca), gui=True)\n",
    "#view.add_representation(repr_type='spacefill', radius=0.7, selection='all')\n",
    "view.add_representation(repr_type='surface', selection='all')\n",
    "view.center()\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"bfactor\"></a>\n",
    "***\n",
    "## B-Factor x Principal Components\n",
    " \n",
    "The **B-factor** is the standard measure of **residue/atom flexibility**. It is determined from the **oscillations** of a **residue** with respect to its **equilibrium position**:\n",
    "\n",
    "**B-factor** profiles represent the distribution of residue **harmonic oscillations**. They can be compared with **X-ray data**, but caution is needed, since crystal **lattice effects** tend to **rigidify** exposed protein residues. Very large **B-factors** should be taken with caution since indicate very **flexible residues** that might display **conformational changes** along the trajectory, which is difficult to follow within the **harmonic approximation** implicit to **B-factor** analysis.\n",
    "\n",
    "The generated **PDB file** can be used to plot an **animation** of the **protein backbone** coloured by the **B-factor** corresponding to the selected **Principal Component**. Such visualization makes easier to evaluate which **region of the protein** is involved in the movement. \n",
    "\n",
    "The **pcz_bfactor** building block returns the **B-factor** values associated to a given **eigenvector**. \n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [pcz_bfactor](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#module-pcasuite.pcz_bfactor) from **biobb_flexserv.pcasuite.pcz_bfactor**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_bfactor import pcz_bfactor\n",
    "\n",
    "bfactor_all_dat = pdbCode + \"_bfactor_all.dat\"\n",
    "bfactor_all_pdb = pdbCode + \"_bfactor_all.pdb\"\n",
    "\n",
    "prop = {\n",
    "    'eigenvector': 0,\n",
    "    'pdb': True\n",
    "}\n",
    "\n",
    "pcz_bfactor( \n",
    "    input_pcz_path=concat_pcz,\n",
    "    output_dat_path=bfactor_all_dat,\n",
    "    output_pdb_path=bfactor_all_pdb,\n",
    "    properties=prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting B-factors x Residue x PCA mode\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "# Plotting the B-factors x Residue x PCA mode\n",
    "y = np.loadtxt(bfactor_all_dat)\n",
    "x = list(range(1,len(y)+1))\n",
    "\n",
    "# Create a scatter plot\n",
    "fig = go.Figure(data=[go.Scatter(x=x, y=y)])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Bfactor x Residue x PCA Modes (All)\",\n",
    "                  xaxis_title=\"Residue Number\",\n",
    "                  yaxis_title=\"Bfactor (\" + '\\u00C5' +'\\u00B2' + \")\",\n",
    "                  height=600)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the B-factor along the first principal component \n",
    "***\n",
    "\n",
    "Visualizing the **trajectory** highlighting the **residues** with higher **B-factor** values, the most **flexible residues** of the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view = nglview.show_simpletraj(nglview.SimpletrajTrajectory(proj1_xtc, bfactor_all_pdb))\n",
    "view.add_representation(repr_type='spacefill', selection='all', colorScheme='bfactor')\n",
    "view.add_representation(repr_type='tube', radius='0.4', selection='all', color='white')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "\n",
    "stop = False\n",
    "def loop(view):\n",
    "    import time\n",
    "    def do():\n",
    "        while True and not stop:\n",
    "            if view.frame == view.max_frame:\n",
    "                direction = -1\n",
    "            if view.frame == 0:\n",
    "                direction = 1\n",
    "            view.frame = view.frame + direction\n",
    "            time.sleep(0.2)\n",
    "    view._run_on_another_thread(do)\n",
    "        \n",
    "#view.on_displayed(loop)\n",
    "\n",
    "#view._iplayer.children[0].disabled = True\n",
    "#view._iplayer.children[1].disabled = True\n",
    "\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"hinges\"></a>\n",
    "***\n",
    "## Hinge Points Prediction\n",
    "\n",
    "**Hinge point detection** is a process to determine residues around which **large protein movements** are organized. Analysis implemented in the **biobb_flexserv module** can be performed using **three different methodologies**: *B-Factor slope change method*, *Force constant method* and *Dynamic domain detection method*.\n",
    "\n",
    "The **pcz_hinges** building block computes the **hinge points** of the **macromolecules** associated to a given **eigenvector**. \n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [pcz_hinges](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#module-pcasuite.pcz_hinges) from **biobb_flexserv.pcasuite.pcz_hinges**\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_hinges import pcz_hinges\n",
    "\n",
    "hinges_bfactor_report = pdbCode + \"_hinges_bfactor_report.json\"\n",
    "hinges_dyndom_report = pdbCode + \"_hinges_dyndom_report.json\"\n",
    "hinges_fcte_report = pdbCode + \"_hinges_fcte_report.json\"\n",
    "\n",
    "bfactor_method = \"Bfactor_slope\"\n",
    "dyndom_method = \"Dynamic_domain\"\n",
    "fcte_method = \"Force_constant\"\n",
    "\n",
    "bfactor_prop = {\n",
    "    'eigenvector': 0, # 0 = All modes\n",
    "    'method': bfactor_method\n",
    "}\n",
    "\n",
    "dyndom_prop = {\n",
    "    'eigenvector': 0, # 0 = All modes\n",
    "    'method': dyndom_method\n",
    "}\n",
    "\n",
    "fcte_prop = {\n",
    "    'eigenvector': 0, # 0 = All modes\n",
    "    'method': fcte_method\n",
    "}\n",
    "\n",
    "pcz_hinges( \n",
    "        input_pcz_path=concat_pcz_gaussian,\n",
    "        output_json_path=hinges_bfactor_report,\n",
    "        properties=bfactor_prop\n",
    ")\n",
    "\n",
    "pcz_hinges( \n",
    "        input_pcz_path=concat_pcz_gaussian,\n",
    "        output_json_path=hinges_dyndom_report,\n",
    "        properties=dyndom_prop\n",
    ")\n",
    "\n",
    "pcz_hinges( \n",
    "        input_pcz_path=concat_pcz_gaussian,\n",
    "        output_json_path=hinges_fcte_report,\n",
    "        properties=fcte_prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Visualizing the hinge points predictions\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "with open(hinges_bfactor_report, 'r') as f:\n",
    "  hinges_bfactor = json.load(f)\n",
    "print(json.dumps(hinges_bfactor, indent=2))\n",
    "\n",
    "with open(hinges_dyndom_report, 'r') as f:\n",
    "  hinges_dyndom = json.load(f)\n",
    "print(json.dumps(hinges_dyndom, indent=2))\n",
    "\n",
    "with open(hinges_fcte_report, 'r') as f:\n",
    "  hinges_fcte = json.load(f)\n",
    "print(json.dumps(hinges_fcte, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show trajectory\n",
    "view1 = nglview.show_simpletraj(nglview.SimpletrajTrajectory(proj1_xtc, bfactor_all_pdb), gui=True)\n",
    "view1.add_representation(repr_type='surface', selection=hinges_dyndom[\"clusters\"][0][\"residues\"], color='red')\n",
    "view1.add_representation(repr_type='surface', selection=hinges_dyndom[\"clusters\"][1][\"residues\"], color='green')\n",
    "#view1.add_representation(repr_type='surface', selection=hinges_dyndom[\"clusters\"][2][\"residues\"], color='yellow')\n",
    "#view1.add_representation(repr_type='surface', selection=hinges_dyndom[\"hinge_residues\"], color='red')\n",
    "view1._remote_call('setSize', target='Widget', args=['350px','350px'])\n",
    "view1\n",
    "view2 = nglview.show_simpletraj(nglview.SimpletrajTrajectory(proj1_xtc, bfactor_all_pdb), gui=True)\n",
    "view2.add_representation(repr_type='surface', selection=hinges_bfactor[\"hinge_residues\"], color='red')\n",
    "view2._remote_call('setSize', target='Widget', args=['350px','350px'])\n",
    "view2\n",
    "view3 = nglview.show_simpletraj(nglview.SimpletrajTrajectory(proj1_xtc, bfactor_all_pdb), gui=True)\n",
    "view3.add_representation(repr_type='surface', selection=str(hinges_fcte[\"hinge_residues\"]), color='red')    \n",
    "view3._remote_call('setSize', target='Widget', args=['350px','350px'])\n",
    "view3\n",
    "ipywidgets.HBox([view1, view2, view3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_stiffness import pcz_stiffness\n",
    "\n",
    "stiffness_report = pdbCode + \"_pcz_stiffness.json\"\n",
    "\n",
    "prop = {\n",
    "    'eigenvector': 0 # 0 = All modes\n",
    "}\n",
    "\n",
    "pcz_stiffness( \n",
    "        input_pcz_path=concat_pcz,\n",
    "        output_json_path=stiffness_report,\n",
    "        properties=prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"stiffness\"></a>\n",
    "***\n",
    "## Apparent Stiffness\n",
    "\n",
    "**Stiffness** is defined as the **force-constant** acting between **two residues** in the case of completely disconnected oscillators. The index helps to detect **strong interactions between residues**, which might indicate physically-intense **direct contacts** or **strong chain-related interactions**. Results are usually plot as a **2D NxN heatmap plot** (with N being the number of residues).\n",
    "\n",
    "The **pcz_stiffness** building block returns the **stiffness** force-constants associated to a given **eigenvector**. \n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [pcz_stiffness](https://biobb-flexserv.readthedocs.io/en/latest/pcasuite.html#pcasuite-pcz-stiffness-module) from **biobb_flexserv.pcasuite.pcz_stiffness**\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting the Apparent Stiffness matrix\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "with open(stiffness_report, 'r') as f:\n",
    "  pcz_stiffness_report = json.load(f)\n",
    "print(json.dumps(pcz_stiffness_report, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "y = np.array(pcz_stiffness_report['stiffness'])\n",
    "x = list(range(1,len(y)))\n",
    "\n",
    "# Create a heatmap plot\n",
    "fig = go.Figure(data=[go.Heatmap(x=x, y=x, z=y, type = 'heatmap', colorscale = 'reds')])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Apparent Stiffness\",\n",
    "                  xaxis_title=\"Residue Number\",\n",
    "                  yaxis_title=\"Residue Number\",\n",
    "                  width=800,\n",
    "                  height=800)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting the Apparent Stiffness matrix (Logarithmic Scale)\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "y = np.array(pcz_stiffness_report['stiffness_log'])\n",
    "x = list(range(1,len(y)))\n",
    "\n",
    "# Create a heatmap plot\n",
    "fig = go.Figure(data=[go.Heatmap(x=x, y=x, z=y, type = 'heatmap', colorscale = 'reds')])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Apparent Stiffness (Logarithmic Scale)\",\n",
    "                  xaxis_title=\"Residue Number\",\n",
    "                  yaxis_title=\"Residue Number\",\n",
    "                  width=800,\n",
    "                  height=800)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"collectivity\"></a>\n",
    "***\n",
    "## Collectivity Index\n",
    "\n",
    "**Collectivity index** is a numerical measure of **how many atoms are affected by a given mode** (see the work by Rafael Brüschweiler: [Collective protein dynamics and nuclear spin relaxation](https://doi.org/10.1063/1.469213)).\n",
    "\n",
    "The **pcz_collectivity** building block returns the **collectivity index** of the **macromolecule** associated to a given **eigenvector**. \n",
    "\n",
    "***\n",
    "**Building Blocks** used:\n",
    " - [pcz_collectivity](https://biobb-flexserv.readthedocs.io/en/latest/flexserv.html#module-pcasuite.pcz_collectivity) from **biobb_flexserv.pcasuite.pcz_collectivity**\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biobb_flexserv.pcasuite.pcz_collectivity import pcz_collectivity\n",
    "\n",
    "pcz_collectivity_report = pdbCode + \"_pcz_collectivity.json\"\n",
    "\n",
    "prop = {\n",
    "    'eigenvector':0 # 0 = All modes\n",
    "}\n",
    "\n",
    "pcz_collectivity( \n",
    "    input_pcz_path=concat_pcz,\n",
    "    output_json_path=pcz_collectivity_report,\n",
    "    properties=prop\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### Plotting the Collectivity indexes associated to the first 10 PCA modes \n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pcz_collectivity_report, 'r') as f:\n",
    "  pcz_collectivity_report_json = json.load(f)\n",
    "print(json.dumps(pcz_collectivity_report_json, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in sys.modules: use_colab_plotly()\n",
    "\n",
    "z = np.array(pcz_collectivity_report_json['collectivity'])\n",
    "x = list(range(1,len(z)+1))\n",
    "x = [\"PC\" + str(pc) for pc in x]\n",
    "\n",
    "y = [\"\"]\n",
    "\n",
    "# Create a heatmap plot\n",
    "fig = go.Figure(data=[go.Heatmap(x=x, y=y, z=[z], type = 'heatmap', colorscale = 'reds')])\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(title=\"Collectivity Index\",\n",
    "                  xaxis_title=\"Principal Component\",\n",
    "                  yaxis_title=\"Collectivity\",\n",
    "                  width=1000,\n",
    "                  height=300)\n",
    "\n",
    "# Show the figure (renderer changes for colab and jupyter)\n",
    "rend = 'colab' if 'google.colab' in sys.modules else ''\n",
    "fig.show(renderer=rend)\n",
    "\n",
    "if 'google.colab' in sys.modules: use_conda_env()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<a id=\"questions\"></a>\n",
    "\n",
    "## Questions & Comments\n",
    "\n",
    "Questions, issues, suggestions and comments are really welcome!\n",
    "\n",
    "* GitHub issues:\n",
    "    * [https://github.com/bioexcel/biobb](https://github.com/bioexcel/biobb)\n",
    "\n",
    "* BioExcel forum:\n",
    "    * [https://ask.bioexcel.eu/c/BioExcel-Building-Blocks-library](https://ask.bioexcel.eu/c/BioExcel-Building-Blocks-library)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
